{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Capstone Project - Connections between Lifestyle & Disease"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data wrangling & Exploratory Data Analysis (EDA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages.\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3071: DtypeWarning: Columns (34,197) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    }
   ],
   "source": [
    "# Import data.\n",
    "\n",
    "df = pd.read_csv(\"AG_copy_1.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>NONFOODALLERGIES_DRUG</th>\n",
       "      <th>NONFOODALLERGIES_BEESTINGS</th>\n",
       "      <th>ASSIGNED_FROM_GEO</th>\n",
       "      <th>RUN_DATE</th>\n",
       "      <th>PET_LOCATIONS</th>\n",
       "      <th>AGE</th>\n",
       "      <th>...</th>\n",
       "      <th>LIBRARY_CONSTRUCTION_PROTOCOL</th>\n",
       "      <th>PKU</th>\n",
       "      <th>DECEASED_PARENT</th>\n",
       "      <th>SEQ_RUN</th>\n",
       "      <th>CAT</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>FOODALLERGIES_PEANUTS</th>\n",
       "      <th>CHICKENPOX</th>\n",
       "      <th>SEASONAL_ALLERGIES</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11224.12109</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>55</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>51</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>39.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5034.107635</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>56</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>32.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11351.12107</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>16</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>56</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>40</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2013.113155</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>40.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4731.107641</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>61</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>72</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>43.1</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #SampleID BarcodeSequence LinkerPrimerSequence CARBOHYDRATE_PER  \\\n",
       "0  11224.12109         no_data              no_data               55   \n",
       "1  5034.107635         no_data              no_data               56   \n",
       "2  11351.12107         no_data              no_data               16   \n",
       "3  2013.113155         no_data              no_data          no_data   \n",
       "4  4731.107641         no_data              no_data               61   \n",
       "\n",
       "  NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS ASSIGNED_FROM_GEO RUN_DATE  \\\n",
       "0               no_data                    no_data                 n  no_data   \n",
       "1               no_data                    no_data                 n  no_data   \n",
       "2               no_data                    no_data                 n  no_data   \n",
       "3               no_data                    no_data                 n  no_data   \n",
       "4                   yes                    no_data                 n  no_data   \n",
       "\n",
       "  PET_LOCATIONS AGE  ... LIBRARY_CONSTRUCTION_PROTOCOL PKU DECEASED_PARENT  \\\n",
       "0    Housing...  51  ...                       no_data  no         no_data   \n",
       "1       no_data  35  ...                       no_data  no         no_data   \n",
       "2    Housing...  56  ...                       no_data  no         no_data   \n",
       "3    Housing...  37  ...                       no_data  no         no_data   \n",
       "4       no_data  72  ...                       no_data  no         no_data   \n",
       "\n",
       "   SEQ_RUN  CAT LATITUDE FOODALLERGIES_PEANUTS CHICKENPOX SEASONAL_ALLERGIES  \\\n",
       "0  no_data  yes     39.7               no_data        yes                 no   \n",
       "1  no_data  yes     32.7               no_data        yes                 no   \n",
       "2  no_data   no       40               no_data        yes                 no   \n",
       "3  no_data   no     40.7               no_data        yes                 no   \n",
       "4  no_data   no     43.1               no_data        yes                yes   \n",
       "\n",
       "                         Description  \n",
       "0  American Gut Project Stool sample  \n",
       "1  American Gut Project Stool sample  \n",
       "2  American Gut Project Stool sample  \n",
       "3  American Gut Project Stool sample  \n",
       "4  American Gut Project Stool sample  \n",
       "\n",
       "[5 rows x 202 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show top of data frame.\n",
    "\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show object type.\n",
    "\n",
    "type(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4815, 202)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show shape of dataframe. \n",
    "\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data set has 4815 rows and 202 features. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#SampleID                object\n",
       "BarcodeSequence          object\n",
       "LinkerPrimerSequence     object\n",
       "CARBOHYDRATE_PER         object\n",
       "NONFOODALLERGIES_DRUG    object\n",
       "                          ...  \n",
       "LATITUDE                 object\n",
       "FOODALLERGIES_PEANUTS    object\n",
       "CHICKENPOX               object\n",
       "SEASONAL_ALLERGIES       object\n",
       "Description              object\n",
       "Length: 202, dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show the data types of all columns.\n",
    "\n",
    "df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#SampleID                object\n",
       "BarcodeSequence          object\n",
       "LinkerPrimerSequence     object\n",
       "CARBOHYDRATE_PER         object\n",
       "NONFOODALLERGIES_DRUG    object\n",
       "                          ...  \n",
       "LATITUDE                 object\n",
       "FOODALLERGIES_PEANUTS    object\n",
       "CHICKENPOX               object\n",
       "SEASONAL_ALLERGIES       object\n",
       "Description              object\n",
       "Length: 200, dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show all the non-numerical/categorical columns.\n",
    "\n",
    "df.dtypes[df.dtypes != 'int64'][df.dtypes != 'float64']\n",
    "\n",
    "# https://stackoverflow.com/questions/40353079/pandas-how-to-check-dtype-for-all-columns-in-a-dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "200 features are categorical (object)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ALTITUDE    int64\n",
       "DEPTH       int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show all the numerical columns.\n",
    "\n",
    "df.dtypes[df.dtypes != 'object']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2 features are numerical (int64). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ALTITUDE</th>\n",
       "      <th>DEPTH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>4815.000000</td>\n",
       "      <td>4815.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.014411</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          ALTITUDE   DEPTH\n",
       "count  4815.000000  4815.0\n",
       "mean      0.000208     0.0\n",
       "std       0.014411     0.0\n",
       "min       0.000000     0.0\n",
       "25%       0.000000     0.0\n",
       "50%       0.000000     0.0\n",
       "75%       0.000000     0.0\n",
       "max       1.000000     0.0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show info about numerical columns.\n",
    "\n",
    "df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      ALTITUDE  DEPTH\n",
      "0            0      0\n",
      "1            0      0\n",
      "2            0      0\n",
      "3            0      0\n",
      "4            0      0\n",
      "...        ...    ...\n",
      "4810         0      0\n",
      "4811         0      0\n",
      "4812         0      0\n",
      "4813         0      0\n",
      "4814         0      0\n",
      "\n",
      "[4815 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# Show vales of numerical columns. \n",
    "\n",
    "print(df[['ALTITUDE', 'DEPTH']])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Initial inspection shows that the altitude and depth columns might be (mostly or only) zeros."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count amount of distinct values in a column.\n",
    "\n",
    "df['ALTITUDE'].nunique()\n",
    "\n",
    "# https://stackoverflow.com/questions/45759966/counting-unique-values-in-a-column-in-pandas-dataframe-like-in-qlik"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    4814\n",
       "1       1\n",
       "Name: ALTITUDE, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show frequency of distinct values in a column.\n",
    "\n",
    "df['ALTITUDE'].value_counts()\n",
    "\n",
    "# https://stackoverflow.com/questions/22391433/count-the-frequency-that-a-value-occurs-in-a-dataframe-column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count amount of distinct values in a column.\n",
    "\n",
    "df['DEPTH'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    4815\n",
       "Name: DEPTH, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show frequency of distinct values in a column.\n",
    "\n",
    "df['DEPTH'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Further inspection shows that the altitude and depth columns are (mostly or only) zeros. Therefore these columns will be deleted."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>NONFOODALLERGIES_DRUG</th>\n",
       "      <th>NONFOODALLERGIES_BEESTINGS</th>\n",
       "      <th>ASSIGNED_FROM_GEO</th>\n",
       "      <th>RUN_DATE</th>\n",
       "      <th>PET_LOCATIONS</th>\n",
       "      <th>AGE</th>\n",
       "      <th>...</th>\n",
       "      <th>LIBRARY_CONSTRUCTION_PROTOCOL</th>\n",
       "      <th>PKU</th>\n",
       "      <th>DECEASED_PARENT</th>\n",
       "      <th>SEQ_RUN</th>\n",
       "      <th>CAT</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>FOODALLERGIES_PEANUTS</th>\n",
       "      <th>CHICKENPOX</th>\n",
       "      <th>SEASONAL_ALLERGIES</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>...</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>size</th>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>...</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "      <td>4815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nunique</th>\n",
       "      <td>4815</td>\n",
       "      <td>257</td>\n",
       "      <td>2</td>\n",
       "      <td>159</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>48</td>\n",
       "      <td>90</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>430</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #SampleID  BarcodeSequence  LinkerPrimerSequence  CARBOHYDRATE_PER  \\\n",
       "count         4815             4815                  4815              4815   \n",
       "size          4815             4815                  4815              4815   \n",
       "nunique       4815              257                     2               159   \n",
       "\n",
       "         NONFOODALLERGIES_DRUG  NONFOODALLERGIES_BEESTINGS  ASSIGNED_FROM_GEO  \\\n",
       "count                     4815                        4815               4815   \n",
       "size                      4815                        4815               4815   \n",
       "nunique                      3                           3                  2   \n",
       "\n",
       "         RUN_DATE  PET_LOCATIONS   AGE  ...  LIBRARY_CONSTRUCTION_PROTOCOL  \\\n",
       "count        4815           4815  4815  ...                           4815   \n",
       "size         4815           4815  4815  ...                           4815   \n",
       "nunique         3             48    90  ...                              2   \n",
       "\n",
       "          PKU  DECEASED_PARENT  SEQ_RUN   CAT  LATITUDE  \\\n",
       "count    4815             4815     4815  4815      4815   \n",
       "size     4815             4815     4815  4815      4815   \n",
       "nunique     4                3        5     4       430   \n",
       "\n",
       "         FOODALLERGIES_PEANUTS  CHICKENPOX  SEASONAL_ALLERGIES  Description  \n",
       "count                     4815        4815                4815         4815  \n",
       "size                      4815        4815                4815         4815  \n",
       "nunique                      3           4                   4           39  \n",
       "\n",
       "[3 rows x 202 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count only non-null values, use 'count'.\n",
    "# Count total values including null values, use 'size'.\n",
    "# Count distinct values, use 'nunique'.\n",
    "\n",
    "\n",
    "df_agg = df.agg(['count', 'size', 'nunique'])\n",
    "df_agg\n",
    "\n",
    "# https://stackoverflow.com/questions/45759966/counting-unique-values-in-a-column-in-pandas-dataframe-like-in-qlik"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         #SampleID  BarcodeSequence  LinkerPrimerSequence  CARBOHYDRATE_PER  \\\n",
      "count         4815             4815                  4815              4815   \n",
      "size          4815             4815                  4815              4815   \n",
      "nunique       4815              257                     2               159   \n",
      "\n",
      "         NONFOODALLERGIES_DRUG  NONFOODALLERGIES_BEESTINGS  ASSIGNED_FROM_GEO  \\\n",
      "count                     4815                        4815               4815   \n",
      "size                      4815                        4815               4815   \n",
      "nunique                      3                           3                  2   \n",
      "\n",
      "         RUN_DATE  PET_LOCATIONS   AGE  TOT_MASS  GENERAL_MEDS  SAMPLE_ID  \\\n",
      "count        4815           4815  4815      4815          4815       4815   \n",
      "size         4815           4815  4815      4815          4815       4815   \n",
      "nunique         3             48    90       132             3       3737   \n",
      "\n",
      "         ORIG_SAMPLE_NAME  QUINOLINE  BODY_SITE  MIGRAINE_FACTOR_2  \\\n",
      "count                4815       4815       4815               4815   \n",
      "size                 4815       4815       4815               4815   \n",
      "nunique               749          7         10                 12   \n",
      "\n",
      "         MIGRAINE_FACTOR_3  TONSILS_REMOVED  MIGRAINE_FACTOR_1  ELEVATION  \\\n",
      "count                 4815             4815               4815       4815   \n",
      "size                  4815             4815               4815       4815   \n",
      "nunique                 12                4                 12       1014   \n",
      "\n",
      "         FLOSSING_FREQUENCY  PROTEIN_PER  CSECTION  PCR_PRIMERS  \\\n",
      "count                  4815         4815      4815         4815   \n",
      "size                   4815         4815      4815         4815   \n",
      "nunique                   7          141         5            2   \n",
      "\n",
      "         COUNTRY_OF_BIRTH  MACRONUTRIENT_PCT_TOTAL  COSMETICS_FREQUENCY  \\\n",
      "count                4815                     4815                 4815   \n",
      "size                 4815                     4815                 4815   \n",
      "nunique                65                       74                    7   \n",
      "\n",
      "         MASTERMIX_LOT  HOST_SUBJECT_ID  ENA-CHECKLIST  GLUTEN  SAMPLE_NAME  \\\n",
      "count             4815             4815           4815    4815         4815   \n",
      "size              4815             4815           4815    4815         4815   \n",
      "nunique              8             3831              2       4         3620   \n",
      "\n",
      "         TM1000_8_TOOL  LONGITUDE   DOG  MIGRAINE_AGGRAVATION  ORIG_NAME  \\\n",
      "count             4815       4815  4815                  4815       4815   \n",
      "size              4815       4815  4815                  4815       4815   \n",
      "nunique              5        694     4                     3        900   \n",
      "\n",
      "         FRAT  DIABETES_MEDICATION  SAMPLE_PLATE  PENICILLIN  \\\n",
      "count    4815                 4815          4815        4815   \n",
      "size     4815                 4815          4815        4815   \n",
      "nunique     4                    4            45           5   \n",
      "\n",
      "         FOODALLERGIES_OTHER_TEXT  KEY_SEQ  BODY_PRODUCT  REGION  SAMPLE_TYPE  \\\n",
      "count                        4814     4815          4815    4815         4815   \n",
      "size                         4815     4815          4815    4815         4815   \n",
      "nunique                       403        2             6       2           12   \n",
      "\n",
      "         MIGRAINE_FREQUENCY  MIGRAINE_RELATIVES  EXPERIMENT_TITLE  PLANT_PER  \\\n",
      "count                  4815                4815              4815       4815   \n",
      "size                   4815                4815              4815       4815   \n",
      "nunique                   6                   5                 2        112   \n",
      "\n",
      "         EXTRACTIONKIT_LOT  SAMPLE  MIGRAINE_PHOTOPHOBIA  SITE_SAMPLED  \\\n",
      "count                 4815    4815                  4815          4815   \n",
      "size                  4815    4815                  4815          4815   \n",
      "nunique                284     638                     3            13   \n",
      "\n",
      "         DRINKING_WATER_SOURCE  CENTER_NAME  COUNTRY  ANTIBIOTIC_SELECT  \\\n",
      "count                     4815         4815     4815               4815   \n",
      "size                      4815         4815     4815               4815   \n",
      "nunique                      7            2       19                  7   \n",
      "\n",
      "         MIGRAINE_NAUSEA  COMMON_NAME  FOODALLERGIES_TREENUTS  MIGRAINE_AURA  \\\n",
      "count               4815         4815                    4815           4815   \n",
      "size                4815         4815                    4815           4815   \n",
      "nunique                3            8                       3              3   \n",
      "\n",
      "         SEQUENCING_METH  FLU_VACCINE_DATE  PREGNANT  MIGRAINE   BMI  \\\n",
      "count               4815              4815      4815      4815  4815   \n",
      "size                4815              4815      4815      4815  4815   \n",
      "nunique                2                 7         5         4  1202   \n",
      "\n",
      "         ENV_BIOME  WELL_ID  TEETHBRUSHING_FREQUENCY  MIGRAINE_MEDICATIONS  \\\n",
      "count         4815     4815                     4815                  4815   \n",
      "size          4815     4815                     4815                  4815   \n",
      "nunique          4       97                        7                    84   \n",
      "\n",
      "         CONTRACEPTIVE  SAMPLE_CENTER  PHYSICAL_LOCATION  LIVINGWITH  LINKER  \\\n",
      "count             4815           4815               4815        4815    4815   \n",
      "size              4815           4815               4815        4815    4815   \n",
      "nunique              8              2                  3           5       2   \n",
      "\n",
      "         RELATIONS  DIABETES_MEDICATIONS  SUPPLEMENTS  AGE_UNIT  \\\n",
      "count         4815                  4815         4815      4815   \n",
      "size          4815                  4815         4815      4815   \n",
      "nunique        292                    38            4         2   \n",
      "\n",
      "         TANNING_SPRAYS  EXERCISE_FREQUENCY  HOST_COMMON_NAME  SULFA_DRUG  \\\n",
      "count              4815                4815              4815        4815   \n",
      "size               4815                4815              4815        4815   \n",
      "nunique               4                   7                 6           5   \n",
      "\n",
      "         NONFOODALLERGIES_NO  ENV_MATTER  RACE_OTHER  SKIN_CONDITION  \\\n",
      "count                   4815        4815        4815            4815   \n",
      "size                    4815        4815        4815            4815   \n",
      "nunique                    3           6          56               6   \n",
      "\n",
      "         ALTITUDE  NONFOODALLERGIES_DANDER  CONDITIONS_MEDICATION  \\\n",
      "count        4815                     4815                   4815   \n",
      "size         4815                     4815                   4815   \n",
      "nunique         2                        3                      4   \n",
      "\n",
      "         LAST_TRAVEL  TARGET_SUBFRAGMENT  TITLE  HEIGHT_IN  \\\n",
      "count           4815                4815   4815       4815   \n",
      "size            4815                4815   4815       4815   \n",
      "nunique            8                   2      1         80   \n",
      "\n",
      "         ANTIBIOTIC_CONDITION  PET_CONTACT  MIGRAINE_PAIN  ROOMMATES  \\\n",
      "count                    4814         4815           4815       4815   \n",
      "size                     4815         4815           4815       4815   \n",
      "nunique                   522           69              3          7   \n",
      "\n",
      "         CENTER_PROJECT_NAME  NONFOODALLERGIES_SUN  HAS_EXTRACTED_DATA  \\\n",
      "count                   4815                  4815                4815   \n",
      "size                    4815                  4815                4815   \n",
      "nunique                    2                     3                   3   \n",
      "\n",
      "         ENA-SPOT-COUNT  EXERCISE_LOCATION  RUN_ID  COLLECTION_DATE  \\\n",
      "count              4815               4815    4815             4815   \n",
      "size               4815               4815    4815             4815   \n",
      "nunique            3702                  7       3              655   \n",
      "\n",
      "         NITROIMIDAZOLE  DEODORANT_USE  FOODALLERGIES_SHELLFISH   SEX  \\\n",
      "count              4815           4815                     4815  4815   \n",
      "size               4815           4815                     4815  4815   \n",
      "nunique               6              9                        3     5   \n",
      "\n",
      "         PERCENTAGE_FROM_CARBS  FAT_PER  TARGET_GENE  PREGNANT_DUE_DATE  \\\n",
      "count                     4815     4815         4815               4815   \n",
      "size                      4815     4815         4815               4815   \n",
      "nunique                      7      164            2                 22   \n",
      "\n",
      "         SHARED_HOUSING  SOFTENER  MIGRAINE_PHONOPHOBIA  SAMPLE_DESCRIPTION  \\\n",
      "count              4815      4815                  4815                4815   \n",
      "size               4815      4815                  4815                4815   \n",
      "nunique               4         5                     3                 982   \n",
      "\n",
      "         SAMP_SIZE  SMOKING_FREQUENCY  WEIGHT_LBS  TANNING_BEDS  \\\n",
      "count         4815               4815        4815          4815   \n",
      "size          4815               4815        4815          4815   \n",
      "nunique          4                  7         245             3   \n",
      "\n",
      "         ALCOHOL_FREQUENCY  STATE  HAS_PHYSICAL_SPECIMEN  STUDY_CENTER  \\\n",
      "count                 4815   4811                   4815          4815   \n",
      "size                  4815   4815                   4815          4815   \n",
      "nunique                  7    119                      3             2   \n",
      "\n",
      "         BODY_HABITAT  DOMINANT_HAND  PRIMARY_VEGETABLE  WATER_LOT  \\\n",
      "count            4815           4815               4813       4815   \n",
      "size             4815           4815               4815       4815   \n",
      "nunique             8              5                582        580   \n",
      "\n",
      "         SPECIAL_RESTRICTIONS  CEPHALOSPORIN  ENV_FEATURE  SUPPLEMENTS_FIELDS  \\\n",
      "count                    4815           4815         4815                4815   \n",
      "size                     4815           4815         4815                4815   \n",
      "nunique                     4              5            3                1821   \n",
      "\n",
      "         ENA-BASE-COUNT  SAMPLE_TIME  APPENDIX_REMOVED  \\\n",
      "count              4815         4815              4815   \n",
      "size               4815         4815              4815   \n",
      "nunique            3816          383                 4   \n",
      "\n",
      "         CURRENT_RESIDENCE_DURATION  EMP_STATUS  COMMUNAL_DINING  \\\n",
      "count                          4815        4815             4815   \n",
      "size                           4815        4815             4815   \n",
      "nunique                           7           2                4   \n",
      "\n",
      "         SLEEP_DURATION  ANONYMIZED_NAME  PUBLIC  DIET_TYPE  RUN_CENTER  \\\n",
      "count              4815             4815    4815       4815        4815   \n",
      "size               4815             4815    4815       4815        4815   \n",
      "nunique               7             4670       2          7           2   \n",
      "\n",
      "         DIET_RESTRICTIONS  EXPERIMENT_CENTER  LACTOSE  HEIGHT_OR_LENGTH  \\\n",
      "count                 4813               4815     4815              4815   \n",
      "size                  4815               4815     4815              4815   \n",
      "nunique                802                  2        4               115   \n",
      "\n",
      "         ANTIBIOTIC_MEDS   IBD  TM50_8_TOOL  DEPTH  PROCESSING_ROBOT  \\\n",
      "count               4815  4815         4815   4815              4815   \n",
      "size                4815  4815         4815   4815              4815   \n",
      "nunique              441     5            8      1                 7   \n",
      "\n",
      "         HOST_TAXID  ILLUMINA_TECHNOLOGY  TAXON_ID  DIABETES_DIAGNOSE_DATE  \\\n",
      "count          4815                 4815      4815                    4815   \n",
      "size           4815                 4815      4815                    4815   \n",
      "nunique           3                    2         7                      76   \n",
      "\n",
      "         REQUIRED_SAMPLE_INFO_STATUS  TYPES_OF_PLANTS  MIGRAINEMEDS  \\\n",
      "count                           4815             4815          4815   \n",
      "size                            4815             4815          4815   \n",
      "nunique                            2                8             4   \n",
      "\n",
      "         FIBER_GRAMS  MAINFACTOR_OTHER_3  PLATING  PROJECT_NAME  \\\n",
      "count           4815                4815     4815          4815   \n",
      "size            4815                4815     4815          4815   \n",
      "nunique          602                  53        7             2   \n",
      "\n",
      "         COLLECTION_TIMESTAMP  NAILS  RUN_PREFIX  DESCRIPTION_WELL  PLATFORM  \\\n",
      "count                    4815   4815        4815              4815      4815   \n",
      "size                     4815   4815        4815              4815      4815   \n",
      "nunique                   654      4           3                 7         2   \n",
      "\n",
      "         RACE  PETS  MULTIVITAMIN  EXTRACTION_ROBOT  \\\n",
      "count    4815  4815          4815              4815   \n",
      "size     4815  4815          4815              4815   \n",
      "nunique     7   525             4                 5   \n",
      "\n",
      "         NONFOODALLERGIES_POISONIVY  ACNE_MEDICATION  PRIMER_PLATE  \\\n",
      "count                          4815             4815          4815   \n",
      "size                           4815             4815          4815   \n",
      "nunique                           3                4            12   \n",
      "\n",
      "         ACNE_MEDICATION_OTC  ANIMAL_PER  MACHINE  TM300_8_TOOL  \\\n",
      "count                   4815        4815     4815          4815   \n",
      "size                    4815        4815     4815          4815   \n",
      "nunique                    4         112        3             7   \n",
      "\n",
      "         MAINFACTOR_OTHER_1  EXPERIMENT_DESIGN_DESCRIPTION  \\\n",
      "count                  4815                           4815   \n",
      "size                   4815                           4815   \n",
      "nunique                 100                              2   \n",
      "\n",
      "         FOODALLERGIES_OTHER  WEIGHT_CHANGE  PRIMER_DATE  PRIMARY_CARB  \\\n",
      "count                   4815           4815         4815          4815   \n",
      "size                    4815           4815         4815          4815   \n",
      "nunique                    3              5            5           782   \n",
      "\n",
      "         DIABETES  MAINFACTOR_OTHER_2  ASTHMA  POOL_FREQUENCY  \\\n",
      "count        4815                4815    4815            4815   \n",
      "size         4815                4815    4815            4815   \n",
      "nunique         5                  65       4               7   \n",
      "\n",
      "         LIBRARY_CONSTRUCTION_PROTOCOL   PKU  DECEASED_PARENT  SEQ_RUN   CAT  \\\n",
      "count                             4815  4815             4815     4815  4815   \n",
      "size                              4815  4815             4815     4815  4815   \n",
      "nunique                              2     4                3        5     4   \n",
      "\n",
      "         LATITUDE  FOODALLERGIES_PEANUTS  CHICKENPOX  SEASONAL_ALLERGIES  \\\n",
      "count        4815                   4815        4815                4815   \n",
      "size         4815                   4815        4815                4815   \n",
      "nunique       430                      3           4                   4   \n",
      "\n",
      "         Description  \n",
      "count           4815  \n",
      "size            4815  \n",
      "nunique           39  \n"
     ]
    }
   ],
   "source": [
    "# Show the all rows of all features. \n",
    "\n",
    "with pd.option_context('display.max_columns', None):\n",
    "    print (df_agg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the columns, there is a large range for the number of unique values 'nunique'."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ask if there are any Nan in the data frame.\n",
    "\n",
    "df.isnull().values.any()\n",
    "\n",
    "# https://datatofish.com/check-nan-pandas-dataframe/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are NaN values in the data frame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ask how many Nan there are in the data frame.\n",
    "\n",
    "df.isnull().sum().sum()\n",
    "\n",
    "# https://datatofish.com/check-nan-pandas-dataframe/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 10 rows that contain NaN values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FOODALLERGIES_OTHER_TEXT</th>\n",
       "      <th>ANTIBIOTIC_CONDITION</th>\n",
       "      <th>STATE</th>\n",
       "      <th>PRIMARY_VEGETABLE</th>\n",
       "      <th>DIET_RESTRICTIONS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>CO</td>\n",
       "      <td>spinach</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>CA</td>\n",
       "      <td>tomatoes (I eat a lot of salsa)</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cats, dogs, certain grasses</td>\n",
       "      <td>no_data</td>\n",
       "      <td>CO</td>\n",
       "      <td>leafy greens</td>\n",
       "      <td>Grain free</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>NY</td>\n",
       "      <td>Potato</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dust</td>\n",
       "      <td>dental cleaning</td>\n",
       "      <td>NY</td>\n",
       "      <td>broccoli</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4810</th>\n",
       "      <td>unknown</td>\n",
       "      <td>UTI</td>\n",
       "      <td>IL</td>\n",
       "      <td>Romaine Lettuce</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4811</th>\n",
       "      <td>unknown</td>\n",
       "      <td>Streptococcus A</td>\n",
       "      <td>unknown</td>\n",
       "      <td>tomatoes, cuncumber</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4812</th>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>WA</td>\n",
       "      <td>potato</td>\n",
       "      <td>Ray Peat diet. no grains, beans, minimal veget...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4813</th>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>IL</td>\n",
       "      <td>Leafy Greens</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4814</th>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>BANKOK</td>\n",
       "      <td>Leaves and fruits</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4815 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         FOODALLERGIES_OTHER_TEXT ANTIBIOTIC_CONDITION    STATE  \\\n",
       "0                         no_data              no_data       CO   \n",
       "1                         no_data              no_data       CA   \n",
       "2     Cats, dogs, certain grasses              no_data       CO   \n",
       "3                         no_data              no_data       NY   \n",
       "4                            dust      dental cleaning       NY   \n",
       "...                           ...                  ...      ...   \n",
       "4810                      unknown                  UTI       IL   \n",
       "4811                      unknown      Streptococcus A  unknown   \n",
       "4812                      no_data              no_data       WA   \n",
       "4813                      no_data              no_data       IL   \n",
       "4814                      no_data              no_data   BANKOK   \n",
       "\n",
       "                    PRIMARY_VEGETABLE  \\\n",
       "0                             spinach   \n",
       "1     tomatoes (I eat a lot of salsa)   \n",
       "2                        leafy greens   \n",
       "3                              Potato   \n",
       "4                            broccoli   \n",
       "...                               ...   \n",
       "4810                  Romaine Lettuce   \n",
       "4811              tomatoes, cuncumber   \n",
       "4812                           potato   \n",
       "4813                     Leafy Greens   \n",
       "4814                Leaves and fruits   \n",
       "\n",
       "                                      DIET_RESTRICTIONS  \n",
       "0                                               no_data  \n",
       "1                                               no_data  \n",
       "2                                            Grain free  \n",
       "3                                               no_data  \n",
       "4                                               no_data  \n",
       "...                                                 ...  \n",
       "4810                                            unknown  \n",
       "4811                                            unknown  \n",
       "4812  Ray Peat diet. no grains, beans, minimal veget...  \n",
       "4813                                            no_data  \n",
       "4814                                            no_data  \n",
       "\n",
       "[4815 rows x 5 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show columns with NaN.\n",
    "\n",
    "df.loc[:, df.isna().any()]\n",
    "\n",
    "# https://stackoverflow.com/questions/36226083/how-to-find-which-columns-contain-any-nan-value-in-pandas-dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above are the 5 columns that contain NaN values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "               #SampleID BarcodeSequence LinkerPrimerSequence  \\\n",
      "190          13018.12575         no_data              no_data   \n",
      "208   000014794.fixed934         no_data              no_data   \n",
      "522   000014791.fixed933         no_data              no_data   \n",
      "1369  000014793.fixed875         no_data              no_data   \n",
      "1980          2578.10762         no_data              no_data   \n",
      "2500         14493.12597         no_data              no_data   \n",
      "3051         14483.12596         no_data              no_data   \n",
      "3306  000014792.fixed931         no_data              no_data   \n",
      "3720         2321.121307         no_data              no_data   \n",
      "3997         3147.113054         no_data              no_data   \n",
      "\n",
      "     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS  \\\n",
      "190                69               no_data                    no_data   \n",
      "208                20               unknown                    unknown   \n",
      "522                20                   yes                    unknown   \n",
      "1369               40               unknown                    unknown   \n",
      "1980          no_data               no_data                    no_data   \n",
      "2500          no_data               no_data                    no_data   \n",
      "3051          no_data               no_data                    no_data   \n",
      "3306               20               unknown                    unknown   \n",
      "3720               49               no_data                    no_data   \n",
      "3997               60               no_data                    no_data   \n",
      "\n",
      "     ASSIGNED_FROM_GEO RUN_DATE                               PET_LOCATIONS  \\\n",
      "190                  y  no_data                           Outdoor, Confined   \n",
      "208                  y  no_data                                      Indoor   \n",
      "522                  y  no_data  Indoor, Outdoor, Outdoor, Outdoor, Outdoor   \n",
      "1369                 y  no_data                                  Housing...   \n",
      "1980                 n  no_data                                      Indoor   \n",
      "2500                 y  no_data                                  Housing...   \n",
      "3051                 y  no_data                                  Housing...   \n",
      "3306                 y  no_data  Indoor, Outdoor, Outdoor, Outdoor, Outdoor   \n",
      "3720                 n  no_data                                     no_data   \n",
      "3997                 n  no_data                                  Housing...   \n",
      "\n",
      "     AGE  ... LIBRARY_CONSTRUCTION_PROTOCOL      PKU DECEASED_PARENT  SEQ_RUN  \\\n",
      "190   14  ...                       no_data       no         no_data  no_data   \n",
      "208   37  ...                       no_data       no         unknown  no_data   \n",
      "522   39  ...                       no_data       no         unknown  no_data   \n",
      "1369  37  ...                       no_data       no         unknown  no_data   \n",
      "1980  45  ...                       no_data       no         no_data  no_data   \n",
      "2500  34  ...                       no_data  no_data         no_data   round9   \n",
      "3051  34  ...                       no_data  no_data         no_data   round9   \n",
      "3306  38  ...                       no_data       no         unknown  no_data   \n",
      "3720  72  ...                       no_data       no         no_data  no_data   \n",
      "3997   1  ...                       no_data       no         no_data  no_data   \n",
      "\n",
      "      CAT LATITUDE FOODALLERGIES_PEANUTS CHICKENPOX SEASONAL_ALLERGIES  \\\n",
      "190    no       42               no_data         no                 no   \n",
      "208   yes     49.4               unknown        yes                 no   \n",
      "522   yes     49.4               unknown        yes                yes   \n",
      "1369  yes     49.4               unknown        yes                 no   \n",
      "1980   no     39.7               no_data        yes                 no   \n",
      "2500   no     51.5               no_data    no_data                yes   \n",
      "3051   no     51.5               no_data    no_data                yes   \n",
      "3306  yes     49.4               unknown        yes                 no   \n",
      "3720   no       41               no_data         no                 no   \n",
      "3997   no     33.8               no_data         no                 no   \n",
      "\n",
      "                               Description  \n",
      "190      American Gut Project Stool sample  \n",
      "208                            Description  \n",
      "522                            Description  \n",
      "1369                           Description  \n",
      "1980     American Gut Project Stool sample  \n",
      "2500     American Gut Project Stool sample  \n",
      "3051  American Gut Project Forehead sample  \n",
      "3306                           Description  \n",
      "3720     American Gut Project Stool sample  \n",
      "3997     American Gut Project Stool sample  \n",
      "\n",
      "[10 rows x 202 columns]\n"
     ]
    }
   ],
   "source": [
    "# Show rows with NaN.\n",
    "\n",
    "is_NaN = df.isnull()\n",
    "row_has_NaN = is_NaN.any(axis=1)\n",
    "rows_with_NaN = df[row_has_NaN]\n",
    "\n",
    "print(rows_with_NaN)\n",
    "\n",
    "# https://www.kite.com/python/answers/how-to-find-rows-with-nan-values-in-a-pandas-dataframe-in-python"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above are the 10 rows that contain NaN values."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First some columns will be deleted. Then the 10 rows with NaN values will be deleted since otherwise, in the inverse order, an error message appears.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the value range of a few specially important columns. Here I am choosing columns relevant for diabetes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['0', '1', '10', '11', '12', '13', '14', '15', '16', '17', '18',\n",
       "       '19', '2', '20', '21', '22', '23', '24', '25', '26', '27', '28',\n",
       "       '29', '3', '30', '31', '32', '33', '34', '35', '36', '37', '38',\n",
       "       '39', '4', '40', '41', '42', '43', '44', '45', '46', '47', '48',\n",
       "       '49', '5', '50', '51', '52', '53', '54', '55', '56', '57', '58',\n",
       "       '59', '6', '60', '61', '62', '63', '64', '65', '66', '67', '68',\n",
       "       '69', '7', '70', '71', '72', '73', '74', '75', '76', '77', '78',\n",
       "       '79', '8', '80', '81', '82', '83', '84', '88', '9', '91', '93',\n",
       "       'no_data', 'unknown'], dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show unique values in a column.\n",
    "# Age range.\n",
    "\n",
    "AGE_unique = np.sort(df.AGE.unique())\n",
    "AGE_unique\n",
    "\n",
    "# https://stackoverflow.com/questions/32072076/find-the-unique-values-in-a-column-and-then-sort-them/32072252"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['I do not have diabetes', 'Type I', 'Type II', 'no_data',\n",
       "       'unknown'], dtype=object)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DIABETES_unique = np.sort(df.DIABETES.unique())\n",
    "DIABETES_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['0', '1', '10', '11', '11.4', '12', '13', '13.5', '14', '15', '16',\n",
       "       '17', '17.3', '17.6', '18', '19', '19.2', '19.9', '2', '2.9', '20',\n",
       "       '20.1', '21', '21.3', '22', '22.2', '23', '24', '25', '25.4', '26',\n",
       "       '26.8', '27', '27.8', '28', '28.2', '28.3', '29', '29.2', '3',\n",
       "       '30', '31', '31.4', '31.5', '31.7', '32', '33', '34', '34.2', '35',\n",
       "       '35.01', '35.6', '36', '36.2', '36.3', '36.6', '36.61', '37', '38',\n",
       "       '38.1', '38.7', '38.81', '39', '39.1', '39.4', '39.6', '39.8', '4',\n",
       "       '4.2', '40', '40.2', '41', '41.6', '41.8', '42', '42.1', '42.2',\n",
       "       '42.3', '42.6', '43', '43.3', '43.4', '43.7', '44', '44.3', '44.5',\n",
       "       '44.7', '45', '45.4', '45.5', '46', '46.4', '46.5', '46.7', '47',\n",
       "       '47.1', '47.2', '47.5', '47.6', '47.7', '48', '48.3', '48.5',\n",
       "       '48.7', '49', '49.6', '49.8', '5', '50', '50.2', '50.8', '51',\n",
       "       '51.7', '52', '53', '53.5', '54', '55', '55.5', '56', '56.1', '57',\n",
       "       '57.7', '58', '59', '59.7', '6', '60', '61', '61.76', '62', '63',\n",
       "       '63.2', '64', '65', '66', '67', '68', '69', '7', '70', '71', '72',\n",
       "       '73', '74', '75', '76', '77', '79', '8', '80', '83', '85', '88',\n",
       "       '9', '90', '95', 'no_data', 'unknown'], dtype=object)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CARBOHYDRATE_PER_unique = np.sort(df.CARBOHYDRATE_PER.unique())\n",
    "CARBOHYDRATE_PER_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['0', '10', '11', '11.7', '12', '13', '13.9', '14', '15', '16',\n",
       "       '17', '17.7', '18', '19', '19.01', '2', '20', '21', '22', '22.1',\n",
       "       '23', '24', '25', '25.3', '25.5', '26', '26.8', '26.9', '27', '28',\n",
       "       '28.3', '28.6', '29', '29.1', '29.4', '30', '30.5', '30.7', '30.9',\n",
       "       '31', '31.2', '31.4', '31.6', '31.9', '32', '32.4', '32.7', '32.8',\n",
       "       '33', '33.7', '33.8', '33.9', '34', '34.2', '34.5', '34.6', '35',\n",
       "       '35.1', '35.4', '35.6', '36', '36.1', '36.2', '36.3', '36.5',\n",
       "       '36.6', '37', '37.4', '37.5', '37.9', '38', '38.1', '38.3', '38.5',\n",
       "       '38.8', '39', '39.1', '39.3', '39.5', '39.9', '4', '40', '40.3',\n",
       "       '40.5', '40.94', '41', '41.2', '41.3', '41.4', '41.7', '42',\n",
       "       '42.4', '42.6', '43', '43.2', '43.3', '44', '44.1', '44.8', '44.9',\n",
       "       '45', '45.12', '45.3', '46', '46.58', '46.7', '47', '47.1', '47.2',\n",
       "       '47.9', '48', '48.7', '49', '5', '50', '51', '51.3', '52', '53',\n",
       "       '53.5', '54', '54.7', '55', '55.3', '55.8', '56', '56.1', '57',\n",
       "       '58', '58.9', '59', '6', '60', '61', '62', '62.2', '63', '63.3',\n",
       "       '64', '65', '65.8', '66', '67', '68', '69', '7', '70', '70.6',\n",
       "       '71', '72', '73', '74', '75', '75.3', '76', '78', '79', '8', '80',\n",
       "       '81', '85', '9', 'no_data', 'unknown'], dtype=object)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FAT_PER_unique = np.sort(df.FAT_PER.unique())\n",
    "FAT_PER_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  res_values = method(rvalues)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>NONFOODALLERGIES_DRUG</th>\n",
       "      <th>NONFOODALLERGIES_BEESTINGS</th>\n",
       "      <th>ASSIGNED_FROM_GEO</th>\n",
       "      <th>RUN_DATE</th>\n",
       "      <th>PET_LOCATIONS</th>\n",
       "      <th>AGE</th>\n",
       "      <th>...</th>\n",
       "      <th>LIBRARY_CONSTRUCTION_PROTOCOL</th>\n",
       "      <th>PKU</th>\n",
       "      <th>DECEASED_PARENT</th>\n",
       "      <th>SEQ_RUN</th>\n",
       "      <th>CAT</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>FOODALLERGIES_PEANUTS</th>\n",
       "      <th>CHICKENPOX</th>\n",
       "      <th>SEASONAL_ALLERGIES</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11224.12109</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>55</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>51</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>39.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5034.107635</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>56</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>32.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11351.12107</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>16</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>56</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>40</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2013.113155</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>40.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4731.107641</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>61</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>72</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>43.1</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4810</th>\n",
       "      <td>000013296.fixed213</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>50</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>71</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>40</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4811</th>\n",
       "      <td>000007822.fixed1101</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>17</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>50.8</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>Description</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4812</th>\n",
       "      <td>9074.121304</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>34</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>29</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>47.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4813</th>\n",
       "      <td>2418.128313</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>30</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>46</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Knight_American_Gut_Freeze_Thaw</td>\n",
       "      <td>yes</td>\n",
       "      <td>42</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4814</th>\n",
       "      <td>7179.126173</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>60</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>13.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4815 rows × 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                #SampleID BarcodeSequence LinkerPrimerSequence  \\\n",
       "0             11224.12109         no_data              no_data   \n",
       "1             5034.107635         no_data              no_data   \n",
       "2             11351.12107         no_data              no_data   \n",
       "3             2013.113155         no_data              no_data   \n",
       "4             4731.107641         no_data              no_data   \n",
       "...                   ...             ...                  ...   \n",
       "4810   000013296.fixed213         no_data              no_data   \n",
       "4811  000007822.fixed1101         no_data              no_data   \n",
       "4812          9074.121304         no_data              no_data   \n",
       "4813          2418.128313         no_data              no_data   \n",
       "4814          7179.126173         no_data              no_data   \n",
       "\n",
       "     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS  \\\n",
       "0                  55               no_data                    no_data   \n",
       "1                  56               no_data                    no_data   \n",
       "2                  16               no_data                    no_data   \n",
       "3             no_data               no_data                    no_data   \n",
       "4                  61                   yes                    no_data   \n",
       "...               ...                   ...                        ...   \n",
       "4810               50               unknown                    unknown   \n",
       "4811               17               unknown                    unknown   \n",
       "4812               34               no_data                    no_data   \n",
       "4813               30               no_data                    no_data   \n",
       "4814               60               no_data                    no_data   \n",
       "\n",
       "     ASSIGNED_FROM_GEO RUN_DATE PET_LOCATIONS AGE  ...  \\\n",
       "0                    n  no_data    Housing...  51  ...   \n",
       "1                    n  no_data       no_data  35  ...   \n",
       "2                    n  no_data    Housing...  56  ...   \n",
       "3                    n  no_data    Housing...  37  ...   \n",
       "4                    n  no_data       no_data  72  ...   \n",
       "...                ...      ...           ...  ..  ...   \n",
       "4810                 y  no_data    Housing...  71  ...   \n",
       "4811                 y  no_data    Housing...   9  ...   \n",
       "4812                 n  no_data    Housing...  29  ...   \n",
       "4813                 y  no_data       no_data  46  ...   \n",
       "4814                 y  no_data    Housing...   1  ...   \n",
       "\n",
       "     LIBRARY_CONSTRUCTION_PROTOCOL      PKU DECEASED_PARENT  \\\n",
       "0                          no_data       no         no_data   \n",
       "1                          no_data       no         no_data   \n",
       "2                          no_data       no         no_data   \n",
       "3                          no_data       no         no_data   \n",
       "4                          no_data       no         no_data   \n",
       "...                            ...      ...             ...   \n",
       "4810                       no_data       no         unknown   \n",
       "4811                       no_data  unknown         unknown   \n",
       "4812                       no_data       no         no_data   \n",
       "4813                       no_data       no         no_data   \n",
       "4814                       no_data       no         no_data   \n",
       "\n",
       "                              SEQ_RUN  CAT LATITUDE FOODALLERGIES_PEANUTS  \\\n",
       "0                             no_data  yes     39.7               no_data   \n",
       "1                             no_data  yes     32.7               no_data   \n",
       "2                             no_data   no       40               no_data   \n",
       "3                             no_data   no     40.7               no_data   \n",
       "4                             no_data   no     43.1               no_data   \n",
       "...                               ...  ...      ...                   ...   \n",
       "4810                          no_data  yes       40               unknown   \n",
       "4811                          no_data   no     50.8               unknown   \n",
       "4812                          no_data  yes     47.7               no_data   \n",
       "4813  Knight_American_Gut_Freeze_Thaw  yes       42               no_data   \n",
       "4814                          no_data   no     13.7               no_data   \n",
       "\n",
       "     CHICKENPOX SEASONAL_ALLERGIES                        Description  \n",
       "0           yes                 no  American Gut Project Stool sample  \n",
       "1           yes                 no  American Gut Project Stool sample  \n",
       "2           yes                 no  American Gut Project Stool sample  \n",
       "3           yes                 no  American Gut Project Stool sample  \n",
       "4           yes                yes  American Gut Project Stool sample  \n",
       "...         ...                ...                                ...  \n",
       "4810        yes                 no                               None  \n",
       "4811         no                 no                        Description  \n",
       "4812        yes                 no  American Gut Project Stool sample  \n",
       "4813        yes                 no  American Gut Project Stool sample  \n",
       "4814        yes                yes  American Gut Project Stool sample  \n",
       "\n",
       "[4815 rows x 202 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look for rows, that contain cells labelled as 'no_data'.\n",
    "\n",
    "df[(df == 'no_data').any(axis=1)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  res_values = method(rvalues)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(4815, 202)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df == 'no_data').any(axis=1)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>NONFOODALLERGIES_DRUG</th>\n",
       "      <th>NONFOODALLERGIES_BEESTINGS</th>\n",
       "      <th>ASSIGNED_FROM_GEO</th>\n",
       "      <th>RUN_DATE</th>\n",
       "      <th>PET_LOCATIONS</th>\n",
       "      <th>AGE</th>\n",
       "      <th>...</th>\n",
       "      <th>LIBRARY_CONSTRUCTION_PROTOCOL</th>\n",
       "      <th>PKU</th>\n",
       "      <th>DECEASED_PARENT</th>\n",
       "      <th>SEQ_RUN</th>\n",
       "      <th>CAT</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>FOODALLERGIES_PEANUTS</th>\n",
       "      <th>CHICKENPOX</th>\n",
       "      <th>SEASONAL_ALLERGIES</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [#SampleID, BarcodeSequence, LinkerPrimerSequence, CARBOHYDRATE_PER, NONFOODALLERGIES_DRUG, NONFOODALLERGIES_BEESTINGS, ASSIGNED_FROM_GEO, RUN_DATE, PET_LOCATIONS, AGE, TOT_MASS, GENERAL_MEDS, SAMPLE_ID, ORIG_SAMPLE_NAME, QUINOLINE, BODY_SITE, MIGRAINE_FACTOR_2, MIGRAINE_FACTOR_3, TONSILS_REMOVED, MIGRAINE_FACTOR_1, ELEVATION, FLOSSING_FREQUENCY, PROTEIN_PER, CSECTION, PCR_PRIMERS, COUNTRY_OF_BIRTH, MACRONUTRIENT_PCT_TOTAL, COSMETICS_FREQUENCY, MASTERMIX_LOT, HOST_SUBJECT_ID, ENA-CHECKLIST, GLUTEN, SAMPLE_NAME, TM1000_8_TOOL, LONGITUDE, DOG, MIGRAINE_AGGRAVATION, ORIG_NAME, FRAT, DIABETES_MEDICATION, SAMPLE_PLATE, PENICILLIN, FOODALLERGIES_OTHER_TEXT, KEY_SEQ, BODY_PRODUCT, REGION, SAMPLE_TYPE, MIGRAINE_FREQUENCY, MIGRAINE_RELATIVES, EXPERIMENT_TITLE, PLANT_PER, EXTRACTIONKIT_LOT, SAMPLE, MIGRAINE_PHOTOPHOBIA, SITE_SAMPLED, DRINKING_WATER_SOURCE, CENTER_NAME, COUNTRY, ANTIBIOTIC_SELECT, MIGRAINE_NAUSEA, COMMON_NAME, FOODALLERGIES_TREENUTS, MIGRAINE_AURA, SEQUENCING_METH, FLU_VACCINE_DATE, PREGNANT, MIGRAINE, BMI, ENV_BIOME, WELL_ID, TEETHBRUSHING_FREQUENCY, MIGRAINE_MEDICATIONS, CONTRACEPTIVE, SAMPLE_CENTER, PHYSICAL_LOCATION, LIVINGWITH, LINKER, RELATIONS, DIABETES_MEDICATIONS, SUPPLEMENTS, AGE_UNIT, TANNING_SPRAYS, EXERCISE_FREQUENCY, HOST_COMMON_NAME, SULFA_DRUG, NONFOODALLERGIES_NO, ENV_MATTER, RACE_OTHER, SKIN_CONDITION, ALTITUDE, NONFOODALLERGIES_DANDER, CONDITIONS_MEDICATION, LAST_TRAVEL, TARGET_SUBFRAGMENT, TITLE, HEIGHT_IN, ANTIBIOTIC_CONDITION, PET_CONTACT, MIGRAINE_PAIN, ROOMMATES, ...]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 202 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look for rows, that contain cells labelled as 'No_data'.\n",
    "\n",
    "df[(df == 'No_data').any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  res_values = method(rvalues)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>NONFOODALLERGIES_DRUG</th>\n",
       "      <th>NONFOODALLERGIES_BEESTINGS</th>\n",
       "      <th>ASSIGNED_FROM_GEO</th>\n",
       "      <th>RUN_DATE</th>\n",
       "      <th>PET_LOCATIONS</th>\n",
       "      <th>AGE</th>\n",
       "      <th>...</th>\n",
       "      <th>LIBRARY_CONSTRUCTION_PROTOCOL</th>\n",
       "      <th>PKU</th>\n",
       "      <th>DECEASED_PARENT</th>\n",
       "      <th>SEQ_RUN</th>\n",
       "      <th>CAT</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>FOODALLERGIES_PEANUTS</th>\n",
       "      <th>CHICKENPOX</th>\n",
       "      <th>SEASONAL_ALLERGIES</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5034.107635</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>56</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>32.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4731.107641</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>61</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>72</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>43.1</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>000015569.fixed647</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>37</td>\n",
       "      <td>yes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Outdoor</td>\n",
       "      <td>65</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>44.7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Description</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>000011031.fixed716</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>50</td>\n",
       "      <td>yes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>54</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>unknown</td>\n",
       "      <td>37.8</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>Description</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>000016722.fixed480</td>\n",
       "      <td>TTCCTGTTAACC</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>52</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>8/21/14</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>32</td>\n",
       "      <td>...</td>\n",
       "      <td>This analysis was done as in Caporaso et al 20...</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>40</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4805</th>\n",
       "      <td>000016360.fixed111</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>19</td>\n",
       "      <td>yes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Indoor</td>\n",
       "      <td>40</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>34.1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4807</th>\n",
       "      <td>3093.107579</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>40.3</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4809</th>\n",
       "      <td>000015346.fixed136</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>65</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Outdoor</td>\n",
       "      <td>28</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>39.3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4810</th>\n",
       "      <td>000013296.fixed213</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>50</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>71</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>40</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4811</th>\n",
       "      <td>000007822.fixed1101</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>17</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>50.8</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>Description</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1557 rows × 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                #SampleID BarcodeSequence LinkerPrimerSequence  \\\n",
       "1             5034.107635         no_data              no_data   \n",
       "4             4731.107641         no_data              no_data   \n",
       "7      000015569.fixed647         no_data              no_data   \n",
       "9      000011031.fixed716         no_data              no_data   \n",
       "10     000016722.fixed480    TTCCTGTTAACC  GTGCCAGCMGCCGCGGTAA   \n",
       "...                   ...             ...                  ...   \n",
       "4805   000016360.fixed111         no_data              no_data   \n",
       "4807          3093.107579         no_data              no_data   \n",
       "4809   000015346.fixed136         no_data              no_data   \n",
       "4810   000013296.fixed213         no_data              no_data   \n",
       "4811  000007822.fixed1101         no_data              no_data   \n",
       "\n",
       "     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS  \\\n",
       "1                  56               no_data                    no_data   \n",
       "4                  61                   yes                    no_data   \n",
       "7                  37                   yes                    unknown   \n",
       "9                  50                   yes                    unknown   \n",
       "10                 52               unknown                    unknown   \n",
       "...               ...                   ...                        ...   \n",
       "4805               19                   yes                    unknown   \n",
       "4807          no_data               no_data                    no_data   \n",
       "4809               65               unknown                    unknown   \n",
       "4810               50               unknown                    unknown   \n",
       "4811               17               unknown                    unknown   \n",
       "\n",
       "     ASSIGNED_FROM_GEO RUN_DATE PET_LOCATIONS AGE  ...  \\\n",
       "1                    n  no_data       no_data  35  ...   \n",
       "4                    n  no_data       no_data  72  ...   \n",
       "7                    y  no_data       Outdoor  65  ...   \n",
       "9                    y  no_data    Housing...  54  ...   \n",
       "10                   y  8/21/14    Housing...  32  ...   \n",
       "...                ...      ...           ...  ..  ...   \n",
       "4805                 y  no_data        Indoor  40  ...   \n",
       "4807                 n  no_data       no_data   2  ...   \n",
       "4809                 y  no_data       Outdoor  28  ...   \n",
       "4810                 y  no_data    Housing...  71  ...   \n",
       "4811                 y  no_data    Housing...   9  ...   \n",
       "\n",
       "                          LIBRARY_CONSTRUCTION_PROTOCOL      PKU  \\\n",
       "1                                               no_data       no   \n",
       "4                                               no_data       no   \n",
       "7                                               no_data       no   \n",
       "9                                               no_data       no   \n",
       "10    This analysis was done as in Caporaso et al 20...       no   \n",
       "...                                                 ...      ...   \n",
       "4805                                            no_data       no   \n",
       "4807                                            no_data       no   \n",
       "4809                                            no_data       no   \n",
       "4810                                            no_data       no   \n",
       "4811                                            no_data  unknown   \n",
       "\n",
       "     DECEASED_PARENT  SEQ_RUN      CAT LATITUDE FOODALLERGIES_PEANUTS  \\\n",
       "1            no_data  no_data      yes     32.7               no_data   \n",
       "4            no_data  no_data       no     43.1               no_data   \n",
       "7            unknown  no_data      yes     44.7               unknown   \n",
       "9            unknown  no_data  unknown     37.8               unknown   \n",
       "10           unknown  no_data      yes       40               unknown   \n",
       "...              ...      ...      ...      ...                   ...   \n",
       "4805         unknown  no_data      yes     34.1               unknown   \n",
       "4807         no_data  no_data      yes     40.3               no_data   \n",
       "4809         unknown  no_data       no     39.3               unknown   \n",
       "4810         unknown  no_data      yes       40               unknown   \n",
       "4811         unknown  no_data       no     50.8               unknown   \n",
       "\n",
       "     CHICKENPOX SEASONAL_ALLERGIES                        Description  \n",
       "1           yes                 no  American Gut Project Stool sample  \n",
       "4           yes                yes  American Gut Project Stool sample  \n",
       "7           yes                yes                        Description  \n",
       "9           yes                 no                        Description  \n",
       "10          yes                 no  American Gut Project Stool sample  \n",
       "...         ...                ...                                ...  \n",
       "4805        yes                yes                               None  \n",
       "4807         no                 no  American Gut Project Stool sample  \n",
       "4809        yes                yes                               None  \n",
       "4810        yes                 no                               None  \n",
       "4811         no                 no                        Description  \n",
       "\n",
       "[1557 rows x 202 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look for rows, that contain cells labelled as 'unknown'.\n",
    "\n",
    "df[(df == 'unknown').any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  res_values = method(rvalues)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(1557, 202)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df == 'unknown').any(axis=1)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>NONFOODALLERGIES_DRUG</th>\n",
       "      <th>NONFOODALLERGIES_BEESTINGS</th>\n",
       "      <th>ASSIGNED_FROM_GEO</th>\n",
       "      <th>RUN_DATE</th>\n",
       "      <th>PET_LOCATIONS</th>\n",
       "      <th>AGE</th>\n",
       "      <th>...</th>\n",
       "      <th>LIBRARY_CONSTRUCTION_PROTOCOL</th>\n",
       "      <th>PKU</th>\n",
       "      <th>DECEASED_PARENT</th>\n",
       "      <th>SEQ_RUN</th>\n",
       "      <th>CAT</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>FOODALLERGIES_PEANUTS</th>\n",
       "      <th>CHICKENPOX</th>\n",
       "      <th>SEASONAL_ALLERGIES</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2724</th>\n",
       "      <td>000012073.fixed49</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>60</td>\n",
       "      <td>yes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>37.7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              #SampleID BarcodeSequence LinkerPrimerSequence CARBOHYDRATE_PER  \\\n",
       "2724  000012073.fixed49         no_data              no_data               60   \n",
       "\n",
       "     NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS ASSIGNED_FROM_GEO  \\\n",
       "2724                   yes                    unknown                 y   \n",
       "\n",
       "     RUN_DATE PET_LOCATIONS AGE  ... LIBRARY_CONSTRUCTION_PROTOCOL PKU  \\\n",
       "2724  no_data    Housing...  35  ...                       no_data  no   \n",
       "\n",
       "     DECEASED_PARENT  SEQ_RUN CAT LATITUDE FOODALLERGIES_PEANUTS CHICKENPOX  \\\n",
       "2724         unknown  no_data  no     37.7               unknown        yes   \n",
       "\n",
       "     SEASONAL_ALLERGIES Description  \n",
       "2724                 no        None  \n",
       "\n",
       "[1 rows x 202 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look for rows, that contain cells labelled as 'Unknown'.\n",
    "\n",
    "df[(df == 'Unknown').any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  res_values = method(rvalues)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>NONFOODALLERGIES_DRUG</th>\n",
       "      <th>NONFOODALLERGIES_BEESTINGS</th>\n",
       "      <th>ASSIGNED_FROM_GEO</th>\n",
       "      <th>RUN_DATE</th>\n",
       "      <th>PET_LOCATIONS</th>\n",
       "      <th>AGE</th>\n",
       "      <th>...</th>\n",
       "      <th>LIBRARY_CONSTRUCTION_PROTOCOL</th>\n",
       "      <th>PKU</th>\n",
       "      <th>DECEASED_PARENT</th>\n",
       "      <th>SEQ_RUN</th>\n",
       "      <th>CAT</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>FOODALLERGIES_PEANUTS</th>\n",
       "      <th>CHICKENPOX</th>\n",
       "      <th>SEASONAL_ALLERGIES</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>000016722.fixed480</td>\n",
       "      <td>TTCCTGTTAACC</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>52</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>8/21/14</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>32</td>\n",
       "      <td>...</td>\n",
       "      <td>This analysis was done as in Caporaso et al 20...</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>40</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>000016449.fixed444</td>\n",
       "      <td>TGACAACCGAAT</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>20</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>9/18/14</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>49</td>\n",
       "      <td>...</td>\n",
       "      <td>This analysis was done as in Caporaso et al 20...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>64.8</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>000018297.fixed577</td>\n",
       "      <td>GTAACCACCACC</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>20</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>9/18/14</td>\n",
       "      <td>Indoor</td>\n",
       "      <td>60</td>\n",
       "      <td>...</td>\n",
       "      <td>This analysis was done as in Caporaso et al 20...</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>unknown</td>\n",
       "      <td>41.7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>000016738.fixed493</td>\n",
       "      <td>CTGTAAAGGTTG</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>8/21/14</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>This analysis was done as in Caporaso et al 20...</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>40</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>000013636.fixed67</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>51</td>\n",
       "      <td>yes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>55</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>39</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4800</th>\n",
       "      <td>000016290.fixed430</td>\n",
       "      <td>GTACTGAAGATC</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>30</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>8/21/14</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>This analysis was done as in Caporaso et al 20...</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>52.4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4802</th>\n",
       "      <td>BLANK.2.10A.fixed595</td>\n",
       "      <td>AAGACGTAGCGG</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>8/21/14</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>This analysis was done as in Caporaso et al 20...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>unknown</td>\n",
       "      <td>40</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4805</th>\n",
       "      <td>000016360.fixed111</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>19</td>\n",
       "      <td>yes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Indoor</td>\n",
       "      <td>40</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>34.1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4809</th>\n",
       "      <td>000015346.fixed136</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>65</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Outdoor</td>\n",
       "      <td>28</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>39.3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4810</th>\n",
       "      <td>000013296.fixed213</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>50</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>71</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>40</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>623 rows × 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 #SampleID BarcodeSequence LinkerPrimerSequence  \\\n",
       "10      000016722.fixed480    TTCCTGTTAACC  GTGCCAGCMGCCGCGGTAA   \n",
       "12      000016449.fixed444    TGACAACCGAAT  GTGCCAGCMGCCGCGGTAA   \n",
       "17      000018297.fixed577    GTAACCACCACC  GTGCCAGCMGCCGCGGTAA   \n",
       "44      000016738.fixed493    CTGTAAAGGTTG  GTGCCAGCMGCCGCGGTAA   \n",
       "57       000013636.fixed67         no_data              no_data   \n",
       "...                    ...             ...                  ...   \n",
       "4800    000016290.fixed430    GTACTGAAGATC  GTGCCAGCMGCCGCGGTAA   \n",
       "4802  BLANK.2.10A.fixed595    AAGACGTAGCGG  GTGCCAGCMGCCGCGGTAA   \n",
       "4805    000016360.fixed111         no_data              no_data   \n",
       "4809    000015346.fixed136         no_data              no_data   \n",
       "4810    000013296.fixed213         no_data              no_data   \n",
       "\n",
       "     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS  \\\n",
       "10                 52               unknown                    unknown   \n",
       "12                 20               unknown                    unknown   \n",
       "17                 20               unknown                    unknown   \n",
       "44            unknown               unknown                    unknown   \n",
       "57                 51                   yes                    unknown   \n",
       "...               ...                   ...                        ...   \n",
       "4800               30               unknown                    unknown   \n",
       "4802          unknown               unknown                    unknown   \n",
       "4805               19                   yes                    unknown   \n",
       "4809               65               unknown                    unknown   \n",
       "4810               50               unknown                    unknown   \n",
       "\n",
       "     ASSIGNED_FROM_GEO RUN_DATE PET_LOCATIONS      AGE  ...  \\\n",
       "10                   y  8/21/14    Housing...       32  ...   \n",
       "12                   y  9/18/14    Housing...       49  ...   \n",
       "17                   y  9/18/14        Indoor       60  ...   \n",
       "44                   y  8/21/14    Housing...       35  ...   \n",
       "57                   y  no_data    Housing...       55  ...   \n",
       "...                ...      ...           ...      ...  ...   \n",
       "4800                 y  8/21/14    Housing...        1  ...   \n",
       "4802                 y  8/21/14       unknown  unknown  ...   \n",
       "4805                 y  no_data        Indoor       40  ...   \n",
       "4809                 y  no_data       Outdoor       28  ...   \n",
       "4810                 y  no_data    Housing...       71  ...   \n",
       "\n",
       "                          LIBRARY_CONSTRUCTION_PROTOCOL      PKU  \\\n",
       "10    This analysis was done as in Caporaso et al 20...       no   \n",
       "12    This analysis was done as in Caporaso et al 20...  unknown   \n",
       "17    This analysis was done as in Caporaso et al 20...       no   \n",
       "44    This analysis was done as in Caporaso et al 20...       no   \n",
       "57                                              no_data       no   \n",
       "...                                                 ...      ...   \n",
       "4800  This analysis was done as in Caporaso et al 20...       no   \n",
       "4802  This analysis was done as in Caporaso et al 20...  unknown   \n",
       "4805                                            no_data       no   \n",
       "4809                                            no_data       no   \n",
       "4810                                            no_data       no   \n",
       "\n",
       "     DECEASED_PARENT  SEQ_RUN      CAT LATITUDE FOODALLERGIES_PEANUTS  \\\n",
       "10           unknown  no_data      yes       40               unknown   \n",
       "12           unknown  no_data      yes     64.8               unknown   \n",
       "17           unknown  no_data  unknown     41.7               unknown   \n",
       "44           unknown  no_data       no       40               unknown   \n",
       "57           unknown  no_data       no       39               unknown   \n",
       "...              ...      ...      ...      ...                   ...   \n",
       "4800         unknown  no_data       no     52.4               unknown   \n",
       "4802         unknown  no_data  unknown       40               unknown   \n",
       "4805         unknown  no_data      yes     34.1               unknown   \n",
       "4809         unknown  no_data       no     39.3               unknown   \n",
       "4810         unknown  no_data      yes       40               unknown   \n",
       "\n",
       "     CHICKENPOX SEASONAL_ALLERGIES                        Description  \n",
       "10          yes                 no  American Gut Project Stool sample  \n",
       "12      unknown            unknown  American Gut Project Stool sample  \n",
       "17          yes                 no  American Gut Project Stool sample  \n",
       "44           no                 no  American Gut Project Stool sample  \n",
       "57          yes                 no                               None  \n",
       "...         ...                ...                                ...  \n",
       "4800        yes                yes  American Gut Project Stool sample  \n",
       "4802    unknown            unknown  American Gut Project Stool sample  \n",
       "4805        yes                yes                               None  \n",
       "4809        yes                yes                               None  \n",
       "4810        yes                 no                               None  \n",
       "\n",
       "[623 rows x 202 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look for rows, that contain cells labelled as 'None'.\n",
    "\n",
    "df[(df == 'None').any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  res_values = method(rvalues)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(623, 202)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df == 'None').any(axis=1)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>NONFOODALLERGIES_DRUG</th>\n",
       "      <th>NONFOODALLERGIES_BEESTINGS</th>\n",
       "      <th>ASSIGNED_FROM_GEO</th>\n",
       "      <th>RUN_DATE</th>\n",
       "      <th>PET_LOCATIONS</th>\n",
       "      <th>AGE</th>\n",
       "      <th>...</th>\n",
       "      <th>LIBRARY_CONSTRUCTION_PROTOCOL</th>\n",
       "      <th>PKU</th>\n",
       "      <th>DECEASED_PARENT</th>\n",
       "      <th>SEQ_RUN</th>\n",
       "      <th>CAT</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>FOODALLERGIES_PEANUTS</th>\n",
       "      <th>CHICKENPOX</th>\n",
       "      <th>SEASONAL_ALLERGIES</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>000016470.fixed656</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>66</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>34.1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>Description</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>11335.12105</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>63</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>71</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>42.4</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764</th>\n",
       "      <td>000016471.fixed968</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>66</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>34.1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>Description</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788</th>\n",
       "      <td>7106.107567</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>40</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>51</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>38</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Forehead sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985</th>\n",
       "      <td>1763.107581</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>5</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>39.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1114</th>\n",
       "      <td>7102.107567</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>40</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>51</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>38</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Mouth sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1275</th>\n",
       "      <td>3812.113184</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>47</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>39.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1364</th>\n",
       "      <td>000016334.fixed235</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>53</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>48</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>38.9</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>11190.12107</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>53</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>62</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>45.5</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1687</th>\n",
       "      <td>6668.113033</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>10</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>69</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>30</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2045</th>\n",
       "      <td>6673.105393</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>77</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>64</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>38.5</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2646</th>\n",
       "      <td>2665.113158</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>30</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>70</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>35.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3502</th>\n",
       "      <td>7103.10757</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>40</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>51</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>38</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3515</th>\n",
       "      <td>5016.107642</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>48</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Indoor</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>33.8</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3664</th>\n",
       "      <td>7105.107565</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>40</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>51</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>38</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3725</th>\n",
       "      <td>3891.113168</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>20</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Indoor, Outdoor, Outdoor</td>\n",
       "      <td>66</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>33.2</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4084</th>\n",
       "      <td>14126.12619</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>42</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Indoor</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>36</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4266</th>\n",
       "      <td>7104.107568</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>40</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>51</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>38</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4580</th>\n",
       "      <td>13079.12571</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>y</td>\n",
       "      <td>no_data</td>\n",
       "      <td>Housing...</td>\n",
       "      <td>47</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>36.9</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4592</th>\n",
       "      <td>6805.113051</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>35</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>n</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>62</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>39.1</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>American Gut Project Stool sample</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               #SampleID BarcodeSequence LinkerPrimerSequence  \\\n",
       "209   000016470.fixed656         no_data              no_data   \n",
       "224          11335.12105         no_data              no_data   \n",
       "764   000016471.fixed968         no_data              no_data   \n",
       "788          7106.107567         no_data              no_data   \n",
       "985          1763.107581         no_data              no_data   \n",
       "1114         7102.107567         no_data              no_data   \n",
       "1275         3812.113184         no_data              no_data   \n",
       "1364  000016334.fixed235         no_data              no_data   \n",
       "1526         11190.12107         no_data              no_data   \n",
       "1687         6668.113033         no_data              no_data   \n",
       "2045         6673.105393         no_data              no_data   \n",
       "2646         2665.113158         no_data              no_data   \n",
       "3502          7103.10757         no_data              no_data   \n",
       "3515         5016.107642         no_data              no_data   \n",
       "3664         7105.107565         no_data              no_data   \n",
       "3725         3891.113168         no_data              no_data   \n",
       "4084         14126.12619         no_data              no_data   \n",
       "4266         7104.107568         no_data              no_data   \n",
       "4580         13079.12571         no_data              no_data   \n",
       "4592         6805.113051         no_data              no_data   \n",
       "\n",
       "     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS  \\\n",
       "209                66               unknown                    unknown   \n",
       "224                63               no_data                    no_data   \n",
       "764                66               unknown                    unknown   \n",
       "788                40               no_data                    no_data   \n",
       "985                 5               no_data                    no_data   \n",
       "1114               40               no_data                    no_data   \n",
       "1275               47               no_data                    no_data   \n",
       "1364               53               unknown                    unknown   \n",
       "1526               53               no_data                    no_data   \n",
       "1687               10               no_data                    no_data   \n",
       "2045               77                   yes                    no_data   \n",
       "2646               30               no_data                    no_data   \n",
       "3502               40               no_data                    no_data   \n",
       "3515               48               no_data                    no_data   \n",
       "3664               40               no_data                    no_data   \n",
       "3725               20               no_data                        yes   \n",
       "4084               42               no_data                    no_data   \n",
       "4266               40               no_data                    no_data   \n",
       "4580          no_data               no_data                    no_data   \n",
       "4592               35               no_data                    no_data   \n",
       "\n",
       "     ASSIGNED_FROM_GEO RUN_DATE             PET_LOCATIONS AGE  ...  \\\n",
       "209                  y  no_data                Housing...  21  ...   \n",
       "224                  n  no_data                Housing...  71  ...   \n",
       "764                  y  no_data                Housing...  21  ...   \n",
       "788                  n  no_data                   no_data  51  ...   \n",
       "985                  n  no_data                   no_data  35  ...   \n",
       "1114                 n  no_data                   no_data  51  ...   \n",
       "1275                 n  no_data                   no_data   2  ...   \n",
       "1364                 y  no_data                Housing...  48  ...   \n",
       "1526                 n  no_data                Housing...  62  ...   \n",
       "1687                 n  no_data                   no_data  69  ...   \n",
       "2045                 n  no_data                   no_data  64  ...   \n",
       "2646                 n  no_data                   no_data  70  ...   \n",
       "3502                 n  no_data                   no_data  51  ...   \n",
       "3515                 n  no_data                    Indoor   4  ...   \n",
       "3664                 n  no_data                   no_data  51  ...   \n",
       "3725                 n  no_data  Indoor, Outdoor, Outdoor  66  ...   \n",
       "4084                 y  no_data                    Indoor  37  ...   \n",
       "4266                 n  no_data                   no_data  51  ...   \n",
       "4580                 y  no_data                Housing...  47  ...   \n",
       "4592                 n  no_data                   no_data  62  ...   \n",
       "\n",
       "     LIBRARY_CONSTRUCTION_PROTOCOL      PKU DECEASED_PARENT  SEQ_RUN      CAT  \\\n",
       "209                        no_data       no         unknown  no_data       no   \n",
       "224                        no_data  no_data         no_data  no_data       no   \n",
       "764                        no_data       no         unknown  no_data       no   \n",
       "788                        no_data       no         no_data  no_data      yes   \n",
       "985                        no_data       no         no_data  no_data       no   \n",
       "1114                       no_data       no         no_data  no_data      yes   \n",
       "1275                       no_data       no         no_data  no_data       no   \n",
       "1364                       no_data       no         unknown  no_data       no   \n",
       "1526                       no_data       no         no_data  no_data       no   \n",
       "1687                       no_data       no         no_data  no_data      yes   \n",
       "2045                       no_data       no         no_data  no_data       no   \n",
       "2646                       no_data       no         no_data  no_data      yes   \n",
       "3502                       no_data       no         no_data  no_data      yes   \n",
       "3515                       no_data       no         no_data  no_data      yes   \n",
       "3664                       no_data       no         no_data  no_data      yes   \n",
       "3725                       no_data       no         no_data  no_data  no_data   \n",
       "4084                       no_data       no         no_data  no_data      yes   \n",
       "4266                       no_data       no         no_data  no_data      yes   \n",
       "4580                       no_data       no         no_data  no_data       no   \n",
       "4592                       no_data       no         no_data  no_data       no   \n",
       "\n",
       "     LATITUDE FOODALLERGIES_PEANUTS CHICKENPOX SEASONAL_ALLERGIES  \\\n",
       "209      34.1               unknown         no                 no   \n",
       "224      42.4               no_data        yes            no_data   \n",
       "764      34.1               unknown         no                 no   \n",
       "788        38               no_data        yes                 no   \n",
       "985      39.7               no_data        yes                yes   \n",
       "1114       38               no_data        yes                 no   \n",
       "1275     39.7               no_data         no                 no   \n",
       "1364     38.9               unknown        yes                 no   \n",
       "1526     45.5               no_data        yes                yes   \n",
       "1687       30               no_data        yes                yes   \n",
       "2045     38.5               no_data        yes            no_data   \n",
       "2646     35.7               no_data        yes                yes   \n",
       "3502       38               no_data        yes                 no   \n",
       "3515     33.8               no_data         no                 no   \n",
       "3664       38               no_data        yes                 no   \n",
       "3725     33.2               no_data        yes                 no   \n",
       "4084       36               no_data        yes                 no   \n",
       "4266       38               no_data        yes                 no   \n",
       "4580     36.9               no_data         no                 no   \n",
       "4592     39.1               no_data        yes                yes   \n",
       "\n",
       "                               Description  \n",
       "209                            Description  \n",
       "224      American Gut Project Stool sample  \n",
       "764                            Description  \n",
       "788   American Gut Project Forehead sample  \n",
       "985      American Gut Project Stool sample  \n",
       "1114     American Gut Project Mouth sample  \n",
       "1275     American Gut Project Stool sample  \n",
       "1364                                  None  \n",
       "1526     American Gut Project Stool sample  \n",
       "1687     American Gut Project Stool sample  \n",
       "2045     American Gut Project Stool sample  \n",
       "2646     American Gut Project Stool sample  \n",
       "3502                               no_data  \n",
       "3515     American Gut Project Stool sample  \n",
       "3664                               no_data  \n",
       "3725     American Gut Project Stool sample  \n",
       "4084     American Gut Project Stool sample  \n",
       "4266                               no_data  \n",
       "4580     American Gut Project Stool sample  \n",
       "4592     American Gut Project Stool sample  \n",
       "\n",
       "[20 rows x 202 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look for rows, that contain cells labelled as 'none'.\n",
    "\n",
    "df[(df == 'none').any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  res_values = method(rvalues)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(20, 202)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df == 'none').any(axis=1)].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Inspect all columns using **domain knowledge to decide which ones are relevant**, in order to select which columns to delete. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['#SampleID',\n",
       " 'BarcodeSequence',\n",
       " 'LinkerPrimerSequence',\n",
       " 'CARBOHYDRATE_PER',\n",
       " 'NONFOODALLERGIES_DRUG',\n",
       " 'NONFOODALLERGIES_BEESTINGS',\n",
       " 'ASSIGNED_FROM_GEO',\n",
       " 'RUN_DATE',\n",
       " 'PET_LOCATIONS',\n",
       " 'AGE',\n",
       " 'TOT_MASS',\n",
       " 'GENERAL_MEDS',\n",
       " 'SAMPLE_ID',\n",
       " 'ORIG_SAMPLE_NAME',\n",
       " 'QUINOLINE',\n",
       " 'BODY_SITE',\n",
       " 'MIGRAINE_FACTOR_2',\n",
       " 'MIGRAINE_FACTOR_3',\n",
       " 'TONSILS_REMOVED',\n",
       " 'MIGRAINE_FACTOR_1',\n",
       " 'ELEVATION',\n",
       " 'FLOSSING_FREQUENCY',\n",
       " 'PROTEIN_PER',\n",
       " 'CSECTION',\n",
       " 'PCR_PRIMERS',\n",
       " 'COUNTRY_OF_BIRTH',\n",
       " 'MACRONUTRIENT_PCT_TOTAL',\n",
       " 'COSMETICS_FREQUENCY',\n",
       " 'MASTERMIX_LOT',\n",
       " 'HOST_SUBJECT_ID',\n",
       " 'ENA-CHECKLIST',\n",
       " 'GLUTEN',\n",
       " 'SAMPLE_NAME',\n",
       " 'TM1000_8_TOOL',\n",
       " 'LONGITUDE',\n",
       " 'DOG',\n",
       " 'MIGRAINE_AGGRAVATION',\n",
       " 'ORIG_NAME',\n",
       " 'FRAT',\n",
       " 'DIABETES_MEDICATION',\n",
       " 'SAMPLE_PLATE',\n",
       " 'PENICILLIN',\n",
       " 'FOODALLERGIES_OTHER_TEXT',\n",
       " 'KEY_SEQ',\n",
       " 'BODY_PRODUCT',\n",
       " 'REGION',\n",
       " 'SAMPLE_TYPE',\n",
       " 'MIGRAINE_FREQUENCY',\n",
       " 'MIGRAINE_RELATIVES',\n",
       " 'EXPERIMENT_TITLE',\n",
       " 'PLANT_PER',\n",
       " 'EXTRACTIONKIT_LOT',\n",
       " 'SAMPLE',\n",
       " 'MIGRAINE_PHOTOPHOBIA',\n",
       " 'SITE_SAMPLED',\n",
       " 'DRINKING_WATER_SOURCE',\n",
       " 'CENTER_NAME',\n",
       " 'COUNTRY',\n",
       " 'ANTIBIOTIC_SELECT',\n",
       " 'MIGRAINE_NAUSEA',\n",
       " 'COMMON_NAME',\n",
       " 'FOODALLERGIES_TREENUTS',\n",
       " 'MIGRAINE_AURA',\n",
       " 'SEQUENCING_METH',\n",
       " 'FLU_VACCINE_DATE',\n",
       " 'PREGNANT',\n",
       " 'MIGRAINE',\n",
       " 'BMI',\n",
       " 'ENV_BIOME',\n",
       " 'WELL_ID',\n",
       " 'TEETHBRUSHING_FREQUENCY',\n",
       " 'MIGRAINE_MEDICATIONS',\n",
       " 'CONTRACEPTIVE',\n",
       " 'SAMPLE_CENTER',\n",
       " 'PHYSICAL_LOCATION',\n",
       " 'LIVINGWITH',\n",
       " 'LINKER',\n",
       " 'RELATIONS',\n",
       " 'DIABETES_MEDICATIONS',\n",
       " 'SUPPLEMENTS',\n",
       " 'AGE_UNIT',\n",
       " 'TANNING_SPRAYS',\n",
       " 'EXERCISE_FREQUENCY',\n",
       " 'HOST_COMMON_NAME',\n",
       " 'SULFA_DRUG',\n",
       " 'NONFOODALLERGIES_NO',\n",
       " 'ENV_MATTER',\n",
       " 'RACE_OTHER',\n",
       " 'SKIN_CONDITION',\n",
       " 'ALTITUDE',\n",
       " 'NONFOODALLERGIES_DANDER',\n",
       " 'CONDITIONS_MEDICATION',\n",
       " 'LAST_TRAVEL',\n",
       " 'TARGET_SUBFRAGMENT',\n",
       " 'TITLE',\n",
       " 'HEIGHT_IN',\n",
       " 'ANTIBIOTIC_CONDITION',\n",
       " 'PET_CONTACT',\n",
       " 'MIGRAINE_PAIN',\n",
       " 'ROOMMATES',\n",
       " 'CENTER_PROJECT_NAME',\n",
       " 'NONFOODALLERGIES_SUN',\n",
       " 'HAS_EXTRACTED_DATA',\n",
       " 'ENA-SPOT-COUNT',\n",
       " 'EXERCISE_LOCATION',\n",
       " 'RUN_ID',\n",
       " 'COLLECTION_DATE',\n",
       " 'NITROIMIDAZOLE',\n",
       " 'DEODORANT_USE',\n",
       " 'FOODALLERGIES_SHELLFISH',\n",
       " 'SEX',\n",
       " 'PERCENTAGE_FROM_CARBS',\n",
       " 'FAT_PER',\n",
       " 'TARGET_GENE',\n",
       " 'PREGNANT_DUE_DATE',\n",
       " 'SHARED_HOUSING',\n",
       " 'SOFTENER',\n",
       " 'MIGRAINE_PHONOPHOBIA',\n",
       " 'SAMPLE_DESCRIPTION',\n",
       " 'SAMP_SIZE',\n",
       " 'SMOKING_FREQUENCY',\n",
       " 'WEIGHT_LBS',\n",
       " 'TANNING_BEDS',\n",
       " 'ALCOHOL_FREQUENCY',\n",
       " 'STATE',\n",
       " 'HAS_PHYSICAL_SPECIMEN',\n",
       " 'STUDY_CENTER',\n",
       " 'BODY_HABITAT',\n",
       " 'DOMINANT_HAND',\n",
       " 'PRIMARY_VEGETABLE',\n",
       " 'WATER_LOT',\n",
       " 'SPECIAL_RESTRICTIONS',\n",
       " 'CEPHALOSPORIN',\n",
       " 'ENV_FEATURE',\n",
       " 'SUPPLEMENTS_FIELDS',\n",
       " 'ENA-BASE-COUNT',\n",
       " 'SAMPLE_TIME',\n",
       " 'APPENDIX_REMOVED',\n",
       " 'CURRENT_RESIDENCE_DURATION',\n",
       " 'EMP_STATUS',\n",
       " 'COMMUNAL_DINING',\n",
       " 'SLEEP_DURATION',\n",
       " 'ANONYMIZED_NAME',\n",
       " 'PUBLIC',\n",
       " 'DIET_TYPE',\n",
       " 'RUN_CENTER',\n",
       " 'DIET_RESTRICTIONS',\n",
       " 'EXPERIMENT_CENTER',\n",
       " 'LACTOSE',\n",
       " 'HEIGHT_OR_LENGTH',\n",
       " 'ANTIBIOTIC_MEDS',\n",
       " 'IBD',\n",
       " 'TM50_8_TOOL',\n",
       " 'DEPTH',\n",
       " 'PROCESSING_ROBOT',\n",
       " 'HOST_TAXID',\n",
       " 'ILLUMINA_TECHNOLOGY',\n",
       " 'TAXON_ID',\n",
       " 'DIABETES_DIAGNOSE_DATE',\n",
       " 'REQUIRED_SAMPLE_INFO_STATUS',\n",
       " 'TYPES_OF_PLANTS',\n",
       " 'MIGRAINEMEDS',\n",
       " 'FIBER_GRAMS',\n",
       " 'MAINFACTOR_OTHER_3',\n",
       " 'PLATING',\n",
       " 'PROJECT_NAME',\n",
       " 'COLLECTION_TIMESTAMP',\n",
       " 'NAILS',\n",
       " 'RUN_PREFIX',\n",
       " 'DESCRIPTION_WELL',\n",
       " 'PLATFORM',\n",
       " 'RACE',\n",
       " 'PETS',\n",
       " 'MULTIVITAMIN',\n",
       " 'EXTRACTION_ROBOT',\n",
       " 'NONFOODALLERGIES_POISONIVY',\n",
       " 'ACNE_MEDICATION',\n",
       " 'PRIMER_PLATE',\n",
       " 'ACNE_MEDICATION_OTC',\n",
       " 'ANIMAL_PER',\n",
       " 'MACHINE',\n",
       " 'TM300_8_TOOL',\n",
       " 'MAINFACTOR_OTHER_1',\n",
       " 'EXPERIMENT_DESIGN_DESCRIPTION',\n",
       " 'FOODALLERGIES_OTHER',\n",
       " 'WEIGHT_CHANGE',\n",
       " 'PRIMER_DATE',\n",
       " 'PRIMARY_CARB',\n",
       " 'DIABETES',\n",
       " 'MAINFACTOR_OTHER_2',\n",
       " 'ASTHMA',\n",
       " 'POOL_FREQUENCY',\n",
       " 'LIBRARY_CONSTRUCTION_PROTOCOL',\n",
       " 'PKU',\n",
       " 'DECEASED_PARENT',\n",
       " 'SEQ_RUN',\n",
       " 'CAT',\n",
       " 'LATITUDE',\n",
       " 'FOODALLERGIES_PEANUTS',\n",
       " 'CHICKENPOX',\n",
       " 'SEASONAL_ALLERGIES',\n",
       " 'Description']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the column names.\n",
    "   \n",
    "list(df.columns.values) \n",
    "\n",
    "# https://www.geeksforgeeks.org/how-to-get-column-names-in-pandas-dataframe/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                #SampleID BarcodeSequence LinkerPrimerSequence  \\\n",
      "0             11224.12109         no_data              no_data   \n",
      "1             5034.107635         no_data              no_data   \n",
      "2             11351.12107         no_data              no_data   \n",
      "3             2013.113155         no_data              no_data   \n",
      "4             4731.107641         no_data              no_data   \n",
      "...                   ...             ...                  ...   \n",
      "4810   000013296.fixed213         no_data              no_data   \n",
      "4811  000007822.fixed1101         no_data              no_data   \n",
      "4812          9074.121304         no_data              no_data   \n",
      "4813          2418.128313         no_data              no_data   \n",
      "4814          7179.126173         no_data              no_data   \n",
      "\n",
      "     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS  \\\n",
      "0                  55               no_data                    no_data   \n",
      "1                  56               no_data                    no_data   \n",
      "2                  16               no_data                    no_data   \n",
      "3             no_data               no_data                    no_data   \n",
      "4                  61                   yes                    no_data   \n",
      "...               ...                   ...                        ...   \n",
      "4810               50               unknown                    unknown   \n",
      "4811               17               unknown                    unknown   \n",
      "4812               34               no_data                    no_data   \n",
      "4813               30               no_data                    no_data   \n",
      "4814               60               no_data                    no_data   \n",
      "\n",
      "     ASSIGNED_FROM_GEO RUN_DATE PET_LOCATIONS AGE TOT_MASS GENERAL_MEDS  \\\n",
      "0                    n  no_data    Housing...  51       50      no_data   \n",
      "1                    n  no_data       no_data  35       63      no_data   \n",
      "2                    n  no_data    Housing...  56       64      no_data   \n",
      "3                    n  no_data    Housing...  37       90      no_data   \n",
      "4                    n  no_data       no_data  72       86      no_data   \n",
      "...                ...      ...           ...  ..      ...          ...   \n",
      "4810                 y  no_data    Housing...  71       80      unknown   \n",
      "4811                 y  no_data    Housing...   9       18      unknown   \n",
      "4812                 n  no_data    Housing...  29       79      no_data   \n",
      "4813                 y  no_data       no_data  46       83      no_data   \n",
      "4814                 y  no_data    Housing...   1       73      no_data   \n",
      "\n",
      "     SAMPLE_ID ORIG_SAMPLE_NAME QUINOLINE      BODY_SITE MIGRAINE_FACTOR_2  \\\n",
      "0      1202271          no_data        no   UBERON:feces           no_data   \n",
      "1      1068981          no_data   no_data   UBERON:feces           no_data   \n",
      "2      1202422          no_data        no   UBERON:feces           no_data   \n",
      "3      1123170          no_data        no   UBERON:feces           no_data   \n",
      "4      1068783          no_data   no_data   UBERON:feces           no_data   \n",
      "...        ...              ...       ...            ...               ...   \n",
      "4810   no_data          no_data       yes   UBERON:feces           unknown   \n",
      "4811   no_data          no_data        no  UBERON:tongue           unknown   \n",
      "4812   1204707          no_data        no   UBERON:feces            Stress   \n",
      "4813   1269945          no_data        no   UBERON:feces           no_data   \n",
      "4814   1249633          no_data        no   UBERON:feces           no_data   \n",
      "\n",
      "     MIGRAINE_FACTOR_3 TONSILS_REMOVED MIGRAINE_FACTOR_1 ELEVATION  \\\n",
      "0              no_data         no_data           no_data    1654.9   \n",
      "1              no_data              no           no_data   no_data   \n",
      "2              no_data             yes           no_data    1624.1   \n",
      "3              no_data              no           no_data   no_data   \n",
      "4              no_data              no           no_data   no_data   \n",
      "...                ...             ...               ...       ...   \n",
      "4810           unknown             yes           unknown       223   \n",
      "4811           unknown              no           unknown     101.9   \n",
      "4812           no_data              no             Other      61.8   \n",
      "4813           no_data              no           no_data     181.7   \n",
      "4814           no_data              no           no_data       4.4   \n",
      "\n",
      "              FLOSSING_FREQUENCY PROTEIN_PER  CSECTION PCR_PRIMERS  \\\n",
      "0                          Daily          15        no     no_data   \n",
      "1     Regularly (3-5 times/week)          22        no     no_data   \n",
      "2                          Daily          17        no     no_data   \n",
      "3                          Never     no_data       yes     no_data   \n",
      "4                          Daily          16        no     no_data   \n",
      "...                          ...         ...       ...         ...   \n",
      "4810                       Daily          29        no     no_data   \n",
      "4811                       Never          47       yes     no_data   \n",
      "4812                       Never          18        no     no_data   \n",
      "4813    Rarely (few times/month)          40  not_sure     no_data   \n",
      "4814                       Daily          16        no     no_data   \n",
      "\n",
      "     COUNTRY_OF_BIRTH MACRONUTRIENT_PCT_TOTAL       COSMETICS_FREQUENCY  \\\n",
      "0                 USA                     100                     Daily   \n",
      "1                 USA                      99  Rarely (few times/month)   \n",
      "2                 USA                     100                     Never   \n",
      "3                 USA                      80                     Never   \n",
      "4                 USA                     100                     Never   \n",
      "...               ...                     ...                       ...   \n",
      "4810              USA                     100                     Never   \n",
      "4811          Belgium                     100                     Never   \n",
      "4812              USA                     100                     Never   \n",
      "4813        Australia                     100                     Never   \n",
      "4814         Thailand                     100                     Never   \n",
      "\n",
      "     MASTERMIX_LOT  HOST_SUBJECT_ID ENA-CHECKLIST GLUTEN SAMPLE_NAME  \\\n",
      "0        145042558  2186:4193172313     ERC000011     no       11224   \n",
      "1          no_data  1978:3545090407     ERC000011     no        5034   \n",
      "2        145042558  2186:3977480147     ERC000011    yes       11351   \n",
      "3        145032369  2037:2553448961     ERC000011     no        2013   \n",
      "4          no_data  1978:1009910171     ERC000011     no        4731   \n",
      "...            ...              ...           ...    ...         ...   \n",
      "4810       no_data       3141083688     ERC000011     no     no_data   \n",
      "4811       no_data       2880131424     ERC000011     no     no_data   \n",
      "4812       no_data  2191:2907824909     ERC000011    yes        9074   \n",
      "4813       no_data  2381:1882060662     ERC000011     no        2418   \n",
      "4814     148014030  2254:2556430608     ERC000011     no        7179   \n",
      "\n",
      "     TM1000_8_TOOL LONGITUDE      DOG MIGRAINE_AGGRAVATION ORIG_NAME     FRAT  \\\n",
      "0          311169B    -104.9  no_data              no_data   no_data  no_data   \n",
      "1          no_data    -117.2       no              no_data   no_data  no_data   \n",
      "2          311169B    -105.3       no              no_data   no_data  no_data   \n",
      "3          311172B       -74       no              no_data   no_data  no_data   \n",
      "4          no_data     -74.3       no              no_data   no_data      yes   \n",
      "...            ...       ...      ...                  ...       ...      ...   \n",
      "4810       no_data     -88.3       no              unknown   no_data  unknown   \n",
      "4811       no_data       4.4       no              unknown   no_data  unknown   \n",
      "4812       no_data    -122.3       no              no_data   no_data  no_data   \n",
      "4813       no_data     -87.7       no              no_data      2418  no_data   \n",
      "4814       311172B     100.5      yes              no_data      7179  no_data   \n",
      "\n",
      "     DIABETES_MEDICATION                SAMPLE_PLATE PENICILLIN  \\\n",
      "0                no_data                  AG_Plate36         no   \n",
      "1                no_data               AmericanGut10    no_data   \n",
      "2                no_data                  AG_Plate36         no   \n",
      "3                no_data                     Fecal19         no   \n",
      "4                no_data               AmericanGut15    no_data   \n",
      "...                  ...                         ...        ...   \n",
      "4810             unknown                     no_data         no   \n",
      "4811             unknown                     no_data        yes   \n",
      "4812             no_data                     no_data         no   \n",
      "4813             no_data  American_Gut_Retest_Plate2         no   \n",
      "4814             no_data                       AG_47         no   \n",
      "\n",
      "         FOODALLERGIES_OTHER_TEXT  KEY_SEQ   BODY_PRODUCT   REGION  \\\n",
      "0                         no_data  no_data   UBERON:feces  no_data   \n",
      "1                         no_data  no_data   UBERON:feces  no_data   \n",
      "2     Cats, dogs, certain grasses  no_data   UBERON:feces  no_data   \n",
      "3                         no_data  no_data   UBERON:feces  no_data   \n",
      "4                            dust  no_data   UBERON:feces  no_data   \n",
      "...                           ...      ...            ...      ...   \n",
      "4810                      unknown  no_data   UBERON:feces  no_data   \n",
      "4811                      unknown  no_data  UBERON:saliva  no_data   \n",
      "4812                      no_data  no_data   UBERON:feces  no_data   \n",
      "4813                      no_data  no_data   UBERON:feces  no_data   \n",
      "4814                      no_data  no_data   UBERON:feces  no_data   \n",
      "\n",
      "     SAMPLE_TYPE MIGRAINE_FREQUENCY MIGRAINE_RELATIVES EXPERIMENT_TITLE  \\\n",
      "0        no_data            no_data            no_data          no_data   \n",
      "1        no_data            no_data            no_data          no_data   \n",
      "2        no_data            no_data            no_data          no_data   \n",
      "3        no_data            no_data            no_data          no_data   \n",
      "4        no_data            no_data            no_data          no_data   \n",
      "...          ...                ...                ...              ...   \n",
      "4810       Stool            unknown            unknown          no_data   \n",
      "4811       Mouth            unknown            unknown          no_data   \n",
      "4812     no_data            Monthly                yes          no_data   \n",
      "4813     no_data            no_data            no_data          no_data   \n",
      "4814     no_data            no_data            no_data          no_data   \n",
      "\n",
      "     PLANT_PER EXTRACTIONKIT_LOT   SAMPLE MIGRAINE_PHOTOPHOBIA SITE_SAMPLED  \\\n",
      "0          100            PS13J2  no_data              no_data        Stool   \n",
      "1           67           no_data  no_data              no_data        Stool   \n",
      "2           26            PS13J2  no_data              no_data        Stool   \n",
      "3           99          12411155     2013              no_data        Stool   \n",
      "4           34           no_data  no_data              no_data        Stool   \n",
      "...        ...               ...      ...                  ...          ...   \n",
      "4810   unknown           no_data  no_data              unknown      no_data   \n",
      "4811        30           no_data  no_data              unknown      no_data   \n",
      "4812        10           no_data     9074              no_data        Stool   \n",
      "4813        90           no_data  no_data              no_data        Stool   \n",
      "4814        50          PS14C255  no_data              no_data        Stool   \n",
      "\n",
      "     DRINKING_WATER_SOURCE CENTER_NAME                       COUNTRY  \\\n",
      "0                     City     no_data  GAZ:United States of America   \n",
      "1                 Filtered     no_data  GAZ:United States of America   \n",
      "2                     City     no_data  GAZ:United States of America   \n",
      "3                     City     no_data  GAZ:United States of America   \n",
      "4                     City     no_data  GAZ:United States of America   \n",
      "...                    ...         ...                           ...   \n",
      "4810                  City     no_data  GAZ:United States of America   \n",
      "4811              Filtered     no_data                   GAZ:Belgium   \n",
      "4812              Filtered     no_data  GAZ:United States of America   \n",
      "4813                  City     no_data  GAZ:United States of America   \n",
      "4814              Filtered     no_data                  GAZ:Thailand   \n",
      "\n",
      "         ANTIBIOTIC_SELECT MIGRAINE_NAUSEA            COMMON_NAME  \\\n",
      "0     Not in the last year         no_data   human gut metagenome   \n",
      "1     Not in the last year         no_data   human gut metagenome   \n",
      "2     Not in the last year         no_data   human gut metagenome   \n",
      "3     Not in the last year         no_data   human gut metagenome   \n",
      "4     In the past 6 months         no_data   human gut metagenome   \n",
      "...                    ...             ...                    ...   \n",
      "4810  In the past 6 months         unknown   human gut metagenome   \n",
      "4811      In the past year         unknown  human oral metagenome   \n",
      "4812  Not in the last year             yes   human gut metagenome   \n",
      "4813  Not in the last year         no_data   human gut metagenome   \n",
      "4814  Not in the last year         no_data   human gut metagenome   \n",
      "\n",
      "     FOODALLERGIES_TREENUTS MIGRAINE_AURA SEQUENCING_METH  \\\n",
      "0                   no_data       no_data         no_data   \n",
      "1                   no_data       no_data         no_data   \n",
      "2                       yes       no_data         no_data   \n",
      "3                   no_data       no_data         no_data   \n",
      "4                   no_data       no_data         no_data   \n",
      "...                     ...           ...             ...   \n",
      "4810                unknown       unknown         no_data   \n",
      "4811                unknown       unknown         no_data   \n",
      "4812                no_data           yes         no_data   \n",
      "4813                no_data       no_data         no_data   \n",
      "4814                no_data       no_data         no_data   \n",
      "\n",
      "          FLU_VACCINE_DATE PREGNANT MIGRAINE          BMI  \\\n",
      "0     Not in the last year       no       no  18.87939453   \n",
      "1     In the past 6 months  unknown       no  24.79717813   \n",
      "2     Not in the last year       no       no  21.74070069   \n",
      "3         In the past year       no       no  26.38393695   \n",
      "4         In the past year  unknown       no  32.60986328   \n",
      "...                    ...      ...      ...          ...   \n",
      "4810  In the past 6 months       no       no  25.53755102   \n",
      "4811  Not in the last year       no       no  12.72974196   \n",
      "4812  Not in the last year       no      yes  21.87111111   \n",
      "4813      In the past year       no       no  26.54183673   \n",
      "4814  In the past 6 months       no       no  24.47729239   \n",
      "\n",
      "                          ENV_BIOME  WELL_ID TEETHBRUSHING_FREQUENCY  \\\n",
      "0                  ENVO:urban biome      B11                   Daily   \n",
      "1     ENVO:human-associated habitat       B1                   Daily   \n",
      "2                  ENVO:urban biome      D10                   Daily   \n",
      "3     ENVO:human-associated habitat      B10                   Daily   \n",
      "4     ENVO:human-associated habitat       D4                   Daily   \n",
      "...                             ...      ...                     ...   \n",
      "4810               ENVO:urban biome  no_data                   Daily   \n",
      "4811               ENVO:urban biome  no_data                   Daily   \n",
      "4812               ENVO:urban biome  no_data                   Daily   \n",
      "4813               ENVO:urban biome  no_data                   Daily   \n",
      "4814               ENVO:urban biome       E6                   Daily   \n",
      "\n",
      "     MIGRAINE_MEDICATIONS CONTRACEPTIVE SAMPLE_CENTER PHYSICAL_LOCATION  \\\n",
      "0                 no_data            No       no_data           no_data   \n",
      "1                 no_data            No       no_data           no_data   \n",
      "2                 no_data       no_data       no_data           no_data   \n",
      "3                 no_data       no_data       no_data           no_data   \n",
      "4                 no_data            No       no_data           no_data   \n",
      "...                   ...           ...           ...               ...   \n",
      "4810              unknown       unknown       no_data              CCME   \n",
      "4811              unknown       unknown       no_data              UCSD   \n",
      "4812              no_data       no_data       no_data           no_data   \n",
      "4813              no_data       no_data       no_data           no_data   \n",
      "4814              no_data       no_data       no_data           no_data   \n",
      "\n",
      "     LIVINGWITH   LINKER               RELATIONS DIABETES_MEDICATIONS  \\\n",
      "0            no  no_data                 no_data              no_data   \n",
      "1            no  no_data                 no_data              no_data   \n",
      "2            no  no_data                 no_data              no_data   \n",
      "3           yes  no_data              Girlfriend              no_data   \n",
      "4           yes  no_data                daughter              no_data   \n",
      "...         ...      ...                     ...                  ...   \n",
      "4810         no  no_data                 unknown              unknown   \n",
      "4811        yes  no_data  Father, Mother, Sister              unknown   \n",
      "4812         no  no_data                 no_data              no_data   \n",
      "4813        yes  no_data                    Wife              no_data   \n",
      "4814         no  no_data                 no_data              no_data   \n",
      "\n",
      "     SUPPLEMENTS AGE_UNIT TANNING_SPRAYS             EXERCISE_FREQUENCY  \\\n",
      "0            yes    years        no_data                          Daily   \n",
      "1            yes    years        no_data     Regularly (3-5 times/week)   \n",
      "2            yes    years        no_data                          Daily   \n",
      "3            yes    years        no_data                          Never   \n",
      "4            yes    years        no_data                          Daily   \n",
      "...          ...      ...            ...                            ...   \n",
      "4810         yes    years        unknown                          Daily   \n",
      "4811          no    years        unknown  Occasionally (1-2 times/week)   \n",
      "4812         yes    years        no_data                          Daily   \n",
      "4813         yes    years        no_data                          Daily   \n",
      "4814          no    years        no_data                          Daily   \n",
      "\n",
      "     HOST_COMMON_NAME SULFA_DRUG NONFOODALLERGIES_NO   ENV_MATTER RACE_OTHER  \\\n",
      "0               human         no             no_data   ENVO:feces    no_data   \n",
      "1               human    no_data                 yes   ENVO:feces    no_data   \n",
      "2               human         no             no_data   ENVO:feces    no_data   \n",
      "3               human         no             no_data   ENVO:feces    no_data   \n",
      "4               human    no_data             no_data   ENVO:feces    no_data   \n",
      "...               ...        ...                 ...          ...        ...   \n",
      "4810            human         no             unknown   ENVO:feces    unknown   \n",
      "4811            human         no                 yes  ENVO:saliva    unknown   \n",
      "4812            human         no             no_data   ENVO:feces    no_data   \n",
      "4813            human         no             no_data   ENVO:feces    no_data   \n",
      "4814            human         no                 yes   ENVO:feces    no_data   \n",
      "\n",
      "                      SKIN_CONDITION  ALTITUDE NONFOODALLERGIES_DANDER  \\\n",
      "0     I do not have a skin condition         0                 no_data   \n",
      "1     I do not have a skin condition         0                 no_data   \n",
      "2                            Rosacea         0                     yes   \n",
      "3     I do not have a skin condition         0                 no_data   \n",
      "4     I do not have a skin condition         0                     yes   \n",
      "...                              ...       ...                     ...   \n",
      "4810  I do not have a skin condition         0                 unknown   \n",
      "4811  I do not have a skin condition         0                 unknown   \n",
      "4812                          Eczema         0                 no_data   \n",
      "4813  I do not have a skin condition         0                 no_data   \n",
      "4814  I do not have a skin condition         0                 no_data   \n",
      "\n",
      "     CONDITIONS_MEDICATION               LAST_TRAVEL TARGET_SUBFRAGMENT  \\\n",
      "0                       no    More than one year ago            no_data   \n",
      "1                       no      Within the past year            no_data   \n",
      "2                      yes    More than one year ago            no_data   \n",
      "3                       no  Within the past 3 months            no_data   \n",
      "4                       no    More than one year ago            no_data   \n",
      "...                    ...                       ...                ...   \n",
      "4810                    no    More than one year ago            no_data   \n",
      "4811                    no                   unknown            no_data   \n",
      "4812                    no  Within the past 6 months            no_data   \n",
      "4813                    no      Within the past year            no_data   \n",
      "4814                   yes  Within the past 3 months            no_data   \n",
      "\n",
      "                     TITLE HEIGHT_IN ANTIBIOTIC_CONDITION PET_CONTACT  \\\n",
      "0     American Gut Project        64              no_data  Contact...   \n",
      "1     American Gut Project        63              no_data     no_data   \n",
      "2     American Gut Project        68              no_data  Contact...   \n",
      "3     American Gut Project        73              no_data  Contact...   \n",
      "4     American Gut Project        64      dental cleaning     no_data   \n",
      "...                    ...       ...                  ...         ...   \n",
      "4810  American Gut Project        70                  UTI  Contact...   \n",
      "4811  American Gut Project        47      Streptococcus A  Contact...   \n",
      "4812  American Gut Project        75              no_data  Contact...   \n",
      "4813  American Gut Project        70              no_data     no_data   \n",
      "4814  American Gut Project        68              no_data  Contact...   \n",
      "\n",
      "     MIGRAINE_PAIN ROOMMATES CENTER_PROJECT_NAME NONFOODALLERGIES_SUN  \\\n",
      "0          no_data   no_data             no_data              no_data   \n",
      "1          no_data   no_data             no_data              no_data   \n",
      "2          no_data   no_data             no_data              no_data   \n",
      "3          no_data   no_data             no_data              no_data   \n",
      "4          no_data   no_data             no_data              no_data   \n",
      "...            ...       ...                 ...                  ...   \n",
      "4810       unknown   unknown             no_data              unknown   \n",
      "4811       unknown   unknown             no_data              unknown   \n",
      "4812       no_data   no_data             no_data              no_data   \n",
      "4813       no_data   no_data             no_data              no_data   \n",
      "4814       no_data   no_data             no_data              no_data   \n",
      "\n",
      "     HAS_EXTRACTED_DATA ENA-SPOT-COUNT EXERCISE_LOCATION   RUN_ID  \\\n",
      "0               no_data          49024              Both  no_data   \n",
      "1               no_data          38689              Both  no_data   \n",
      "2               no_data          31111          Outdoors  no_data   \n",
      "3               no_data          19149           no_data  no_data   \n",
      "4               no_data          42137          Outdoors  no_data   \n",
      "...                 ...            ...               ...      ...   \n",
      "4810               TRUE          73468           Indoors  no_data   \n",
      "4811               TRUE        no_data           Indoors  no_data   \n",
      "4812            no_data          25152              Both  no_data   \n",
      "4813            no_data          58676              Both  no_data   \n",
      "4814            no_data          22130          Outdoors  no_data   \n",
      "\n",
      "     COLLECTION_DATE NITROIMIDAZOLE  \\\n",
      "0           12/22/13             no   \n",
      "1         05/29/2013        no_data   \n",
      "2           12/12/13             no   \n",
      "3           08/05/13             no   \n",
      "4         05/13/2013        no_data   \n",
      "...              ...            ...   \n",
      "4810         1/29/14             no   \n",
      "4811      08/20/2013             no   \n",
      "4812         9/21/13             no   \n",
      "4813         4/28/13             no   \n",
      "4814         9/18/13             no   \n",
      "\n",
      "                                          DEODORANT_USE  \\\n",
      "0                                       I use deoderant   \n",
      "1     Not sure, but I use some form of deoderant/ant...   \n",
      "2                                       I use deoderant   \n",
      "3                                       I use deoderant   \n",
      "4                                       I use deoderant   \n",
      "...                                                 ...   \n",
      "4810           I do not use deoderant or antiperspirant   \n",
      "4811           I do not use deoderant or antiperspirant   \n",
      "4812           I do not use deodorant or antiperspirant   \n",
      "4813           I do not use deoderant or antiperspirant   \n",
      "4814                                    I use deoderant   \n",
      "\n",
      "     FOODALLERGIES_SHELLFISH      SEX PERCENTAGE_FROM_CARBS  FAT_PER  \\\n",
      "0                    no_data   female         Less than 25%       30   \n",
      "1                    no_data   female         Less than 25%       21   \n",
      "2                    no_data     male         Less than 25%       67   \n",
      "3                    no_data     male             51 to 75%  no_data   \n",
      "4                    no_data   female             25 to 50%       23   \n",
      "...                      ...      ...                   ...      ...   \n",
      "4810                 unknown     male         Less than 25%       21   \n",
      "4811                 unknown  unknown             51 to 75%       36   \n",
      "4812                 no_data     male         Less than 25%       48   \n",
      "4813                 no_data     male         Less than 25%       30   \n",
      "4814                 no_data     male         Less than 25%       24   \n",
      "\n",
      "     TARGET_GENE PREGNANT_DUE_DATE SHARED_HOUSING SOFTENER  \\\n",
      "0        no_data           no_data             no      yes   \n",
      "1        no_data           no_data             no      yes   \n",
      "2        no_data           no_data             no       no   \n",
      "3        no_data           no_data             no       no   \n",
      "4        no_data           no_data            yes       no   \n",
      "...          ...               ...            ...      ...   \n",
      "4810     no_data           unknown             no       no   \n",
      "4811     no_data           unknown             no       no   \n",
      "4812     no_data           no_data             no       no   \n",
      "4813     no_data           no_data             no      yes   \n",
      "4814     no_data           no_data             no       no   \n",
      "\n",
      "     MIGRAINE_PHONOPHOBIA SAMPLE_DESCRIPTION  SAMP_SIZE SMOKING_FREQUENCY  \\\n",
      "0                 no_data            no_data       .1,g             Never   \n",
      "1                 no_data            no_data       .1,g             Never   \n",
      "2                 no_data            no_data       .1,g             Never   \n",
      "3                 no_data   Fecal19_2013_B10      0.1,g           no_data   \n",
      "4                 no_data            no_data       .1,g             Never   \n",
      "...                   ...                ...        ...               ...   \n",
      "4810              unknown            no_data    no_data             Never   \n",
      "4811              unknown            no_data    no_data             Never   \n",
      "4812              no_data            no_data       .1,g             Never   \n",
      "4813              no_data            no_data       .1,g             Never   \n",
      "4814              no_data            no_data  0.1, gram             Never   \n",
      "\n",
      "     WEIGHT_LBS TANNING_BEDS              ALCOHOL_FREQUENCY    STATE  \\\n",
      "0           110      no_data                          Never       CO   \n",
      "1           140      no_data                          Daily       CA   \n",
      "2           143      no_data                          Never       CO   \n",
      "3           200      no_data                          Daily       NY   \n",
      "4           190      no_data                          Never       NY   \n",
      "...         ...          ...                            ...      ...   \n",
      "4810        178      unknown                          Never       IL   \n",
      "4811         40      unknown                          Never  unknown   \n",
      "4812        175      no_data       Rarely (few times/month)       WA   \n",
      "4813        185      no_data  Occasionally (1-2 times/week)       IL   \n",
      "4814        161      no_data       Rarely (few times/month)   BANKOK   \n",
      "\n",
      "     HAS_PHYSICAL_SPECIMEN STUDY_CENTER        BODY_HABITAT DOMINANT_HAND  \\\n",
      "0                  no_data      no_data        UBERON:feces          left   \n",
      "1                  no_data      no_data        UBERON:feces          left   \n",
      "2                  no_data      no_data        UBERON:feces         right   \n",
      "3                  no_data      no_data        UBERON:feces         right   \n",
      "4                  no_data      no_data        UBERON:feces         right   \n",
      "...                    ...          ...                 ...           ...   \n",
      "4810                  TRUE      no_data        UBERON:feces          left   \n",
      "4811                  TRUE      no_data  UBERON:oral cavity         right   \n",
      "4812               no_data      no_data        UBERON:feces         right   \n",
      "4813               no_data      no_data        UBERON:feces         right   \n",
      "4814               no_data      no_data        UBERON:feces         right   \n",
      "\n",
      "                    PRIMARY_VEGETABLE        WATER_LOT SPECIAL_RESTRICTIONS  \\\n",
      "0                             spinach         RNBC2791                   no   \n",
      "1     tomatoes (I eat a lot of salsa)          no_data                   no   \n",
      "2                        leafy greens         RNBC2791                  yes   \n",
      "3                              Potato  RNBC2791.091222                   no   \n",
      "4                            broccoli          no_data                   no   \n",
      "...                               ...              ...                  ...   \n",
      "4810                  Romaine Lettuce          no_data                   no   \n",
      "4811              tomatoes, cuncumber          no_data                   no   \n",
      "4812                           potato          no_data                  yes   \n",
      "4813                     Leafy Greens          no_data                   no   \n",
      "4814                Leaves and fruits         RNBC9352                   no   \n",
      "\n",
      "     CEPHALOSPORIN                    ENV_FEATURE  \\\n",
      "0               no  ENVO:human-associated habitat   \n",
      "1          no_data  ENVO:human-associated habitat   \n",
      "2               no  ENVO:human-associated habitat   \n",
      "3               no  ENVO:human-associated habitat   \n",
      "4          no_data  ENVO:human-associated habitat   \n",
      "...            ...                            ...   \n",
      "4810            no  ENVO:human-associated habitat   \n",
      "4811            no  ENVO:human-associated habitat   \n",
      "4812            no  ENVO:human-associated habitat   \n",
      "4813            no  ENVO:human-associated habitat   \n",
      "4814            no  ENVO:human-associated habitat   \n",
      "\n",
      "                                     SUPPLEMENTS_FIELDS ENA-BASE-COUNT  \\\n",
      "0                             calcium, probiotics, maca        7402624   \n",
      "1                vitamin B complex, Omega 3 fatty acids        5842005   \n",
      "2     Kombucha, Probiotic supplements, Niacin, CoEnz...        4697761   \n",
      "3                                        pro/prebiotics        2891499   \n",
      "4                                      calcium, vitamin        6362610   \n",
      "...                                                 ...            ...   \n",
      "4810  Cranberry 500 mg, Probiotic Caps, Lactobacillu...       11093668   \n",
      "4811                                            unknown        no_data   \n",
      "4812                                           b6, zinc        3797952   \n",
      "4813                     Fish Oil, Zyflamend Whole Body        8860076   \n",
      "4814                                            no_data        3341630   \n",
      "\n",
      "     SAMPLE_TIME APPENDIX_REMOVED CURRENT_RESIDENCE_DURATION EMP_STATUS  \\\n",
      "0        7:50 am               no     More than one year ago    no_data   \n",
      "1       10:40 am               no     More than one year ago    no_data   \n",
      "2        7:00 am               no     More than one year ago    no_data   \n",
      "3        8:48 pm               no     More than one year ago    no_data   \n",
      "4        8:00 am               no     More than one year ago    no_data   \n",
      "...          ...              ...                        ...        ...   \n",
      "4810     3:30 pm               no     More than one year ago    no_data   \n",
      "4811     8:40 pm               no                    unknown    no_data   \n",
      "4812    12:20 pm               no     More than one year ago    no_data   \n",
      "4813    10:25 am               no     More than one year ago    no_data   \n",
      "4814     5:45 am               no     More than one year ago    no_data   \n",
      "\n",
      "     COMMUNAL_DINING     SLEEP_DURATION ANONYMIZED_NAME PUBLIC  \\\n",
      "0            no_data          6-7 hours           11224      y   \n",
      "1            no_data  Less than 5 hours            5034      n   \n",
      "2            no_data    8 or more hours           11351      y   \n",
      "3            no_data          6-7 hours            2013      n   \n",
      "4                yes          5-6 hours            4731      n   \n",
      "...              ...                ...             ...    ...   \n",
      "4810         unknown          6-7 hours           13296      y   \n",
      "4811         unknown    8 or more hours            7822      y   \n",
      "4812         no_data          7-8 hours            9074      y   \n",
      "4813         no_data          6-7 hours            2418      y   \n",
      "4814         no_data          6-7 hours            7179      y   \n",
      "\n",
      "                       DIET_TYPE RUN_CENTER  \\\n",
      "0                          Vegan    no_data   \n",
      "1                       Omnivore    no_data   \n",
      "2                       Omnivore    no_data   \n",
      "3     Vegetarian but eat seafood    no_data   \n",
      "4       Omnivore but no red meat    no_data   \n",
      "...                          ...        ...   \n",
      "4810                    Omnivore    no_data   \n",
      "4811                    Omnivore    no_data   \n",
      "4812                    Omnivore    no_data   \n",
      "4813                    Omnivore    no_data   \n",
      "4814                    Omnivore    no_data   \n",
      "\n",
      "                                      DIET_RESTRICTIONS EXPERIMENT_CENTER  \\\n",
      "0                                               no_data           no_data   \n",
      "1                                               no_data           no_data   \n",
      "2                                            Grain free           no_data   \n",
      "3                                               no_data           no_data   \n",
      "4                                               no_data           no_data   \n",
      "...                                                 ...               ...   \n",
      "4810                                            unknown           no_data   \n",
      "4811                                            unknown           no_data   \n",
      "4812  Ray Peat diet. no grains, beans, minimal veget...           no_data   \n",
      "4813                                            no_data           no_data   \n",
      "4814                                            no_data           no_data   \n",
      "\n",
      "     LACTOSE HEIGHT_OR_LENGTH ANTIBIOTIC_MEDS                IBD TM50_8_TOOL  \\\n",
      "0         no              163         no_data  I do not have IBD     311313B   \n",
      "1         no              160         no_data  I do not have IBD     no_data   \n",
      "2        yes              173         no_data  I do not have IBD     311313B   \n",
      "3         no              186         no_data  I do not have IBD     311426B   \n",
      "4         no              163         no_data  I do not have IBD     no_data   \n",
      "...      ...              ...             ...                ...         ...   \n",
      "4810      no              178           CIPRO  I do not have IBD     no_data   \n",
      "4811      no              120     amoxicillin  I do not have IBD     no_data   \n",
      "4812      no              191         no_data  I do not have IBD     no_data   \n",
      "4813      no              178         no_data  I do not have IBD     no_data   \n",
      "4814      no              175         no_data  I do not have IBD     111075B   \n",
      "\n",
      "      DEPTH PROCESSING_ROBOT HOST_TAXID ILLUMINA_TECHNOLOGY TAXON_ID  \\\n",
      "0         0             ROBE       9606             no_data   408170   \n",
      "1         0          no_data       9606             no_data   408170   \n",
      "2         0             ROBE       9606             no_data   408170   \n",
      "3         0             RikE       9606             no_data   408170   \n",
      "4         0          no_data       9606             no_data   408170   \n",
      "...     ...              ...        ...                 ...      ...   \n",
      "4810      0          no_data       9606             no_data  no_data   \n",
      "4811      0          no_data       9606             no_data  no_data   \n",
      "4812      0          no_data       9606               MiSeq   408170   \n",
      "4813      0          no_data       9606               MiSeq   408170   \n",
      "4814      0             ROBE       9606             no_data   408170   \n",
      "\n",
      "     DIABETES_DIAGNOSE_DATE REQUIRED_SAMPLE_INFO_STATUS TYPES_OF_PLANTS  \\\n",
      "0                   no_data                     no_data    More than 30   \n",
      "1                   no_data                     no_data    More than 30   \n",
      "2                   no_data                     no_data    More than 30   \n",
      "3                   no_data                     no_data        21 to 30   \n",
      "4                   no_data                     no_data        21 to 30   \n",
      "...                     ...                         ...             ...   \n",
      "4810                unknown                   completed        11 to 20   \n",
      "4811                unknown                   completed         6 to 10   \n",
      "4812                no_data                     no_data        11 to 20   \n",
      "4813                no_data                     no_data        11 to 20   \n",
      "4814                no_data                     no_data    More than 30   \n",
      "\n",
      "     MIGRAINEMEDS FIBER_GRAMS MAINFACTOR_OTHER_3  PLATING PROJECT_NAME  \\\n",
      "0         no_data         353            no_data       GG      no_data   \n",
      "1         no_data         189            no_data  no_data      no_data   \n",
      "2         no_data         329            no_data       GG      no_data   \n",
      "3         no_data        18.3            no_data       MM      no_data   \n",
      "4         no_data         200            no_data  no_data      no_data   \n",
      "...           ...         ...                ...      ...          ...   \n",
      "4810      unknown         301            unknown  no_data      no_data   \n",
      "4811      unknown          70            unknown  no_data      no_data   \n",
      "4812           no          25            no_data  no_data      no_data   \n",
      "4813      no_data         273            no_data  no_data      no_data   \n",
      "4814      no_data         196            no_data       JG      no_data   \n",
      "\n",
      "     COLLECTION_TIMESTAMP NAILS RUN_PREFIX DESCRIPTION_WELL PLATFORM  \\\n",
      "0                 no_data    no    no_data          no_data  no_data   \n",
      "1                 no_data    no    no_data          no_data  no_data   \n",
      "2                 no_data    no    no_data          no_data  no_data   \n",
      "3                 no_data    no    no_data          no_data  no_data   \n",
      "4                 no_data   yes    no_data          no_data  no_data   \n",
      "...                   ...   ...        ...              ...      ...   \n",
      "4810       29/01/14 15:30    no    no_data          no_data  no_data   \n",
      "4811       20/08/13 08:40    no    no_data          no_data  no_data   \n",
      "4812              no_data   yes    no_data          no_data  no_data   \n",
      "4813              no_data   yes    no_data          no_data  no_data   \n",
      "4814              no_data    no    no_data          no_data  no_data   \n",
      "\n",
      "                           RACE                    PETS MULTIVITAMIN  \\\n",
      "0                     Caucasian  Contact..., Housing...          yes   \n",
      "1                     Caucasian                 no_data           no   \n",
      "2                     Caucasian  Contact..., Housing...          yes   \n",
      "3                     Caucasian  Contact..., Housing...          yes   \n",
      "4              African American                 no_data           no   \n",
      "...                         ...                     ...          ...   \n",
      "4810                  Caucasian  Contact..., Housing...      unknown   \n",
      "4811                  Caucasian  Contact..., Housing...          yes   \n",
      "4812                  Caucasian  Contact..., Housing...           no   \n",
      "4813                  Caucasian                     Cat           no   \n",
      "4814  Asian or Pacific Islander  Contact..., Housing...          yes   \n",
      "\n",
      "     EXTRACTION_ROBOT NONFOODALLERGIES_POISONIVY ACNE_MEDICATION PRIMER_PLATE  \\\n",
      "0                HOWE                    no_data             yes           10   \n",
      "1             no_data                    no_data              no            8   \n",
      "2                HOWE                    no_data              no           10   \n",
      "3                NEWE                    no_data              no            9   \n",
      "4             no_data                        yes              no            9   \n",
      "...               ...                        ...             ...          ...   \n",
      "4810          no_data                    unknown              no      no_data   \n",
      "4811          no_data                    unknown              no      no_data   \n",
      "4812          no_data                    no_data              no      no_data   \n",
      "4813          no_data                    no_data              no      no_data   \n",
      "4814             NEWE                    no_data              no           11   \n",
      "\n",
      "     ACNE_MEDICATION_OTC ANIMAL_PER  MACHINE TM300_8_TOOL  \\\n",
      "0                     no          0  no_data      111075B   \n",
      "1                     no         33   MiSeq1      no_data   \n",
      "2                     no         74  no_data      111075B   \n",
      "3                     no          1  MiSeq 1      311318B   \n",
      "4                     no         66   MiSeq1      no_data   \n",
      "...                  ...        ...      ...          ...   \n",
      "4810                  no    unknown  no_data      no_data   \n",
      "4811                  no         70  no_data      no_data   \n",
      "4812                  no         90   MiSeq1      no_data   \n",
      "4813                  no         10  no_data      no_data   \n",
      "4814                  no         50  no_data      311313B   \n",
      "\n",
      "                     MAINFACTOR_OTHER_1 EXPERIMENT_DESIGN_DESCRIPTION  \\\n",
      "0                               no_data                       no_data   \n",
      "1                               no_data                       no_data   \n",
      "2                               no_data                       no_data   \n",
      "3                               no_data                       no_data   \n",
      "4                               no_data                       no_data   \n",
      "...                                 ...                           ...   \n",
      "4810                            unknown                       no_data   \n",
      "4811                            unknown                       no_data   \n",
      "4812  not sure, may be excess serotonin                       no_data   \n",
      "4813                            no_data                       no_data   \n",
      "4814                            no_data                       no_data   \n",
      "\n",
      "     FOODALLERGIES_OTHER                  WEIGHT_CHANGE PRIMER_DATE  \\\n",
      "0                no_data                Remained stable      101013   \n",
      "1                no_data                Remained stable     no_data   \n",
      "2                no_data                Remained stable      101013   \n",
      "3                no_data                Remained stable       52813   \n",
      "4                    yes  Decreased more than 10 pounds     no_data   \n",
      "...                  ...                            ...         ...   \n",
      "4810             unknown                Remained stable     no_data   \n",
      "4811             unknown                Remained stable     no_data   \n",
      "4812             no_data                Remained stable     no_data   \n",
      "4813             no_data                Remained stable     no_data   \n",
      "4814             no_data                Remained stable       22114   \n",
      "\n",
      "             PRIMARY_CARB                DIABETES MAINFACTOR_OTHER_2 ASTHMA  \\\n",
      "0                  banana  I do not have diabetes            no_data     no   \n",
      "1     corn chips or salsa  I do not have diabetes            no_data     no   \n",
      "2                   fruit  I do not have diabetes            no_data     no   \n",
      "3                   Bread  I do not have diabetes            no_data     no   \n",
      "4                   bread  I do not have diabetes            no_data    yes   \n",
      "...                   ...                     ...                ...    ...   \n",
      "4810               Apples  I do not have diabetes            unknown     no   \n",
      "4811                bread  I do not have diabetes            unknown     no   \n",
      "4812   orange juice, milk  I do not have diabetes            no_data     no   \n",
      "4813                 Rice  I do not have diabetes            no_data     no   \n",
      "4814                 Rice  I do not have diabetes            no_data     no   \n",
      "\n",
      "                     POOL_FREQUENCY LIBRARY_CONSTRUCTION_PROTOCOL      PKU  \\\n",
      "0     Occasionally (1-2 times/week)                       no_data       no   \n",
      "1          Rarely (few times/month)                       no_data       no   \n",
      "2                             Daily                       no_data       no   \n",
      "3                             Never                       no_data       no   \n",
      "4     Occasionally (1-2 times/week)                       no_data       no   \n",
      "...                             ...                           ...      ...   \n",
      "4810                          Never                       no_data       no   \n",
      "4811  Occasionally (1-2 times/week)                       no_data  unknown   \n",
      "4812       Rarely (few times/month)                       no_data       no   \n",
      "4813       Rarely (few times/month)                       no_data       no   \n",
      "4814       Rarely (few times/month)                       no_data       no   \n",
      "\n",
      "     DECEASED_PARENT                          SEQ_RUN  CAT LATITUDE  \\\n",
      "0            no_data                          no_data  yes     39.7   \n",
      "1            no_data                          no_data  yes     32.7   \n",
      "2            no_data                          no_data   no       40   \n",
      "3            no_data                          no_data   no     40.7   \n",
      "4            no_data                          no_data   no     43.1   \n",
      "...              ...                              ...  ...      ...   \n",
      "4810         unknown                          no_data  yes       40   \n",
      "4811         unknown                          no_data   no     50.8   \n",
      "4812         no_data                          no_data  yes     47.7   \n",
      "4813         no_data  Knight_American_Gut_Freeze_Thaw  yes       42   \n",
      "4814         no_data                          no_data   no     13.7   \n",
      "\n",
      "     FOODALLERGIES_PEANUTS CHICKENPOX SEASONAL_ALLERGIES  \\\n",
      "0                  no_data        yes                 no   \n",
      "1                  no_data        yes                 no   \n",
      "2                  no_data        yes                 no   \n",
      "3                  no_data        yes                 no   \n",
      "4                  no_data        yes                yes   \n",
      "...                    ...        ...                ...   \n",
      "4810               unknown        yes                 no   \n",
      "4811               unknown         no                 no   \n",
      "4812               no_data        yes                 no   \n",
      "4813               no_data        yes                 no   \n",
      "4814               no_data        yes                yes   \n",
      "\n",
      "                            Description  \n",
      "0     American Gut Project Stool sample  \n",
      "1     American Gut Project Stool sample  \n",
      "2     American Gut Project Stool sample  \n",
      "3     American Gut Project Stool sample  \n",
      "4     American Gut Project Stool sample  \n",
      "...                                 ...  \n",
      "4810                               None  \n",
      "4811                        Description  \n",
      "4812  American Gut Project Stool sample  \n",
      "4813  American Gut Project Stool sample  \n",
      "4814  American Gut Project Stool sample  \n",
      "\n",
      "[4815 rows x 202 columns]\n"
     ]
    }
   ],
   "source": [
    "# Show the top and bottom 5 rows of all columns. \n",
    "\n",
    "with pd.option_context('display.max_columns', None):\n",
    "    print (df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Columns to delete:** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>delete_in_step_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>#SampleID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BarcodeSequence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LinkerPrimerSequence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASSIGNED_FROM_GEO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RUN_DATE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       delete_in_step_1\n",
       "0             #SampleID\n",
       "1       BarcodeSequence\n",
       "2  LinkerPrimerSequence\n",
       "3     ASSIGNED_FROM_GEO\n",
       "4              RUN_DATE"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import csv with the names of the columns to be deleted in deletion round 1.\n",
    "\n",
    "df_delete_1 = pd.read_csv(\"columns_delete_in_step_1.csv\")\n",
    "df_delete_1.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['#SampleID',\n",
       " 'BarcodeSequence',\n",
       " 'LinkerPrimerSequence',\n",
       " 'ASSIGNED_FROM_GEO',\n",
       " 'RUN_DATE',\n",
       " 'PET_LOCATIONS',\n",
       " 'GENERAL_MEDS',\n",
       " 'SAMPLE_ID',\n",
       " 'ORIG_SAMPLE_NAME',\n",
       " 'PCR_PRIMERS',\n",
       " 'MASTERMIX_LOT',\n",
       " 'HOST_SUBJECT_ID',\n",
       " 'ENA-CHECKLIST',\n",
       " 'SAMPLE_NAME',\n",
       " 'TM1000_8_TOOL',\n",
       " 'ORIG_NAME',\n",
       " 'FRAT',\n",
       " 'SAMPLE_PLATE',\n",
       " 'FOODALLERGIES_OTHER_TEXT',\n",
       " 'KEY_SEQ',\n",
       " 'BODY_PRODUCT',\n",
       " 'REGION',\n",
       " 'SAMPLE_TYPE',\n",
       " 'EXPERIMENT_TITLE',\n",
       " 'EXTRACTIONKIT_LOT',\n",
       " 'SAMPLE',\n",
       " 'CENTER_NAME',\n",
       " 'COMMON_NAME',\n",
       " 'SEQUENCING_METH',\n",
       " 'ENV_BIOME',\n",
       " 'WELL_ID',\n",
       " 'SAMPLE_CENTER',\n",
       " 'PHYSICAL_LOCATION',\n",
       " 'LINKER',\n",
       " 'TANNING_SPRAYS',\n",
       " 'HOST_COMMON_NAME',\n",
       " 'RACE_OTHER',\n",
       " 'ALTITUDE',\n",
       " 'NONFOODALLERGIES_DANDER',\n",
       " 'TARGET_SUBFRAGMENT',\n",
       " 'TITLE',\n",
       " 'MIGRAINE_PAIN',\n",
       " 'ROOMMATES',\n",
       " 'CENTER_PROJECT_NAME',\n",
       " 'HAS_EXTRACTED_DATA',\n",
       " 'ENA-SPOT-COUNT',\n",
       " 'RUN_ID',\n",
       " 'COLLECTION_DATE',\n",
       " 'TARGET_GENE',\n",
       " 'PREGNANT_DUE_DATE',\n",
       " 'SAMPLE_DESCRIPTION',\n",
       " 'SAMP_SIZE',\n",
       " 'HAS_PHYSICAL_SPECIMEN',\n",
       " 'STUDY_CENTER',\n",
       " 'WATER_LOT',\n",
       " 'ENV_FEATURE',\n",
       " 'ENA-BASE-COUNT',\n",
       " 'SAMPLE_TIME',\n",
       " 'EMP_STATUS',\n",
       " 'COMMUNAL_DINING',\n",
       " 'ANONYMIZED_NAME',\n",
       " 'PUBLIC',\n",
       " 'RUN_CENTER',\n",
       " 'EXPERIMENT_CENTER',\n",
       " 'TM50_8_TOOL',\n",
       " 'DEPTH',\n",
       " 'PROCESSING_ROBOT',\n",
       " 'HOST_TAXID',\n",
       " 'ILLUMINA_TECHNOLOGY',\n",
       " 'TAXON_ID',\n",
       " 'DIABETES_DIAGNOSE_DATE',\n",
       " 'REQUIRED_SAMPLE_INFO_STATUS',\n",
       " 'MAINFACTOR_OTHER_3',\n",
       " 'PLATING',\n",
       " 'PROJECT_NAME',\n",
       " 'COLLECTION_TIMESTAMP',\n",
       " 'RUN_PREFIX',\n",
       " 'DESCRIPTION_WELL',\n",
       " 'PLATFORM',\n",
       " 'EXTRACTION_ROBOT',\n",
       " 'PRIMER_PLATE',\n",
       " 'MACHINE',\n",
       " 'TM300_8_TOOL',\n",
       " 'MAINFACTOR_OTHER_1',\n",
       " 'EXPERIMENT_DESIGN_DESCRIPTION',\n",
       " 'PRIMER_DATE',\n",
       " 'MAINFACTOR_OTHER_2',\n",
       " 'LIBRARY_CONSTRUCTION_PROTOCOL',\n",
       " 'PKU',\n",
       " 'DECEASED_PARENT',\n",
       " 'SEQ_RUN',\n",
       " 'Description']"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert dataframe into a list. \n",
    "\n",
    "df_delete_1_list = df_delete_1['delete_in_step_1'].tolist()\n",
    "df_delete_1_list\n",
    "\n",
    "# https://www.geeksforgeeks.org/python-pandas-index-tolist/#:~:text=tolist()%20function%20return%20a,Timedelta%2FInterval%2FPeriod)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "92"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_delete_1_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Delete columns shown in 'df_delete_1_list'. \n",
    "\n",
    "df_2 = df.drop(df_delete_1_list, axis = 1) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4815, 110)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get shape of updated data frame df_2 after deleting 92 columns.\n",
    "\n",
    "df_2.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "df_2 has 4815 rows and 110 columns. This means that 45.5 % of the columns have been deleted at this step. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4815, 202)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking the shape of the original data frame df.\n",
    "\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CARBOHYDRATE_PER',\n",
       " 'NONFOODALLERGIES_DRUG',\n",
       " 'NONFOODALLERGIES_BEESTINGS',\n",
       " 'AGE',\n",
       " 'TOT_MASS',\n",
       " 'QUINOLINE',\n",
       " 'BODY_SITE',\n",
       " 'MIGRAINE_FACTOR_2',\n",
       " 'MIGRAINE_FACTOR_3',\n",
       " 'TONSILS_REMOVED',\n",
       " 'MIGRAINE_FACTOR_1',\n",
       " 'ELEVATION',\n",
       " 'FLOSSING_FREQUENCY',\n",
       " 'PROTEIN_PER',\n",
       " 'CSECTION',\n",
       " 'COUNTRY_OF_BIRTH',\n",
       " 'MACRONUTRIENT_PCT_TOTAL',\n",
       " 'COSMETICS_FREQUENCY',\n",
       " 'GLUTEN',\n",
       " 'LONGITUDE',\n",
       " 'DOG',\n",
       " 'MIGRAINE_AGGRAVATION',\n",
       " 'DIABETES_MEDICATION',\n",
       " 'PENICILLIN',\n",
       " 'MIGRAINE_FREQUENCY',\n",
       " 'MIGRAINE_RELATIVES',\n",
       " 'PLANT_PER',\n",
       " 'MIGRAINE_PHOTOPHOBIA',\n",
       " 'SITE_SAMPLED',\n",
       " 'DRINKING_WATER_SOURCE',\n",
       " 'COUNTRY',\n",
       " 'ANTIBIOTIC_SELECT',\n",
       " 'MIGRAINE_NAUSEA',\n",
       " 'FOODALLERGIES_TREENUTS',\n",
       " 'MIGRAINE_AURA',\n",
       " 'FLU_VACCINE_DATE',\n",
       " 'PREGNANT',\n",
       " 'MIGRAINE',\n",
       " 'BMI',\n",
       " 'TEETHBRUSHING_FREQUENCY',\n",
       " 'MIGRAINE_MEDICATIONS',\n",
       " 'CONTRACEPTIVE',\n",
       " 'LIVINGWITH',\n",
       " 'RELATIONS',\n",
       " 'DIABETES_MEDICATIONS',\n",
       " 'SUPPLEMENTS',\n",
       " 'AGE_UNIT',\n",
       " 'EXERCISE_FREQUENCY',\n",
       " 'SULFA_DRUG',\n",
       " 'NONFOODALLERGIES_NO',\n",
       " 'ENV_MATTER',\n",
       " 'SKIN_CONDITION',\n",
       " 'CONDITIONS_MEDICATION',\n",
       " 'LAST_TRAVEL',\n",
       " 'HEIGHT_IN',\n",
       " 'ANTIBIOTIC_CONDITION',\n",
       " 'PET_CONTACT',\n",
       " 'NONFOODALLERGIES_SUN',\n",
       " 'EXERCISE_LOCATION',\n",
       " 'NITROIMIDAZOLE',\n",
       " 'DEODORANT_USE',\n",
       " 'FOODALLERGIES_SHELLFISH',\n",
       " 'SEX',\n",
       " 'PERCENTAGE_FROM_CARBS',\n",
       " 'FAT_PER',\n",
       " 'SHARED_HOUSING',\n",
       " 'SOFTENER',\n",
       " 'MIGRAINE_PHONOPHOBIA',\n",
       " 'SMOKING_FREQUENCY',\n",
       " 'WEIGHT_LBS',\n",
       " 'TANNING_BEDS',\n",
       " 'ALCOHOL_FREQUENCY',\n",
       " 'STATE',\n",
       " 'BODY_HABITAT',\n",
       " 'DOMINANT_HAND',\n",
       " 'PRIMARY_VEGETABLE',\n",
       " 'SPECIAL_RESTRICTIONS',\n",
       " 'CEPHALOSPORIN',\n",
       " 'SUPPLEMENTS_FIELDS',\n",
       " 'APPENDIX_REMOVED',\n",
       " 'CURRENT_RESIDENCE_DURATION',\n",
       " 'SLEEP_DURATION',\n",
       " 'DIET_TYPE',\n",
       " 'DIET_RESTRICTIONS',\n",
       " 'LACTOSE',\n",
       " 'HEIGHT_OR_LENGTH',\n",
       " 'ANTIBIOTIC_MEDS',\n",
       " 'IBD',\n",
       " 'TYPES_OF_PLANTS',\n",
       " 'MIGRAINEMEDS',\n",
       " 'FIBER_GRAMS',\n",
       " 'NAILS',\n",
       " 'RACE',\n",
       " 'PETS',\n",
       " 'MULTIVITAMIN',\n",
       " 'NONFOODALLERGIES_POISONIVY',\n",
       " 'ACNE_MEDICATION',\n",
       " 'ACNE_MEDICATION_OTC',\n",
       " 'ANIMAL_PER',\n",
       " 'FOODALLERGIES_OTHER',\n",
       " 'WEIGHT_CHANGE',\n",
       " 'PRIMARY_CARB',\n",
       " 'DIABETES',\n",
       " 'ASTHMA',\n",
       " 'POOL_FREQUENCY',\n",
       " 'CAT',\n",
       " 'LATITUDE',\n",
       " 'FOODALLERGIES_PEANUTS',\n",
       " 'CHICKENPOX',\n",
       " 'SEASONAL_ALLERGIES']"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the column names of df_2, the new smaller data frame.\n",
    "   \n",
    "df_2_list = list(df_2.columns.values) \n",
    "df_2_list\n",
    "\n",
    "# https://www.geeksforgeeks.org/how-to-get-column-names-in-pandas-dataframe/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS AGE  \\\n",
      "0                  55               no_data                    no_data  51   \n",
      "1                  56               no_data                    no_data  35   \n",
      "2                  16               no_data                    no_data  56   \n",
      "3             no_data               no_data                    no_data  37   \n",
      "4                  61                   yes                    no_data  72   \n",
      "...               ...                   ...                        ...  ..   \n",
      "4810               50               unknown                    unknown  71   \n",
      "4811               17               unknown                    unknown   9   \n",
      "4812               34               no_data                    no_data  29   \n",
      "4813               30               no_data                    no_data  46   \n",
      "4814               60               no_data                    no_data   1   \n",
      "\n",
      "     TOT_MASS QUINOLINE      BODY_SITE MIGRAINE_FACTOR_2 MIGRAINE_FACTOR_3  \\\n",
      "0          50        no   UBERON:feces           no_data           no_data   \n",
      "1          63   no_data   UBERON:feces           no_data           no_data   \n",
      "2          64        no   UBERON:feces           no_data           no_data   \n",
      "3          90        no   UBERON:feces           no_data           no_data   \n",
      "4          86   no_data   UBERON:feces           no_data           no_data   \n",
      "...       ...       ...            ...               ...               ...   \n",
      "4810       80       yes   UBERON:feces           unknown           unknown   \n",
      "4811       18        no  UBERON:tongue           unknown           unknown   \n",
      "4812       79        no   UBERON:feces            Stress           no_data   \n",
      "4813       83        no   UBERON:feces           no_data           no_data   \n",
      "4814       73        no   UBERON:feces           no_data           no_data   \n",
      "\n",
      "     TONSILS_REMOVED MIGRAINE_FACTOR_1 ELEVATION          FLOSSING_FREQUENCY  \\\n",
      "0            no_data           no_data    1654.9                       Daily   \n",
      "1                 no           no_data   no_data  Regularly (3-5 times/week)   \n",
      "2                yes           no_data    1624.1                       Daily   \n",
      "3                 no           no_data   no_data                       Never   \n",
      "4                 no           no_data   no_data                       Daily   \n",
      "...              ...               ...       ...                         ...   \n",
      "4810             yes           unknown       223                       Daily   \n",
      "4811              no           unknown     101.9                       Never   \n",
      "4812              no             Other      61.8                       Never   \n",
      "4813              no           no_data     181.7    Rarely (few times/month)   \n",
      "4814              no           no_data       4.4                       Daily   \n",
      "\n",
      "     PROTEIN_PER  CSECTION COUNTRY_OF_BIRTH MACRONUTRIENT_PCT_TOTAL  \\\n",
      "0             15        no              USA                     100   \n",
      "1             22        no              USA                      99   \n",
      "2             17        no              USA                     100   \n",
      "3        no_data       yes              USA                      80   \n",
      "4             16        no              USA                     100   \n",
      "...          ...       ...              ...                     ...   \n",
      "4810          29        no              USA                     100   \n",
      "4811          47       yes          Belgium                     100   \n",
      "4812          18        no              USA                     100   \n",
      "4813          40  not_sure        Australia                     100   \n",
      "4814          16        no         Thailand                     100   \n",
      "\n",
      "           COSMETICS_FREQUENCY GLUTEN LONGITUDE      DOG MIGRAINE_AGGRAVATION  \\\n",
      "0                        Daily     no    -104.9  no_data              no_data   \n",
      "1     Rarely (few times/month)     no    -117.2       no              no_data   \n",
      "2                        Never    yes    -105.3       no              no_data   \n",
      "3                        Never     no       -74       no              no_data   \n",
      "4                        Never     no     -74.3       no              no_data   \n",
      "...                        ...    ...       ...      ...                  ...   \n",
      "4810                     Never     no     -88.3       no              unknown   \n",
      "4811                     Never     no       4.4       no              unknown   \n",
      "4812                     Never    yes    -122.3       no              no_data   \n",
      "4813                     Never     no     -87.7       no              no_data   \n",
      "4814                     Never     no     100.5      yes              no_data   \n",
      "\n",
      "     DIABETES_MEDICATION PENICILLIN MIGRAINE_FREQUENCY MIGRAINE_RELATIVES  \\\n",
      "0                no_data         no            no_data            no_data   \n",
      "1                no_data    no_data            no_data            no_data   \n",
      "2                no_data         no            no_data            no_data   \n",
      "3                no_data         no            no_data            no_data   \n",
      "4                no_data    no_data            no_data            no_data   \n",
      "...                  ...        ...                ...                ...   \n",
      "4810             unknown         no            unknown            unknown   \n",
      "4811             unknown        yes            unknown            unknown   \n",
      "4812             no_data         no            Monthly                yes   \n",
      "4813             no_data         no            no_data            no_data   \n",
      "4814             no_data         no            no_data            no_data   \n",
      "\n",
      "     PLANT_PER MIGRAINE_PHOTOPHOBIA SITE_SAMPLED DRINKING_WATER_SOURCE  \\\n",
      "0          100              no_data        Stool                  City   \n",
      "1           67              no_data        Stool              Filtered   \n",
      "2           26              no_data        Stool                  City   \n",
      "3           99              no_data        Stool                  City   \n",
      "4           34              no_data        Stool                  City   \n",
      "...        ...                  ...          ...                   ...   \n",
      "4810   unknown              unknown      no_data                  City   \n",
      "4811        30              unknown      no_data              Filtered   \n",
      "4812        10              no_data        Stool              Filtered   \n",
      "4813        90              no_data        Stool                  City   \n",
      "4814        50              no_data        Stool              Filtered   \n",
      "\n",
      "                           COUNTRY     ANTIBIOTIC_SELECT MIGRAINE_NAUSEA  \\\n",
      "0     GAZ:United States of America  Not in the last year         no_data   \n",
      "1     GAZ:United States of America  Not in the last year         no_data   \n",
      "2     GAZ:United States of America  Not in the last year         no_data   \n",
      "3     GAZ:United States of America  Not in the last year         no_data   \n",
      "4     GAZ:United States of America  In the past 6 months         no_data   \n",
      "...                            ...                   ...             ...   \n",
      "4810  GAZ:United States of America  In the past 6 months         unknown   \n",
      "4811                   GAZ:Belgium      In the past year         unknown   \n",
      "4812  GAZ:United States of America  Not in the last year             yes   \n",
      "4813  GAZ:United States of America  Not in the last year         no_data   \n",
      "4814                  GAZ:Thailand  Not in the last year         no_data   \n",
      "\n",
      "     FOODALLERGIES_TREENUTS MIGRAINE_AURA      FLU_VACCINE_DATE PREGNANT  \\\n",
      "0                   no_data       no_data  Not in the last year       no   \n",
      "1                   no_data       no_data  In the past 6 months  unknown   \n",
      "2                       yes       no_data  Not in the last year       no   \n",
      "3                   no_data       no_data      In the past year       no   \n",
      "4                   no_data       no_data      In the past year  unknown   \n",
      "...                     ...           ...                   ...      ...   \n",
      "4810                unknown       unknown  In the past 6 months       no   \n",
      "4811                unknown       unknown  Not in the last year       no   \n",
      "4812                no_data           yes  Not in the last year       no   \n",
      "4813                no_data       no_data      In the past year       no   \n",
      "4814                no_data       no_data  In the past 6 months       no   \n",
      "\n",
      "     MIGRAINE          BMI TEETHBRUSHING_FREQUENCY MIGRAINE_MEDICATIONS  \\\n",
      "0          no  18.87939453                   Daily              no_data   \n",
      "1          no  24.79717813                   Daily              no_data   \n",
      "2          no  21.74070069                   Daily              no_data   \n",
      "3          no  26.38393695                   Daily              no_data   \n",
      "4          no  32.60986328                   Daily              no_data   \n",
      "...       ...          ...                     ...                  ...   \n",
      "4810       no  25.53755102                   Daily              unknown   \n",
      "4811       no  12.72974196                   Daily              unknown   \n",
      "4812      yes  21.87111111                   Daily              no_data   \n",
      "4813       no  26.54183673                   Daily              no_data   \n",
      "4814       no  24.47729239                   Daily              no_data   \n",
      "\n",
      "     CONTRACEPTIVE LIVINGWITH               RELATIONS DIABETES_MEDICATIONS  \\\n",
      "0               No         no                 no_data              no_data   \n",
      "1               No         no                 no_data              no_data   \n",
      "2          no_data         no                 no_data              no_data   \n",
      "3          no_data        yes              Girlfriend              no_data   \n",
      "4               No        yes                daughter              no_data   \n",
      "...            ...        ...                     ...                  ...   \n",
      "4810       unknown         no                 unknown              unknown   \n",
      "4811       unknown        yes  Father, Mother, Sister              unknown   \n",
      "4812       no_data         no                 no_data              no_data   \n",
      "4813       no_data        yes                    Wife              no_data   \n",
      "4814       no_data         no                 no_data              no_data   \n",
      "\n",
      "     SUPPLEMENTS AGE_UNIT             EXERCISE_FREQUENCY SULFA_DRUG  \\\n",
      "0            yes    years                          Daily         no   \n",
      "1            yes    years     Regularly (3-5 times/week)    no_data   \n",
      "2            yes    years                          Daily         no   \n",
      "3            yes    years                          Never         no   \n",
      "4            yes    years                          Daily    no_data   \n",
      "...          ...      ...                            ...        ...   \n",
      "4810         yes    years                          Daily         no   \n",
      "4811          no    years  Occasionally (1-2 times/week)         no   \n",
      "4812         yes    years                          Daily         no   \n",
      "4813         yes    years                          Daily         no   \n",
      "4814          no    years                          Daily         no   \n",
      "\n",
      "     NONFOODALLERGIES_NO   ENV_MATTER                  SKIN_CONDITION  \\\n",
      "0                no_data   ENVO:feces  I do not have a skin condition   \n",
      "1                    yes   ENVO:feces  I do not have a skin condition   \n",
      "2                no_data   ENVO:feces                         Rosacea   \n",
      "3                no_data   ENVO:feces  I do not have a skin condition   \n",
      "4                no_data   ENVO:feces  I do not have a skin condition   \n",
      "...                  ...          ...                             ...   \n",
      "4810             unknown   ENVO:feces  I do not have a skin condition   \n",
      "4811                 yes  ENVO:saliva  I do not have a skin condition   \n",
      "4812             no_data   ENVO:feces                          Eczema   \n",
      "4813             no_data   ENVO:feces  I do not have a skin condition   \n",
      "4814                 yes   ENVO:feces  I do not have a skin condition   \n",
      "\n",
      "     CONDITIONS_MEDICATION               LAST_TRAVEL HEIGHT_IN  \\\n",
      "0                       no    More than one year ago        64   \n",
      "1                       no      Within the past year        63   \n",
      "2                      yes    More than one year ago        68   \n",
      "3                       no  Within the past 3 months        73   \n",
      "4                       no    More than one year ago        64   \n",
      "...                    ...                       ...       ...   \n",
      "4810                    no    More than one year ago        70   \n",
      "4811                    no                   unknown        47   \n",
      "4812                    no  Within the past 6 months        75   \n",
      "4813                    no      Within the past year        70   \n",
      "4814                   yes  Within the past 3 months        68   \n",
      "\n",
      "     ANTIBIOTIC_CONDITION PET_CONTACT NONFOODALLERGIES_SUN EXERCISE_LOCATION  \\\n",
      "0                 no_data  Contact...              no_data              Both   \n",
      "1                 no_data     no_data              no_data              Both   \n",
      "2                 no_data  Contact...              no_data          Outdoors   \n",
      "3                 no_data  Contact...              no_data           no_data   \n",
      "4         dental cleaning     no_data              no_data          Outdoors   \n",
      "...                   ...         ...                  ...               ...   \n",
      "4810                  UTI  Contact...              unknown           Indoors   \n",
      "4811      Streptococcus A  Contact...              unknown           Indoors   \n",
      "4812              no_data  Contact...              no_data              Both   \n",
      "4813              no_data     no_data              no_data              Both   \n",
      "4814              no_data  Contact...              no_data          Outdoors   \n",
      "\n",
      "     NITROIMIDAZOLE                                      DEODORANT_USE  \\\n",
      "0                no                                    I use deoderant   \n",
      "1           no_data  Not sure, but I use some form of deoderant/ant...   \n",
      "2                no                                    I use deoderant   \n",
      "3                no                                    I use deoderant   \n",
      "4           no_data                                    I use deoderant   \n",
      "...             ...                                                ...   \n",
      "4810             no           I do not use deoderant or antiperspirant   \n",
      "4811             no           I do not use deoderant or antiperspirant   \n",
      "4812             no           I do not use deodorant or antiperspirant   \n",
      "4813             no           I do not use deoderant or antiperspirant   \n",
      "4814             no                                    I use deoderant   \n",
      "\n",
      "     FOODALLERGIES_SHELLFISH      SEX PERCENTAGE_FROM_CARBS  FAT_PER  \\\n",
      "0                    no_data   female         Less than 25%       30   \n",
      "1                    no_data   female         Less than 25%       21   \n",
      "2                    no_data     male         Less than 25%       67   \n",
      "3                    no_data     male             51 to 75%  no_data   \n",
      "4                    no_data   female             25 to 50%       23   \n",
      "...                      ...      ...                   ...      ...   \n",
      "4810                 unknown     male         Less than 25%       21   \n",
      "4811                 unknown  unknown             51 to 75%       36   \n",
      "4812                 no_data     male         Less than 25%       48   \n",
      "4813                 no_data     male         Less than 25%       30   \n",
      "4814                 no_data     male         Less than 25%       24   \n",
      "\n",
      "     SHARED_HOUSING SOFTENER MIGRAINE_PHONOPHOBIA SMOKING_FREQUENCY  \\\n",
      "0                no      yes              no_data             Never   \n",
      "1                no      yes              no_data             Never   \n",
      "2                no       no              no_data             Never   \n",
      "3                no       no              no_data           no_data   \n",
      "4               yes       no              no_data             Never   \n",
      "...             ...      ...                  ...               ...   \n",
      "4810             no       no              unknown             Never   \n",
      "4811             no       no              unknown             Never   \n",
      "4812             no       no              no_data             Never   \n",
      "4813             no      yes              no_data             Never   \n",
      "4814             no       no              no_data             Never   \n",
      "\n",
      "     WEIGHT_LBS TANNING_BEDS              ALCOHOL_FREQUENCY    STATE  \\\n",
      "0           110      no_data                          Never       CO   \n",
      "1           140      no_data                          Daily       CA   \n",
      "2           143      no_data                          Never       CO   \n",
      "3           200      no_data                          Daily       NY   \n",
      "4           190      no_data                          Never       NY   \n",
      "...         ...          ...                            ...      ...   \n",
      "4810        178      unknown                          Never       IL   \n",
      "4811         40      unknown                          Never  unknown   \n",
      "4812        175      no_data       Rarely (few times/month)       WA   \n",
      "4813        185      no_data  Occasionally (1-2 times/week)       IL   \n",
      "4814        161      no_data       Rarely (few times/month)   BANKOK   \n",
      "\n",
      "            BODY_HABITAT DOMINANT_HAND                PRIMARY_VEGETABLE  \\\n",
      "0           UBERON:feces          left                          spinach   \n",
      "1           UBERON:feces          left  tomatoes (I eat a lot of salsa)   \n",
      "2           UBERON:feces         right                     leafy greens   \n",
      "3           UBERON:feces         right                           Potato   \n",
      "4           UBERON:feces         right                         broccoli   \n",
      "...                  ...           ...                              ...   \n",
      "4810        UBERON:feces          left                  Romaine Lettuce   \n",
      "4811  UBERON:oral cavity         right              tomatoes, cuncumber   \n",
      "4812        UBERON:feces         right                           potato   \n",
      "4813        UBERON:feces         right                     Leafy Greens   \n",
      "4814        UBERON:feces         right                Leaves and fruits   \n",
      "\n",
      "     SPECIAL_RESTRICTIONS CEPHALOSPORIN  \\\n",
      "0                      no            no   \n",
      "1                      no       no_data   \n",
      "2                     yes            no   \n",
      "3                      no            no   \n",
      "4                      no       no_data   \n",
      "...                   ...           ...   \n",
      "4810                   no            no   \n",
      "4811                   no            no   \n",
      "4812                  yes            no   \n",
      "4813                   no            no   \n",
      "4814                   no            no   \n",
      "\n",
      "                                     SUPPLEMENTS_FIELDS APPENDIX_REMOVED  \\\n",
      "0                             calcium, probiotics, maca               no   \n",
      "1                vitamin B complex, Omega 3 fatty acids               no   \n",
      "2     Kombucha, Probiotic supplements, Niacin, CoEnz...               no   \n",
      "3                                        pro/prebiotics               no   \n",
      "4                                      calcium, vitamin               no   \n",
      "...                                                 ...              ...   \n",
      "4810  Cranberry 500 mg, Probiotic Caps, Lactobacillu...               no   \n",
      "4811                                            unknown               no   \n",
      "4812                                           b6, zinc               no   \n",
      "4813                     Fish Oil, Zyflamend Whole Body               no   \n",
      "4814                                            no_data               no   \n",
      "\n",
      "     CURRENT_RESIDENCE_DURATION     SLEEP_DURATION  \\\n",
      "0        More than one year ago          6-7 hours   \n",
      "1        More than one year ago  Less than 5 hours   \n",
      "2        More than one year ago    8 or more hours   \n",
      "3        More than one year ago          6-7 hours   \n",
      "4        More than one year ago          5-6 hours   \n",
      "...                         ...                ...   \n",
      "4810     More than one year ago          6-7 hours   \n",
      "4811                    unknown    8 or more hours   \n",
      "4812     More than one year ago          7-8 hours   \n",
      "4813     More than one year ago          6-7 hours   \n",
      "4814     More than one year ago          6-7 hours   \n",
      "\n",
      "                       DIET_TYPE  \\\n",
      "0                          Vegan   \n",
      "1                       Omnivore   \n",
      "2                       Omnivore   \n",
      "3     Vegetarian but eat seafood   \n",
      "4       Omnivore but no red meat   \n",
      "...                          ...   \n",
      "4810                    Omnivore   \n",
      "4811                    Omnivore   \n",
      "4812                    Omnivore   \n",
      "4813                    Omnivore   \n",
      "4814                    Omnivore   \n",
      "\n",
      "                                      DIET_RESTRICTIONS LACTOSE  \\\n",
      "0                                               no_data      no   \n",
      "1                                               no_data      no   \n",
      "2                                            Grain free     yes   \n",
      "3                                               no_data      no   \n",
      "4                                               no_data      no   \n",
      "...                                                 ...     ...   \n",
      "4810                                            unknown      no   \n",
      "4811                                            unknown      no   \n",
      "4812  Ray Peat diet. no grains, beans, minimal veget...      no   \n",
      "4813                                            no_data      no   \n",
      "4814                                            no_data      no   \n",
      "\n",
      "     HEIGHT_OR_LENGTH ANTIBIOTIC_MEDS                IBD TYPES_OF_PLANTS  \\\n",
      "0                 163         no_data  I do not have IBD    More than 30   \n",
      "1                 160         no_data  I do not have IBD    More than 30   \n",
      "2                 173         no_data  I do not have IBD    More than 30   \n",
      "3                 186         no_data  I do not have IBD        21 to 30   \n",
      "4                 163         no_data  I do not have IBD        21 to 30   \n",
      "...               ...             ...                ...             ...   \n",
      "4810              178           CIPRO  I do not have IBD        11 to 20   \n",
      "4811              120     amoxicillin  I do not have IBD         6 to 10   \n",
      "4812              191         no_data  I do not have IBD        11 to 20   \n",
      "4813              178         no_data  I do not have IBD        11 to 20   \n",
      "4814              175         no_data  I do not have IBD    More than 30   \n",
      "\n",
      "     MIGRAINEMEDS FIBER_GRAMS NAILS                       RACE  \\\n",
      "0         no_data         353    no                  Caucasian   \n",
      "1         no_data         189    no                  Caucasian   \n",
      "2         no_data         329    no                  Caucasian   \n",
      "3         no_data        18.3    no                  Caucasian   \n",
      "4         no_data         200   yes           African American   \n",
      "...           ...         ...   ...                        ...   \n",
      "4810      unknown         301    no                  Caucasian   \n",
      "4811      unknown          70    no                  Caucasian   \n",
      "4812           no          25   yes                  Caucasian   \n",
      "4813      no_data         273   yes                  Caucasian   \n",
      "4814      no_data         196    no  Asian or Pacific Islander   \n",
      "\n",
      "                        PETS MULTIVITAMIN NONFOODALLERGIES_POISONIVY  \\\n",
      "0     Contact..., Housing...          yes                    no_data   \n",
      "1                    no_data           no                    no_data   \n",
      "2     Contact..., Housing...          yes                    no_data   \n",
      "3     Contact..., Housing...          yes                    no_data   \n",
      "4                    no_data           no                        yes   \n",
      "...                      ...          ...                        ...   \n",
      "4810  Contact..., Housing...      unknown                    unknown   \n",
      "4811  Contact..., Housing...          yes                    unknown   \n",
      "4812  Contact..., Housing...           no                    no_data   \n",
      "4813                     Cat           no                    no_data   \n",
      "4814  Contact..., Housing...          yes                    no_data   \n",
      "\n",
      "     ACNE_MEDICATION ACNE_MEDICATION_OTC ANIMAL_PER FOODALLERGIES_OTHER  \\\n",
      "0                yes                  no          0             no_data   \n",
      "1                 no                  no         33             no_data   \n",
      "2                 no                  no         74             no_data   \n",
      "3                 no                  no          1             no_data   \n",
      "4                 no                  no         66                 yes   \n",
      "...              ...                 ...        ...                 ...   \n",
      "4810              no                  no    unknown             unknown   \n",
      "4811              no                  no         70             unknown   \n",
      "4812              no                  no         90             no_data   \n",
      "4813              no                  no         10             no_data   \n",
      "4814              no                  no         50             no_data   \n",
      "\n",
      "                      WEIGHT_CHANGE         PRIMARY_CARB  \\\n",
      "0                   Remained stable               banana   \n",
      "1                   Remained stable  corn chips or salsa   \n",
      "2                   Remained stable                fruit   \n",
      "3                   Remained stable                Bread   \n",
      "4     Decreased more than 10 pounds                bread   \n",
      "...                             ...                  ...   \n",
      "4810                Remained stable               Apples   \n",
      "4811                Remained stable                bread   \n",
      "4812                Remained stable   orange juice, milk   \n",
      "4813                Remained stable                 Rice   \n",
      "4814                Remained stable                 Rice   \n",
      "\n",
      "                    DIABETES ASTHMA                 POOL_FREQUENCY  CAT  \\\n",
      "0     I do not have diabetes     no  Occasionally (1-2 times/week)  yes   \n",
      "1     I do not have diabetes     no       Rarely (few times/month)  yes   \n",
      "2     I do not have diabetes     no                          Daily   no   \n",
      "3     I do not have diabetes     no                          Never   no   \n",
      "4     I do not have diabetes    yes  Occasionally (1-2 times/week)   no   \n",
      "...                      ...    ...                            ...  ...   \n",
      "4810  I do not have diabetes     no                          Never  yes   \n",
      "4811  I do not have diabetes     no  Occasionally (1-2 times/week)   no   \n",
      "4812  I do not have diabetes     no       Rarely (few times/month)  yes   \n",
      "4813  I do not have diabetes     no       Rarely (few times/month)  yes   \n",
      "4814  I do not have diabetes     no       Rarely (few times/month)   no   \n",
      "\n",
      "     LATITUDE FOODALLERGIES_PEANUTS CHICKENPOX SEASONAL_ALLERGIES  \n",
      "0        39.7               no_data        yes                 no  \n",
      "1        32.7               no_data        yes                 no  \n",
      "2          40               no_data        yes                 no  \n",
      "3        40.7               no_data        yes                 no  \n",
      "4        43.1               no_data        yes                yes  \n",
      "...       ...                   ...        ...                ...  \n",
      "4810       40               unknown        yes                 no  \n",
      "4811     50.8               unknown         no                 no  \n",
      "4812     47.7               no_data        yes                 no  \n",
      "4813       42               no_data        yes                 no  \n",
      "4814     13.7               no_data        yes                yes  \n",
      "\n",
      "[4815 rows x 110 columns]\n"
     ]
    }
   ],
   "source": [
    "# Show the top and bottom 5 rows of all columns in df_2, the new smaller data frame.\n",
    "\n",
    "with pd.option_context('display.max_columns', None):\n",
    "    print (df_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GLUTEN                 1\n",
       "DIABETES_MEDICATION    1\n",
       "ANIMAL_PER             1\n",
       "SITE_SAMPLED           1\n",
       "ACNE_MEDICATION        1\n",
       "                      ..\n",
       "TANNING_BEDS           1\n",
       "NONFOODALLERGIES_NO    1\n",
       "SEASONAL_ALLERGIES     1\n",
       "TOT_MASS               1\n",
       "FLOSSING_FREQUENCY     1\n",
       "Length: 110, dtype: int64"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if there are any duplicate columns.\n",
    "\n",
    "df_2.columns.value_counts()\n",
    "\n",
    "# https://cmdlinetips.com/2020/08/pandas-value_counts-how-to-get-frequency-counts-of-variables-in-a-dataframe/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Length is 110, i.e each column name appears only once, hence there are no duplicate columns."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The dataset contains several rows that correspond to blank samples and not to  a person. \n",
    "These rows are removed below, creating df_3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>NONFOODALLERGIES_DRUG</th>\n",
       "      <th>NONFOODALLERGIES_BEESTINGS</th>\n",
       "      <th>AGE</th>\n",
       "      <th>TOT_MASS</th>\n",
       "      <th>QUINOLINE</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>MIGRAINE_FACTOR_2</th>\n",
       "      <th>MIGRAINE_FACTOR_3</th>\n",
       "      <th>TONSILS_REMOVED</th>\n",
       "      <th>...</th>\n",
       "      <th>WEIGHT_CHANGE</th>\n",
       "      <th>PRIMARY_CARB</th>\n",
       "      <th>DIABETES</th>\n",
       "      <th>ASTHMA</th>\n",
       "      <th>POOL_FREQUENCY</th>\n",
       "      <th>CAT</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>FOODALLERGIES_PEANUTS</th>\n",
       "      <th>CHICKENPOX</th>\n",
       "      <th>SEASONAL_ALLERGIES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>55</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>51</td>\n",
       "      <td>50</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>banana</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Occasionally (1-2 times/week)</td>\n",
       "      <td>yes</td>\n",
       "      <td>39.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>56</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>35</td>\n",
       "      <td>63</td>\n",
       "      <td>no_data</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>corn chips or salsa</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Rarely (few times/month)</td>\n",
       "      <td>yes</td>\n",
       "      <td>32.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>56</td>\n",
       "      <td>64</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>fruit</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Daily</td>\n",
       "      <td>no</td>\n",
       "      <td>40</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>37</td>\n",
       "      <td>90</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>Bread</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>no</td>\n",
       "      <td>40.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>61</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>72</td>\n",
       "      <td>86</td>\n",
       "      <td>no_data</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Decreased more than 10 pounds</td>\n",
       "      <td>bread</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Occasionally (1-2 times/week)</td>\n",
       "      <td>no</td>\n",
       "      <td>43.1</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>55</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>65</td>\n",
       "      <td>83</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>bread</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Regularly (3-5 times/week)</td>\n",
       "      <td>no</td>\n",
       "      <td>42.8</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>60</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>34</td>\n",
       "      <td>101</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>Depression</td>\n",
       "      <td>Caffeine</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>Grains, pasta, rice</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>yes</td>\n",
       "      <td>46.8</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>37</td>\n",
       "      <td>yes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>65</td>\n",
       "      <td>73</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>raspberries</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>yes</td>\n",
       "      <td>44.7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>32</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>72</td>\n",
       "      <td>58</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:tongue</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>yogurt</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>no_data</td>\n",
       "      <td>33.1</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>50</td>\n",
       "      <td>yes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>54</td>\n",
       "      <td>80</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>Stress</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>Vegetables</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Rarely (few times/month)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>37.8</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>52</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>32</td>\n",
       "      <td>54</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>wheat</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>yes</td>\n",
       "      <td>40</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>50</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>22</td>\n",
       "      <td>70</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>Caffeine</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>Bread</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Rarely (few times/month)</td>\n",
       "      <td>yes</td>\n",
       "      <td>26.1</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>20</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>49</td>\n",
       "      <td>79</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>starch</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>64.8</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>40</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>66</td>\n",
       "      <td>61</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:skin</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>bread</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>no</td>\n",
       "      <td>33.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>63</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>26</td>\n",
       "      <td>59</td>\n",
       "      <td>no_data</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>bread</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Occasionally (1-2 times/week)</td>\n",
       "      <td>no</td>\n",
       "      <td>37.3</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>30</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>35</td>\n",
       "      <td>86</td>\n",
       "      <td>no_data</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>nuts</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>yes</td>\n",
       "      <td>40.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>51</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>50</td>\n",
       "      <td>58</td>\n",
       "      <td>no_data</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>Potato, corn, rice</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>no</td>\n",
       "      <td>35.1</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>20</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>60</td>\n",
       "      <td>64</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Increased more than 10 pounds</td>\n",
       "      <td>vegetables - salad, broccoli, cauliflower</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>unknown</td>\n",
       "      <td>41.7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>35</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>40</td>\n",
       "      <td>63</td>\n",
       "      <td>no_data</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>almonds</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Rarely (few times/month)</td>\n",
       "      <td>no</td>\n",
       "      <td>44.8</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>63</td>\n",
       "      <td>56</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>Other</td>\n",
       "      <td>Foods; other</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>whole grain</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Never</td>\n",
       "      <td>no</td>\n",
       "      <td>40.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>61</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>51</td>\n",
       "      <td>81</td>\n",
       "      <td>no_data</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>wheat</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Occasionally (1-2 times/week)</td>\n",
       "      <td>yes</td>\n",
       "      <td>43</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>46</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>32</td>\n",
       "      <td>73</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Decreased more than 10 pounds</td>\n",
       "      <td>Rice</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>no</td>\n",
       "      <td>40.8</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>62</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>50</td>\n",
       "      <td>90</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:nostril</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>Potato</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>yes</td>\n",
       "      <td>35.8</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>52</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>64</td>\n",
       "      <td>58</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>Other</td>\n",
       "      <td>Other</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>wheat flour</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>yes</td>\n",
       "      <td>41.1</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>60</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>fruit</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>no</td>\n",
       "      <td>40.7</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>35</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>43</td>\n",
       "      <td>68</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Decreased more than 10 pounds</td>\n",
       "      <td>Sweet Potato</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Rarely (few times/month)</td>\n",
       "      <td>no</td>\n",
       "      <td>38.1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>45</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>62</td>\n",
       "      <td>80</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>Other</td>\n",
       "      <td>Lack of sleep</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>wheat</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>yes</td>\n",
       "      <td>30.2</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>12</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>53</td>\n",
       "      <td>75</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>chocolate</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>no</td>\n",
       "      <td>45.5</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>48</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>39</td>\n",
       "      <td>72</td>\n",
       "      <td>no_data</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>Other</td>\n",
       "      <td>Other</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>Oat</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Occasionally (1-2 times/week)</td>\n",
       "      <td>no_data</td>\n",
       "      <td>33.9</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>4</td>\n",
       "      <td>15</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>...</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>51.5</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>15</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>42</td>\n",
       "      <td>89</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Decreased more than 10 pounds</td>\n",
       "      <td>lentil</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Rarely (few times/month)</td>\n",
       "      <td>no</td>\n",
       "      <td>32.8</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>25</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>43</td>\n",
       "      <td>56</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>Sweet potato</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Occasionally (1-2 times/week)</td>\n",
       "      <td>no</td>\n",
       "      <td>33.5</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>50</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>34</td>\n",
       "      <td>63</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>Oats</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>yes</td>\n",
       "      <td>38.6</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>49</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>23</td>\n",
       "      <td>67</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>rice</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Rarely (few times/month)</td>\n",
       "      <td>no</td>\n",
       "      <td>13.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>58</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>64</td>\n",
       "      <td>89</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:tongue</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>bread</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Never</td>\n",
       "      <td>no</td>\n",
       "      <td>39</td>\n",
       "      <td>unknown</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>25</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>48</td>\n",
       "      <td>89</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>corn</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>yes</td>\n",
       "      <td>Never</td>\n",
       "      <td>no</td>\n",
       "      <td>29.7</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>35</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>58</td>\n",
       "      <td>51</td>\n",
       "      <td>no_data</td>\n",
       "      <td>UBERON:tongue</td>\n",
       "      <td>Lack of sleep</td>\n",
       "      <td>Foods; other</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>Milk</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Rarely (few times/month)</td>\n",
       "      <td>no</td>\n",
       "      <td>33.8</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>48.5</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>59</td>\n",
       "      <td>89</td>\n",
       "      <td>no_data</td>\n",
       "      <td>UBERON:hand</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>Potato</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Rarely (few times/month)</td>\n",
       "      <td>no</td>\n",
       "      <td>41.4</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>59</td>\n",
       "      <td>yes</td>\n",
       "      <td>no_data</td>\n",
       "      <td>64</td>\n",
       "      <td>81</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>orange juice</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Rarely (few times/month)</td>\n",
       "      <td>no</td>\n",
       "      <td>38.2</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>48</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>46</td>\n",
       "      <td>70</td>\n",
       "      <td>no</td>\n",
       "      <td>UBERON:feces</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>Remained stable</td>\n",
       "      <td>Bread</td>\n",
       "      <td>I do not have diabetes</td>\n",
       "      <td>no</td>\n",
       "      <td>Rarely (few times/month)</td>\n",
       "      <td>no_data</td>\n",
       "      <td>39.4</td>\n",
       "      <td>no_data</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40 rows × 110 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS AGE  \\\n",
       "0                55               no_data                    no_data  51   \n",
       "1                56               no_data                    no_data  35   \n",
       "2                16               no_data                    no_data  56   \n",
       "3           no_data               no_data                    no_data  37   \n",
       "4                61                   yes                    no_data  72   \n",
       "5                55               no_data                    no_data  65   \n",
       "6                60               no_data                    no_data  34   \n",
       "7                37                   yes                    unknown  65   \n",
       "8                32                   yes                    no_data  72   \n",
       "9                50                   yes                    unknown  54   \n",
       "10               52               unknown                    unknown  32   \n",
       "11               50               no_data                    no_data  22   \n",
       "12               20               unknown                    unknown  49   \n",
       "13               40                   yes                    no_data  66   \n",
       "14               63               no_data                    no_data  26   \n",
       "15               30               no_data                    no_data  35   \n",
       "16               51               no_data                    no_data  50   \n",
       "17               20               unknown                    unknown  60   \n",
       "18               35               no_data                    no_data  40   \n",
       "19          no_data                   yes                    no_data  63   \n",
       "21               61               no_data                    no_data  51   \n",
       "22               46               no_data                    no_data  32   \n",
       "23               62               no_data                    no_data  50   \n",
       "24               52                   yes                    no_data  64   \n",
       "25               60               no_data                    no_data  67   \n",
       "26               35                   yes                        yes  43   \n",
       "27               45               no_data                    no_data  62   \n",
       "28               12               no_data                    no_data  53   \n",
       "29               48               no_data                    no_data  39   \n",
       "30          no_data               no_data                    no_data   4   \n",
       "31               15               no_data                    no_data  42   \n",
       "32               25               no_data                    no_data  43   \n",
       "33               50               no_data                    no_data  34   \n",
       "34               49               no_data                    no_data  23   \n",
       "36               58               unknown                    unknown  64   \n",
       "37               25               no_data                    no_data  48   \n",
       "39               35               no_data                    no_data  58   \n",
       "40             48.5               no_data                    no_data  59   \n",
       "41               59                   yes                    no_data  64   \n",
       "42               48               no_data                    no_data  46   \n",
       "\n",
       "   TOT_MASS QUINOLINE       BODY_SITE MIGRAINE_FACTOR_2 MIGRAINE_FACTOR_3  \\\n",
       "0        50        no    UBERON:feces           no_data           no_data   \n",
       "1        63   no_data    UBERON:feces           no_data           no_data   \n",
       "2        64        no    UBERON:feces           no_data           no_data   \n",
       "3        90        no    UBERON:feces           no_data           no_data   \n",
       "4        86   no_data    UBERON:feces           no_data           no_data   \n",
       "5        83        no    UBERON:feces           no_data           no_data   \n",
       "6       101        no    UBERON:feces        Depression          Caffeine   \n",
       "7        73        no    UBERON:feces           unknown           unknown   \n",
       "8        58        no   UBERON:tongue           no_data           no_data   \n",
       "9        80        no    UBERON:feces            Stress           unknown   \n",
       "10       54        no    UBERON:feces           unknown           unknown   \n",
       "11       70        no    UBERON:feces          Caffeine           no_data   \n",
       "12       79        no    UBERON:feces           unknown           unknown   \n",
       "13       61        no     UBERON:skin           no_data           no_data   \n",
       "14       59   no_data    UBERON:feces           no_data           no_data   \n",
       "15       86   no_data    UBERON:feces           no_data           no_data   \n",
       "16       58   no_data    UBERON:feces           no_data           no_data   \n",
       "17       64        no    UBERON:feces           unknown           unknown   \n",
       "18       63   no_data    UBERON:feces           no_data           no_data   \n",
       "19       56        no    UBERON:feces             Other      Foods; other   \n",
       "21       81   no_data    UBERON:feces           no_data           no_data   \n",
       "22       73        no    UBERON:feces           no_data           no_data   \n",
       "23       90        no  UBERON:nostril           no_data           no_data   \n",
       "24       58        no    UBERON:feces             Other             Other   \n",
       "25       67        no    UBERON:feces           no_data           no_data   \n",
       "26       68        no    UBERON:feces           unknown           unknown   \n",
       "27       80        no    UBERON:feces             Other     Lack of sleep   \n",
       "28       75        no    UBERON:feces           no_data           no_data   \n",
       "29       72   no_data    UBERON:feces             Other             Other   \n",
       "30       15        no    UBERON:feces           no_data           no_data   \n",
       "31       89        no    UBERON:feces           no_data           no_data   \n",
       "32       56        no    UBERON:feces           no_data           no_data   \n",
       "33       63        no    UBERON:feces           no_data           no_data   \n",
       "34       67        no    UBERON:feces           no_data           no_data   \n",
       "36       89        no   UBERON:tongue           unknown           unknown   \n",
       "37       89        no    UBERON:feces           no_data           no_data   \n",
       "39       51   no_data   UBERON:tongue     Lack of sleep      Foods; other   \n",
       "40       89   no_data     UBERON:hand           no_data           no_data   \n",
       "41       81        no    UBERON:feces           no_data           no_data   \n",
       "42       70        no    UBERON:feces           no_data           no_data   \n",
       "\n",
       "   TONSILS_REMOVED  ...                  WEIGHT_CHANGE  \\\n",
       "0          no_data  ...                Remained stable   \n",
       "1               no  ...                Remained stable   \n",
       "2              yes  ...                Remained stable   \n",
       "3               no  ...                Remained stable   \n",
       "4               no  ...  Decreased more than 10 pounds   \n",
       "5               no  ...                Remained stable   \n",
       "6               no  ...                Remained stable   \n",
       "7              yes  ...                Remained stable   \n",
       "8               no  ...                Remained stable   \n",
       "9               no  ...                Remained stable   \n",
       "10              no  ...                Remained stable   \n",
       "11              no  ...                Remained stable   \n",
       "12         unknown  ...                Remained stable   \n",
       "13              no  ...                Remained stable   \n",
       "14              no  ...                Remained stable   \n",
       "15              no  ...                Remained stable   \n",
       "16              no  ...                Remained stable   \n",
       "17              no  ...  Increased more than 10 pounds   \n",
       "18              no  ...                Remained stable   \n",
       "19             yes  ...                Remained stable   \n",
       "21              no  ...                Remained stable   \n",
       "22              no  ...  Decreased more than 10 pounds   \n",
       "23              no  ...                Remained stable   \n",
       "24              no  ...                Remained stable   \n",
       "25              no  ...                Remained stable   \n",
       "26              no  ...  Decreased more than 10 pounds   \n",
       "27              no  ...                Remained stable   \n",
       "28              no  ...                Remained stable   \n",
       "29              no  ...                Remained stable   \n",
       "30         no_data  ...                        no_data   \n",
       "31              no  ...  Decreased more than 10 pounds   \n",
       "32             yes  ...                Remained stable   \n",
       "33              no  ...                Remained stable   \n",
       "34              no  ...                Remained stable   \n",
       "36             yes  ...                Remained stable   \n",
       "37             yes  ...                Remained stable   \n",
       "39             yes  ...                Remained stable   \n",
       "40             yes  ...                Remained stable   \n",
       "41              no  ...                Remained stable   \n",
       "42              no  ...                Remained stable   \n",
       "\n",
       "                                 PRIMARY_CARB                DIABETES  \\\n",
       "0                                      banana  I do not have diabetes   \n",
       "1                         corn chips or salsa  I do not have diabetes   \n",
       "2                                       fruit  I do not have diabetes   \n",
       "3                                       Bread  I do not have diabetes   \n",
       "4                                       bread  I do not have diabetes   \n",
       "5                                       bread  I do not have diabetes   \n",
       "6                         Grains, pasta, rice  I do not have diabetes   \n",
       "7                                 raspberries  I do not have diabetes   \n",
       "8                                      yogurt  I do not have diabetes   \n",
       "9                                  Vegetables  I do not have diabetes   \n",
       "10                                      wheat  I do not have diabetes   \n",
       "11                                      Bread  I do not have diabetes   \n",
       "12                                     starch                 unknown   \n",
       "13                                      bread  I do not have diabetes   \n",
       "14                                      bread  I do not have diabetes   \n",
       "15                                       nuts  I do not have diabetes   \n",
       "16                         Potato, corn, rice  I do not have diabetes   \n",
       "17  vegetables - salad, broccoli, cauliflower  I do not have diabetes   \n",
       "18                                    almonds  I do not have diabetes   \n",
       "19                                whole grain  I do not have diabetes   \n",
       "21                                      wheat  I do not have diabetes   \n",
       "22                                       Rice  I do not have diabetes   \n",
       "23                                     Potato  I do not have diabetes   \n",
       "24                                wheat flour  I do not have diabetes   \n",
       "25                                      fruit  I do not have diabetes   \n",
       "26                               Sweet Potato  I do not have diabetes   \n",
       "27                                      wheat  I do not have diabetes   \n",
       "28                                  chocolate  I do not have diabetes   \n",
       "29                                        Oat  I do not have diabetes   \n",
       "30                                    no_data                 no_data   \n",
       "31                                     lentil  I do not have diabetes   \n",
       "32                               Sweet potato  I do not have diabetes   \n",
       "33                                       Oats  I do not have diabetes   \n",
       "34                                       rice  I do not have diabetes   \n",
       "36                                      bread  I do not have diabetes   \n",
       "37                                       corn  I do not have diabetes   \n",
       "39                                       Milk  I do not have diabetes   \n",
       "40                                     Potato  I do not have diabetes   \n",
       "41                               orange juice  I do not have diabetes   \n",
       "42                                      Bread  I do not have diabetes   \n",
       "\n",
       "     ASTHMA                 POOL_FREQUENCY      CAT LATITUDE  \\\n",
       "0        no  Occasionally (1-2 times/week)      yes     39.7   \n",
       "1        no       Rarely (few times/month)      yes     32.7   \n",
       "2        no                          Daily       no       40   \n",
       "3        no                          Never       no     40.7   \n",
       "4       yes  Occasionally (1-2 times/week)       no     43.1   \n",
       "5        no     Regularly (3-5 times/week)       no     42.8   \n",
       "6        no                          Never      yes     46.8   \n",
       "7        no                          Never      yes     44.7   \n",
       "8        no                          Never  no_data     33.1   \n",
       "9        no       Rarely (few times/month)  unknown     37.8   \n",
       "10       no                          Never      yes       40   \n",
       "11       no       Rarely (few times/month)      yes     26.1   \n",
       "12  unknown                        unknown      yes     64.8   \n",
       "13       no                          Never       no     33.7   \n",
       "14      yes  Occasionally (1-2 times/week)       no     37.3   \n",
       "15       no                          Never      yes     40.7   \n",
       "16       no                          Never       no     35.1   \n",
       "17       no                          Never  unknown     41.7   \n",
       "18       no       Rarely (few times/month)       no     44.8   \n",
       "19      yes                          Never       no     40.7   \n",
       "21       no  Occasionally (1-2 times/week)      yes       43   \n",
       "22       no                          Never       no     40.8   \n",
       "23       no                          Never      yes     35.8   \n",
       "24       no                          Never      yes     41.1   \n",
       "25       no                          Never       no     40.7   \n",
       "26       no       Rarely (few times/month)       no     38.1   \n",
       "27       no                          Never      yes     30.2   \n",
       "28       no                          Never       no     45.5   \n",
       "29      yes  Occasionally (1-2 times/week)  no_data     33.9   \n",
       "30      yes                        no_data       no     51.5   \n",
       "31       no       Rarely (few times/month)       no     32.8   \n",
       "32       no  Occasionally (1-2 times/week)       no     33.5   \n",
       "33       no                          Never      yes     38.6   \n",
       "34       no       Rarely (few times/month)       no     13.7   \n",
       "36       no                          Never       no       39   \n",
       "37      yes                          Never       no     29.7   \n",
       "39       no       Rarely (few times/month)       no     33.8   \n",
       "40       no       Rarely (few times/month)       no     41.4   \n",
       "41       no       Rarely (few times/month)       no     38.2   \n",
       "42       no       Rarely (few times/month)  no_data     39.4   \n",
       "\n",
       "   FOODALLERGIES_PEANUTS CHICKENPOX SEASONAL_ALLERGIES  \n",
       "0                no_data        yes                 no  \n",
       "1                no_data        yes                 no  \n",
       "2                no_data        yes                 no  \n",
       "3                no_data        yes                 no  \n",
       "4                no_data        yes                yes  \n",
       "5                no_data        yes                yes  \n",
       "6                no_data        yes                yes  \n",
       "7                unknown        yes                yes  \n",
       "8                no_data    no_data                 no  \n",
       "9                unknown        yes                 no  \n",
       "10               unknown        yes                 no  \n",
       "11               no_data    no_data                 no  \n",
       "12               unknown    unknown            unknown  \n",
       "13               no_data        yes                 no  \n",
       "14                   yes        yes                yes  \n",
       "15               no_data        yes                 no  \n",
       "16               no_data        yes                 no  \n",
       "17               unknown        yes                 no  \n",
       "18               no_data    no_data                 no  \n",
       "19               no_data        yes                yes  \n",
       "21               no_data        yes                yes  \n",
       "22               no_data        yes                yes  \n",
       "23               no_data        yes                 no  \n",
       "24               no_data        yes                 no  \n",
       "25                   yes    no_data                 no  \n",
       "26               unknown        yes                yes  \n",
       "27               no_data        yes                yes  \n",
       "28               no_data        yes                 no  \n",
       "29               no_data        yes                yes  \n",
       "30                   yes    no_data                yes  \n",
       "31               no_data         no                 no  \n",
       "32               no_data        yes            no_data  \n",
       "33               no_data        yes                yes  \n",
       "34               no_data         no                 no  \n",
       "36               unknown        yes                yes  \n",
       "37               no_data        yes                yes  \n",
       "39               no_data        yes                 no  \n",
       "40               no_data         no                 no  \n",
       "41               no_data        yes                 no  \n",
       "42               no_data        yes                yes  \n",
       "\n",
       "[40 rows x 110 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove rows based on two conditions.\n",
    "\n",
    "df_3 = df_2.drop(df_2[(df_2['AGE'] == 'unknown') | (df_2['AGE'] == 'no_data')].index)\n",
    "df_3.head(40)\n",
    "\n",
    "# https://stackoverflow.com/questions/13851535/how-to-delete-rows-from-a-pandas-dataframe-based-on-a-conditional-expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4241, 110)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_3.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4815, 110)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "574"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "deleted_rows_blank = df_2.shape[0] - df_3.shape[0]\n",
    "\n",
    "deleted_rows_blank"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "574 rows where deleted since they were just blank. The updated data frame df_3 has 4241 rows and 110 columns."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CARBOHYDRATE_PER\n",
      "NONFOODALLERGIES_DRUG\n",
      "NONFOODALLERGIES_BEESTINGS\n",
      "AGE\n",
      "TOT_MASS\n",
      "QUINOLINE\n",
      "BODY_SITE\n",
      "MIGRAINE_FACTOR_2\n",
      "MIGRAINE_FACTOR_3\n",
      "TONSILS_REMOVED\n",
      "MIGRAINE_FACTOR_1\n",
      "ELEVATION\n",
      "FLOSSING_FREQUENCY\n",
      "PROTEIN_PER\n",
      "CSECTION\n",
      "COUNTRY_OF_BIRTH\n",
      "MACRONUTRIENT_PCT_TOTAL\n",
      "COSMETICS_FREQUENCY\n",
      "GLUTEN\n",
      "LONGITUDE\n",
      "DOG\n",
      "MIGRAINE_AGGRAVATION\n",
      "DIABETES_MEDICATION\n",
      "PENICILLIN\n",
      "MIGRAINE_FREQUENCY\n",
      "MIGRAINE_RELATIVES\n",
      "PLANT_PER\n",
      "MIGRAINE_PHOTOPHOBIA\n",
      "SITE_SAMPLED\n",
      "DRINKING_WATER_SOURCE\n",
      "COUNTRY\n",
      "ANTIBIOTIC_SELECT\n",
      "MIGRAINE_NAUSEA\n",
      "FOODALLERGIES_TREENUTS\n",
      "MIGRAINE_AURA\n",
      "FLU_VACCINE_DATE\n",
      "PREGNANT\n",
      "MIGRAINE\n",
      "BMI\n",
      "TEETHBRUSHING_FREQUENCY\n",
      "MIGRAINE_MEDICATIONS\n",
      "CONTRACEPTIVE\n",
      "LIVINGWITH\n",
      "RELATIONS\n",
      "DIABETES_MEDICATIONS\n",
      "SUPPLEMENTS\n",
      "AGE_UNIT\n",
      "EXERCISE_FREQUENCY\n",
      "SULFA_DRUG\n",
      "NONFOODALLERGIES_NO\n",
      "ENV_MATTER\n",
      "SKIN_CONDITION\n",
      "CONDITIONS_MEDICATION\n",
      "LAST_TRAVEL\n",
      "HEIGHT_IN\n",
      "ANTIBIOTIC_CONDITION\n",
      "PET_CONTACT\n",
      "NONFOODALLERGIES_SUN\n",
      "EXERCISE_LOCATION\n",
      "NITROIMIDAZOLE\n",
      "DEODORANT_USE\n",
      "FOODALLERGIES_SHELLFISH\n",
      "SEX\n",
      "PERCENTAGE_FROM_CARBS\n",
      "FAT_PER\n",
      "SHARED_HOUSING\n",
      "SOFTENER\n",
      "MIGRAINE_PHONOPHOBIA\n",
      "SMOKING_FREQUENCY\n",
      "WEIGHT_LBS\n",
      "TANNING_BEDS\n",
      "ALCOHOL_FREQUENCY\n",
      "STATE\n",
      "BODY_HABITAT\n",
      "DOMINANT_HAND\n",
      "PRIMARY_VEGETABLE\n",
      "SPECIAL_RESTRICTIONS\n",
      "CEPHALOSPORIN\n",
      "SUPPLEMENTS_FIELDS\n",
      "APPENDIX_REMOVED\n",
      "CURRENT_RESIDENCE_DURATION\n",
      "SLEEP_DURATION\n",
      "DIET_TYPE\n",
      "DIET_RESTRICTIONS\n",
      "LACTOSE\n",
      "HEIGHT_OR_LENGTH\n",
      "ANTIBIOTIC_MEDS\n",
      "IBD\n",
      "TYPES_OF_PLANTS\n",
      "MIGRAINEMEDS\n",
      "FIBER_GRAMS\n",
      "NAILS\n",
      "RACE\n",
      "PETS\n",
      "MULTIVITAMIN\n",
      "NONFOODALLERGIES_POISONIVY\n",
      "ACNE_MEDICATION\n",
      "ACNE_MEDICATION_OTC\n",
      "ANIMAL_PER\n",
      "FOODALLERGIES_OTHER\n",
      "WEIGHT_CHANGE\n",
      "PRIMARY_CARB\n",
      "DIABETES\n",
      "ASTHMA\n",
      "POOL_FREQUENCY\n",
      "CAT\n",
      "LATITUDE\n",
      "FOODALLERGIES_PEANUTS\n",
      "CHICKENPOX\n",
      "SEASONAL_ALLERGIES\n"
     ]
    }
   ],
   "source": [
    "df_3_columns = list(df_3.columns.values) \n",
    "\n",
    "for c in df_3_columns: \n",
    "    print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "absolute number of unique entries in column: no_data    498\n",
      "50         284\n",
      "40         210\n",
      "30         207\n",
      "unknown    192\n",
      "          ... \n",
      "42.3         1\n",
      "22.2         1\n",
      "51.7         1\n",
      "49.8         1\n",
      "47.1         1\n",
      "Name: CARBOHYDRATE_PER, Length: 157, dtype: int64\n",
      "percentage of unique entries in column: no_data    0.12\n",
      "50         0.07\n",
      "40         0.05\n",
      "30         0.05\n",
      "unknown    0.05\n",
      "           ... \n",
      "42.3       0.00\n",
      "22.2       0.00\n",
      "51.7       0.00\n",
      "49.8       0.00\n",
      "47.1       0.00\n",
      "Name: CARBOHYDRATE_PER, Length: 157, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "abs_no_data_in_col = df_3.CARBOHYDRATE_PER.value_counts()\n",
    "\n",
    "percentage_no_data_in_col = round(df_3.CARBOHYDRATE_PER.value_counts(normalize=True), 2)\n",
    "\n",
    "print(f'absolute number of unique entries in column: {abs_no_data_in_col}')\n",
    "\n",
    "print(f'percentage of unique entries in column: {percentage_no_data_in_col}')\n",
    "\n",
    "# https://cmdlinetips.com/2020/08/pandas-value_counts-how-to-get-frequency-counts-of-variables-in-a-dataframe/\n",
    "# https://cmdlinetips.com/2020/08/pandas-value_counts-how-to-get-frequency-counts-of-variables-in-a-dataframe/\n",
    "# https://www.w3schools.com/python/ref_func_round.asp\n",
    "# https://realpython.com/python-f-strings/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Keep this column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "absolute number of unique entries in column: no_data    2722\n",
      "yes         792\n",
      "unknown     727\n",
      "Name: NONFOODALLERGIES_DRUG, dtype: int64\n",
      "percentage of unique entries in column: no_data    0.64\n",
      "yes        0.19\n",
      "unknown    0.17\n",
      "Name: NONFOODALLERGIES_DRUG, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "abs_no_data_in_col = df_3.NONFOODALLERGIES_DRUG.value_counts()\n",
    "\n",
    "percentage_no_data_in_col = round(df_3.NONFOODALLERGIES_DRUG.value_counts(normalize=True), 2)\n",
    "\n",
    "print(f'absolute number of unique entries in column: {abs_no_data_in_col}')\n",
    "\n",
    "print(f'percentage of unique entries in column: {percentage_no_data_in_col}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Keep this column."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ask if there are any Nan in the data frame.\n",
    "\n",
    "df_3.isnull().values.any()\n",
    "\n",
    "# https://datatofish.com/check-nan-pandas-dataframe/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are NaN values in the data frame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ask how many Nan there are in the data frame.\n",
    "\n",
    "df_3.isnull().sum().sum()\n",
    "\n",
    "# https://datatofish.com/check-nan-pandas-dataframe/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 9 rows that contain NaN values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS AGE  \\\n",
      "190                69               no_data                    no_data  14   \n",
      "208                20               unknown                    unknown  37   \n",
      "522                20                   yes                    unknown  39   \n",
      "1369               40               unknown                    unknown  37   \n",
      "1980          no_data               no_data                    no_data  45   \n",
      "2500          no_data               no_data                    no_data  34   \n",
      "3051          no_data               no_data                    no_data  34   \n",
      "3306               20               unknown                    unknown  38   \n",
      "3997               60               no_data                    no_data   1   \n",
      "\n",
      "     TOT_MASS QUINOLINE     BODY_SITE MIGRAINE_FACTOR_2 MIGRAINE_FACTOR_3  \\\n",
      "190        72        no  UBERON:feces           no_data           no_data   \n",
      "208        84        no  UBERON:feces           unknown           unknown   \n",
      "522        81        no  UBERON:feces           unknown           unknown   \n",
      "1369       71        no  UBERON:feces           unknown           unknown   \n",
      "1980       74   no_data  UBERON:feces           no_data           no_data   \n",
      "2500       64        no  UBERON:feces           no_data           no_data   \n",
      "3051       64        no   UBERON:skin           no_data           no_data   \n",
      "3306       56        no  UBERON:feces           unknown           unknown   \n",
      "3997        8        no  UBERON:feces           no_data           no_data   \n",
      "\n",
      "     TONSILS_REMOVED  ...                  WEIGHT_CHANGE    PRIMARY_CARB  \\\n",
      "190               no  ...  Increased more than 10 pounds           wheat   \n",
      "208              yes  ...                Remained stable  dark chocolate   \n",
      "522               no  ...                Remained stable      white rice   \n",
      "1369              no  ...                Remained stable         cookies   \n",
      "1980              no  ...  Increased more than 10 pounds         no_data   \n",
      "2500         no_data  ...                        no_data         no_data   \n",
      "3051         no_data  ...                        no_data         no_data   \n",
      "3306              no  ...                Remained stable      white rice   \n",
      "3997              no  ...  Increased more than 10 pounds      breastmilk   \n",
      "\n",
      "                    DIABETES ASTHMA            POOL_FREQUENCY  CAT LATITUDE  \\\n",
      "190   I do not have diabetes     no  Rarely (few times/month)   no       42   \n",
      "208   I do not have diabetes     no  Rarely (few times/month)  yes     49.4   \n",
      "522   I do not have diabetes     no  Rarely (few times/month)  yes     49.4   \n",
      "1369  I do not have diabetes     no  Rarely (few times/month)  yes     49.4   \n",
      "1980  I do not have diabetes     no  Rarely (few times/month)   no     39.7   \n",
      "2500                 no_data     no                   no_data   no     51.5   \n",
      "3051                 no_data     no                   no_data   no     51.5   \n",
      "3306  I do not have diabetes     no  Rarely (few times/month)  yes     49.4   \n",
      "3997                 Type II     no  Rarely (few times/month)   no     33.8   \n",
      "\n",
      "     FOODALLERGIES_PEANUTS CHICKENPOX SEASONAL_ALLERGIES  \n",
      "190                no_data         no                 no  \n",
      "208                unknown        yes                 no  \n",
      "522                unknown        yes                yes  \n",
      "1369               unknown        yes                 no  \n",
      "1980               no_data        yes                 no  \n",
      "2500               no_data    no_data                yes  \n",
      "3051               no_data    no_data                yes  \n",
      "3306               unknown        yes                 no  \n",
      "3997               no_data         no                 no  \n",
      "\n",
      "[9 rows x 110 columns]\n"
     ]
    }
   ],
   "source": [
    "# Show rows with NaN.\n",
    "\n",
    "is_NaN = df_3.isnull()\n",
    "row_has_NaN = is_NaN.any(axis=1)\n",
    "rows_with_NaN = df_3[row_has_NaN]\n",
    "\n",
    "print(rows_with_NaN)\n",
    "\n",
    "# https://www.kite.com/python/answers/how-to-find-rows-with-nan-values-in-a-pandas-dataframe-in-python"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above are the 9 rows that contain NaN values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ANTIBIOTIC_CONDITION</th>\n",
       "      <th>STATE</th>\n",
       "      <th>PRIMARY_VEGETABLE</th>\n",
       "      <th>DIET_RESTRICTIONS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>no_data</td>\n",
       "      <td>CO</td>\n",
       "      <td>spinach</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>no_data</td>\n",
       "      <td>CA</td>\n",
       "      <td>tomatoes (I eat a lot of salsa)</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>no_data</td>\n",
       "      <td>CO</td>\n",
       "      <td>leafy greens</td>\n",
       "      <td>Grain free</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>no_data</td>\n",
       "      <td>NY</td>\n",
       "      <td>Potato</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dental cleaning</td>\n",
       "      <td>NY</td>\n",
       "      <td>broccoli</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4810</th>\n",
       "      <td>UTI</td>\n",
       "      <td>IL</td>\n",
       "      <td>Romaine Lettuce</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4811</th>\n",
       "      <td>Streptococcus A</td>\n",
       "      <td>unknown</td>\n",
       "      <td>tomatoes, cuncumber</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4812</th>\n",
       "      <td>no_data</td>\n",
       "      <td>WA</td>\n",
       "      <td>potato</td>\n",
       "      <td>Ray Peat diet. no grains, beans, minimal veget...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4813</th>\n",
       "      <td>no_data</td>\n",
       "      <td>IL</td>\n",
       "      <td>Leafy Greens</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4814</th>\n",
       "      <td>no_data</td>\n",
       "      <td>BANKOK</td>\n",
       "      <td>Leaves and fruits</td>\n",
       "      <td>no_data</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4241 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ANTIBIOTIC_CONDITION    STATE                PRIMARY_VEGETABLE  \\\n",
       "0                 no_data       CO                          spinach   \n",
       "1                 no_data       CA  tomatoes (I eat a lot of salsa)   \n",
       "2                 no_data       CO                     leafy greens   \n",
       "3                 no_data       NY                           Potato   \n",
       "4         dental cleaning       NY                         broccoli   \n",
       "...                   ...      ...                              ...   \n",
       "4810                  UTI       IL                  Romaine Lettuce   \n",
       "4811      Streptococcus A  unknown              tomatoes, cuncumber   \n",
       "4812              no_data       WA                           potato   \n",
       "4813              no_data       IL                     Leafy Greens   \n",
       "4814              no_data   BANKOK                Leaves and fruits   \n",
       "\n",
       "                                      DIET_RESTRICTIONS  \n",
       "0                                               no_data  \n",
       "1                                               no_data  \n",
       "2                                            Grain free  \n",
       "3                                               no_data  \n",
       "4                                               no_data  \n",
       "...                                                 ...  \n",
       "4810                                            unknown  \n",
       "4811                                            unknown  \n",
       "4812  Ray Peat diet. no grains, beans, minimal veget...  \n",
       "4813                                            no_data  \n",
       "4814                                            no_data  \n",
       "\n",
       "[4241 rows x 4 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show columns with NaN.\n",
    "\n",
    "df_3.loc[:, df.isna().any()]\n",
    "\n",
    "# https://stackoverflow.com/questions/36226083/how-to-find-which-columns-contain-any-nan-value-in-pandas-dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above are the 4 columns that contain NaN values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop rows with NaN values. This results in df_4.\n",
    "\n",
    "df_4 = df_3.dropna(how='any') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4232, 110)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_4.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4241, 110)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_3.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "deleted_rows_NaN = df_3.shape[0] - df_4.shape[0]\n",
    "\n",
    "deleted_rows_NaN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "9 rows where deleted since they contained NaN values. The updated data frame **df_4 has 4232 rows and 110 columns**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ask if there are any Nan in the data frame.\n",
    "\n",
    "df_4.isnull().values.any()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are no more NaN values in the data frame df_4."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CARBOHYDRATE_PER              object\n",
       "NONFOODALLERGIES_DRUG         object\n",
       "NONFOODALLERGIES_BEESTINGS    object\n",
       "AGE                           object\n",
       "TOT_MASS                      object\n",
       "                               ...  \n",
       "CAT                           object\n",
       "LATITUDE                      object\n",
       "FOODALLERGIES_PEANUTS         object\n",
       "CHICKENPOX                    object\n",
       "SEASONAL_ALLERGIES            object\n",
       "Length: 110, dtype: object"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show the data types of all columns.\n",
    "\n",
    "df_4.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Keep data types as objects for first round of modelling."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Changing of data type and feature engineering "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Change data type from object to float for 3 columns initially. Later on this will be done with more columns. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**AGE**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('O')"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_4['AGE'].dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_5 = df_4.astype({'AGE': 'float64'})\n",
    "\n",
    "# https://thispointer.com/pandas-change-data-type-of-single-or-multiple-columns-of-dataframe-in-python/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_5['AGE'].dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**LATITUDE** - **IMPORTANT** for **Vitamin D3** levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('O')"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check data type\n",
    "\n",
    "df_5['LATITUDE'].dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "could not convert string to float: 'no_data'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-68-857600eaaf76>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Trying to change data type does not work: \"ValueError: could not convert string to float: 'no_data'.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mdf_6\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_5\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m'LATITUDE'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'float64'\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mastype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m   5681\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mcol_name\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5682\u001b[0m                     results.append(\n\u001b[0;32m-> 5683\u001b[0;31m                         \u001b[0mcol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcol_name\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5684\u001b[0m                     )\n\u001b[1;32m   5685\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mastype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m   5696\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5697\u001b[0m             \u001b[0;31m# else, only a single dtype is given\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5698\u001b[0;31m             \u001b[0mnew_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5699\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_constructor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__finalize__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5700\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mastype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m    580\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    581\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mbool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"raise\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 582\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"astype\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    583\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    584\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, f, filter, **kwargs)\u001b[0m\n\u001b[1;32m    440\u001b[0m                 \u001b[0mapplied\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    441\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 442\u001b[0;31m                 \u001b[0mapplied\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    443\u001b[0m             \u001b[0mresult_blocks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_extend_blocks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mapplied\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresult_blocks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    444\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mastype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m    623\u001b[0m             \u001b[0mvals1d\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mravel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    624\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 625\u001b[0;31m                 \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mastype_nansafe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvals1d\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    626\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mValueError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    627\u001b[0m                 \u001b[0;31m# e.g. astype_nansafe can fail on object-dtype of strings\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/pandas/core/dtypes/cast.py\u001b[0m in \u001b[0;36mastype_nansafe\u001b[0;34m(arr, dtype, copy, skipna)\u001b[0m\n\u001b[1;32m    895\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcopy\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mis_object_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mis_object_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    896\u001b[0m         \u001b[0;31m# Explicit copy, or required since NumPy can't view from / to object.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 897\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    898\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    899\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mview\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: could not convert string to float: 'no_data'"
     ]
    }
   ],
   "source": [
    "# Trying to change data type does not work: \"ValueError: could not convert string to float: 'no_data'.\"\n",
    "\n",
    "# df_6 = df_5.astype({'LATITUDE': 'float64'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('O')"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check data type is unchanged (object).\n",
    "\n",
    "df_5['LATITUDE'].dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "absolute number of no_data in column: 6\n",
      "percentage of no_data in column: 0.125\n"
     ]
    }
   ],
   "source": [
    "abs_no_data_in_col = (df_2.LATITUDE == 'no_data').sum()\n",
    "\n",
    "percentage_no_data_in_col = round(((((df_2.LATITUDE == 'no_data').sum())/df_2.shape[0])*100), 3)\n",
    "\n",
    "print(f'absolute number of no_data in column: {abs_no_data_in_col}')\n",
    "print(f'percentage of no_data in column: {percentage_no_data_in_col}')\n",
    "\n",
    "# https://stackoverflow.com/questions/35277075/python-pandas-counting-the-occurrences-of-a-specific-value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>COUNTRY_OF_BIRTH</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>STATE</th>\n",
       "      <th>CURRENT_RESIDENCE_DURATION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>no_data</td>\n",
       "      <td>no_data</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>678</th>\n",
       "      <td>no_data</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>no_data</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>GAZ:Switzerland</td>\n",
       "      <td>no_data</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1531</th>\n",
       "      <td>no_data</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3325</th>\n",
       "      <td>no_data</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3972</th>\n",
       "      <td>no_data</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>GAZ:Switzerland</td>\n",
       "      <td>no_data</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     LATITUDE COUNTRY_OF_BIRTH                       COUNTRY    STATE  \\\n",
       "217   no_data          no_data  GAZ:United States of America       MA   \n",
       "678   no_data              USA  GAZ:United States of America       MA   \n",
       "863   no_data      Switzerland               GAZ:Switzerland  no_data   \n",
       "1531  no_data              USA  GAZ:United States of America       MA   \n",
       "3325  no_data              USA  GAZ:United States of America       NY   \n",
       "3972  no_data      Switzerland               GAZ:Switzerland  no_data   \n",
       "\n",
       "     CURRENT_RESIDENCE_DURATION  \n",
       "217      More than one year ago  \n",
       "678      More than one year ago  \n",
       "863      More than one year ago  \n",
       "1531     More than one year ago  \n",
       "3325     More than one year ago  \n",
       "3972     More than one year ago  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_5[df_5['LATITUDE'] == 'no_data'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CO', 'CA', 'NY', 'MA', 'MN', 'VERMONT', 'FL', 'AK', 'PA', 'NC',\n",
       "       'MASSACHUSETTS', 'WI', 'NJ', 'OH', 'KY', 'TX', 'OR', '-', 'AZ',\n",
       "       'MO', 'BANKOK', 'MD', 'GA', 'CT', 'VA', 'IN', 'NH', 'ME',\n",
       "       'unknown', 'SC', 'IL', 'KS', 'NM', 'UT', 'WA', 'ILLINOIS',\n",
       "       'LONDON', 'MB', 'ND', 'HI', 'BRITISH COLUMBIA', 'LA', 'OK',\n",
       "       'TEXAS', 'BC', 'ALASKA', 'CALIFORNIA', 'IA', 'no_data', 'TN',\n",
       "       'WASHINGTON', 'CEREDIGION', 'GREATER LONDON', 'CAMDEN TOWN', 'NV',\n",
       "       'DC', 'MI', 'COLORADO', 'WV', 'AL', 'AR', 'ID', 'VIRGINIA', 'AB',\n",
       "       'TENNESSEE', 'VT', 'MARYLAND', 'SD', 'NE', 'MT', 'MS', 'NEW YORK',\n",
       "       'LONDON, CITY OF', 'WY', 'Bangkok', 'OHIO', 'CA - CALIFORNIA',\n",
       "       'QC', 'RI', 'ONTARIO', 'NSW', 'ROTHERHITHE', 'NONE', 'MAINE',\n",
       "       'OREGON', 'WISCONSIN', 'CONNECTICUT', 'SK', 'BUCKINGHAMSHIRE',\n",
       "       'WESTERN AUSTRALIA', 'FLORIDA', 'HAWAII', 'GA.', 'NEW MEXICO',\n",
       "       'VIC', 'MA - MASSACHUSETTS', 'COLORADO (CO)', 'SOUTH CAROLINA',\n",
       "       'KENTUCKY', 'PR', 'ON', 'MAZOWIECKIE', 'ARKANSAS', 'MISSOURI',\n",
       "       'NC NORTH CAROLINA', 'ZURICH', 'MINNESOTA', 'BEDFORDSHIRE',\n",
       "       'GEORGOA', 'OH- OHIO', 'THURINGIA', 'TX-TEXAS', 'MICHIGAN',\n",
       "       'QUEBEC', 'WELLINGTON', 'MONTANA', 'GVA', 'BUSKERUD'], dtype=object)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show unique entries.\n",
    "\n",
    "df_5.STATE.unique()\n",
    "\n",
    "# https://chrisalbon.com/python/data_wrangling/pandas_list_unique_values_in_column/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>COUNTRY_OF_BIRTH</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>STATE</th>\n",
       "      <th>CURRENT_RESIDENCE_DURATION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>42.8</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>41.9</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>42.3</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>41.6</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>42.3</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4665</th>\n",
       "      <td>42.4</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4703</th>\n",
       "      <td>42.3</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4724</th>\n",
       "      <td>42.4</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778</th>\n",
       "      <td>41.6</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4790</th>\n",
       "      <td>42.6</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>197 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     LATITUDE COUNTRY_OF_BIRTH                       COUNTRY STATE  \\\n",
       "5        42.8              USA  GAZ:United States of America    MA   \n",
       "43       41.9              USA  GAZ:United States of America    MA   \n",
       "61       42.3              USA  GAZ:United States of America    MA   \n",
       "79       41.6              USA  GAZ:United States of America    MA   \n",
       "105      42.3              USA  GAZ:United States of America    MA   \n",
       "...       ...              ...                           ...   ...   \n",
       "4665     42.4              USA  GAZ:United States of America    MA   \n",
       "4703     42.3              USA  GAZ:United States of America    MA   \n",
       "4724     42.4   United Kingdom  GAZ:United States of America    MA   \n",
       "4778     41.6              USA  GAZ:United States of America    MA   \n",
       "4790     42.6              USA  GAZ:United States of America    MA   \n",
       "\n",
       "     CURRENT_RESIDENCE_DURATION  \n",
       "5        More than one year ago  \n",
       "43       More than one year ago  \n",
       "61       More than one year ago  \n",
       "79       More than one year ago  \n",
       "105      More than one year ago  \n",
       "...                         ...  \n",
       "4665     More than one year ago  \n",
       "4703     More than one year ago  \n",
       "4724     More than one year ago  \n",
       "4778     More than one year ago  \n",
       "4790     More than one year ago  \n",
       "\n",
       "[197 rows x 5 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_state_MA = df_5[df_5['STATE'] == 'MA'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]\n",
    "df_state_MA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>STATE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>42.8</td>\n",
       "      <td>MA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>41.9</td>\n",
       "      <td>MA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>42.3</td>\n",
       "      <td>MA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>41.6</td>\n",
       "      <td>MA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>42.3</td>\n",
       "      <td>MA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>40.8</td>\n",
       "      <td>MA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>41.6</td>\n",
       "      <td>MA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>42.4</td>\n",
       "      <td>MA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    LATITUDE STATE\n",
       "5       42.8    MA\n",
       "43      41.9    MA\n",
       "61      42.3    MA\n",
       "79      41.6    MA\n",
       "105     42.3    MA\n",
       "112     40.8    MA\n",
       "178     41.6    MA\n",
       "188     42.4    MA"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_state_MA = df_5[df_5['STATE'] == 'MA'][['LATITUDE', 'STATE']]\n",
    "df_state_MA.head(8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_state_MA_8 = df_state_MA.head(8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('O')"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_state_MA_8['LATITUDE'].dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_state_MA_8_float = df_state_MA_8.astype({'LATITUDE': 'float64'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_state_MA_8_float['LATITUDE'].dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LATITUDE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>8.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>41.9625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.6255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>40.8000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>41.6000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>42.1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>42.3250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>42.8000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       LATITUDE\n",
       "count    8.0000\n",
       "mean    41.9625\n",
       "std      0.6255\n",
       "min     40.8000\n",
       "25%     41.6000\n",
       "50%     42.1000\n",
       "75%     42.3250\n",
       "max     42.8000"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_state_MA_8_float.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "LATITUDE    42.0\n",
       "dtype: float64"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_latitude_MA = round(df_state_MA_8_float.mean(axis=0), 1)\n",
    "mean_latitude_MA\n",
    "\n",
    "# https://stackoverflow.com/questions/31037298/pandas-get-column-average-mean"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use this calculated mean value of LATITUDE = 42.0 for the cells that have STATE = MA and have 'no_data' under LATITUDE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "42.0"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Replace value in a cell.\n",
    "\n",
    "df_5_1 = df_5.loc[217, 'LATITUDE'] = 42.0\n",
    "\n",
    "df_5_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>COUNTRY_OF_BIRTH</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>STATE</th>\n",
       "      <th>CURRENT_RESIDENCE_DURATION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>678</th>\n",
       "      <td>no_data</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>no_data</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>GAZ:Switzerland</td>\n",
       "      <td>no_data</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1531</th>\n",
       "      <td>no_data</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>MA</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3325</th>\n",
       "      <td>no_data</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3972</th>\n",
       "      <td>no_data</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>GAZ:Switzerland</td>\n",
       "      <td>no_data</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     LATITUDE COUNTRY_OF_BIRTH                       COUNTRY    STATE  \\\n",
       "678   no_data              USA  GAZ:United States of America       MA   \n",
       "863   no_data      Switzerland               GAZ:Switzerland  no_data   \n",
       "1531  no_data              USA  GAZ:United States of America       MA   \n",
       "3325  no_data              USA  GAZ:United States of America       NY   \n",
       "3972  no_data      Switzerland               GAZ:Switzerland  no_data   \n",
       "\n",
       "     CURRENT_RESIDENCE_DURATION  \n",
       "678      More than one year ago  \n",
       "863      More than one year ago  \n",
       "1531     More than one year ago  \n",
       "3325     More than one year ago  \n",
       "3972     More than one year ago  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_5[df_5['LATITUDE'] == 'no_data'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Replacement worked for that cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace value in a cell.\n",
    "\n",
    "df_5_2 = df_5.loc[678, 'LATITUDE'] = 42.0\n",
    "\n",
    "df_5_3 = df_5.loc[1531, 'LATITUDE'] = 42.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>COUNTRY_OF_BIRTH</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>STATE</th>\n",
       "      <th>CURRENT_RESIDENCE_DURATION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>no_data</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>GAZ:Switzerland</td>\n",
       "      <td>no_data</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3325</th>\n",
       "      <td>no_data</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3972</th>\n",
       "      <td>no_data</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>GAZ:Switzerland</td>\n",
       "      <td>no_data</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     LATITUDE COUNTRY_OF_BIRTH                       COUNTRY    STATE  \\\n",
       "863   no_data      Switzerland               GAZ:Switzerland  no_data   \n",
       "3325  no_data              USA  GAZ:United States of America       NY   \n",
       "3972  no_data      Switzerland               GAZ:Switzerland  no_data   \n",
       "\n",
       "     CURRENT_RESIDENCE_DURATION  \n",
       "863      More than one year ago  \n",
       "3325     More than one year ago  \n",
       "3972     More than one year ago  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_5[df_5['LATITUDE'] == 'no_data'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Replacement worked for those 2 cells."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>COUNTRY_OF_BIRTH</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>STATE</th>\n",
       "      <th>CURRENT_RESIDENCE_DURATION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40.7</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>43.1</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>40.7</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>43</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>40.7</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4761</th>\n",
       "      <td>42.6</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4762</th>\n",
       "      <td>44.3</td>\n",
       "      <td>Greece</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4776</th>\n",
       "      <td>40.7</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4792</th>\n",
       "      <td>40.7</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4798</th>\n",
       "      <td>42.6</td>\n",
       "      <td>USA</td>\n",
       "      <td>GAZ:United States of America</td>\n",
       "      <td>NY</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>240 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     LATITUDE COUNTRY_OF_BIRTH                       COUNTRY STATE  \\\n",
       "3        40.7              USA  GAZ:United States of America    NY   \n",
       "4        43.1              USA  GAZ:United States of America    NY   \n",
       "19       40.7              USA  GAZ:United States of America    NY   \n",
       "21         43              USA  GAZ:United States of America    NY   \n",
       "25       40.7              USA  GAZ:United States of America    NY   \n",
       "...       ...              ...                           ...   ...   \n",
       "4761     42.6              USA  GAZ:United States of America    NY   \n",
       "4762     44.3           Greece  GAZ:United States of America    NY   \n",
       "4776     40.7              USA  GAZ:United States of America    NY   \n",
       "4792     40.7              USA  GAZ:United States of America    NY   \n",
       "4798     42.6              USA  GAZ:United States of America    NY   \n",
       "\n",
       "     CURRENT_RESIDENCE_DURATION  \n",
       "3        More than one year ago  \n",
       "4        More than one year ago  \n",
       "19       More than one year ago  \n",
       "21       More than one year ago  \n",
       "25       More than one year ago  \n",
       "...                         ...  \n",
       "4761     More than one year ago  \n",
       "4762     More than one year ago  \n",
       "4776     More than one year ago  \n",
       "4792     More than one year ago  \n",
       "4798     More than one year ago  \n",
       "\n",
       "[240 rows x 5 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_state_NY = df_5[df_5['STATE'] == 'NY'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]\n",
    "df_state_NY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>STATE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>43.1</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>43</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>40.8</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>42.2</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>40.8</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>40.9</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>40.9</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>43.1</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>40.8</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>42.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>40.8</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>41.1</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>43.1</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>40.7</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>587</th>\n",
       "      <td>41.3</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>43</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>608</th>\n",
       "      <td>40.8</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>40.8</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    LATITUDE STATE\n",
       "3       40.7    NY\n",
       "4       43.1    NY\n",
       "19      40.7    NY\n",
       "21        43    NY\n",
       "25      40.7    NY\n",
       "54      40.8    NY\n",
       "55      40.7    NY\n",
       "160     40.7    NY\n",
       "201     42.2    NY\n",
       "210     40.7    NY\n",
       "232     40.8    NY\n",
       "238     40.9    NY\n",
       "243     40.7    NY\n",
       "254     40.9    NY\n",
       "288     43.1    NY\n",
       "290     40.7    NY\n",
       "294     40.8    NY\n",
       "310     40.7    NY\n",
       "339     40.7    NY\n",
       "345     42.7    NY\n",
       "352     40.8    NY\n",
       "378     40.7    NY\n",
       "431     41.1    NY\n",
       "483     40.7    NY\n",
       "566     43.1    NY\n",
       "567     40.7    NY\n",
       "587     41.3    NY\n",
       "600       43    NY\n",
       "608     40.8    NY\n",
       "618     40.8    NY"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_state_NY = df_5[df_5['STATE'] == 'NY'][['LATITUDE', 'STATE']]\n",
    "df_state_NY.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_state_NY_30 = df_state_NY.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('O')"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_state_NY_30['LATITUDE'].dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_state_NY_30_float = df_state_NY_30.astype({'LATITUDE': 'float64'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_state_NY_30_float['LATITUDE'].dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LATITUDE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>30.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>41.276667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.924637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>40.700000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>40.700000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>40.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>41.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>43.100000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        LATITUDE\n",
       "count  30.000000\n",
       "mean   41.276667\n",
       "std     0.924637\n",
       "min    40.700000\n",
       "25%    40.700000\n",
       "50%    40.800000\n",
       "75%    41.250000\n",
       "max    43.100000"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_state_NY_30_float.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "LATITUDE    41.3\n",
       "dtype: float64"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_latitude_NY = round(df_state_NY_30_float.mean(axis=0), 1)\n",
    "mean_latitude_NY\n",
    "\n",
    "# https://stackoverflow.com/questions/31037298/pandas-get-column-average-mean"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use this calculated mean value of LATITUDE = 41.3 for the cells that have STATE = NY and have 'no_data' under LATITUDE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41.3"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Replace value in a cell.\n",
    "\n",
    "df_5_4 = df_5.loc[3325, 'LATITUDE'] = 41.3\n",
    "\n",
    "df_5_4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>COUNTRY_OF_BIRTH</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>STATE</th>\n",
       "      <th>CURRENT_RESIDENCE_DURATION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>no_data</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>GAZ:Switzerland</td>\n",
       "      <td>no_data</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3972</th>\n",
       "      <td>no_data</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>GAZ:Switzerland</td>\n",
       "      <td>no_data</td>\n",
       "      <td>More than one year ago</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     LATITUDE COUNTRY_OF_BIRTH          COUNTRY    STATE  \\\n",
       "863   no_data      Switzerland  GAZ:Switzerland  no_data   \n",
       "3972  no_data      Switzerland  GAZ:Switzerland  no_data   \n",
       "\n",
       "     CURRENT_RESIDENCE_DURATION  \n",
       "863      More than one year ago  \n",
       "3972     More than one year ago  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_5[df_5['LATITUDE'] == 'no_data'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Replacement worked for that cell."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>NONFOODALLERGIES_DRUG</th>\n",
       "      <th>NONFOODALLERGIES_BEESTINGS</th>\n",
       "      <th>AGE</th>\n",
       "      <th>TOT_MASS</th>\n",
       "      <th>QUINOLINE</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>MIGRAINE_FACTOR_2</th>\n",
       "      <th>MIGRAINE_FACTOR_3</th>\n",
       "      <th>TONSILS_REMOVED</th>\n",
       "      <th>...</th>\n",
       "      <th>WEIGHT_CHANGE</th>\n",
       "      <th>PRIMARY_CARB</th>\n",
       "      <th>DIABETES</th>\n",
       "      <th>ASTHMA</th>\n",
       "      <th>POOL_FREQUENCY</th>\n",
       "      <th>CAT</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>FOODALLERGIES_PEANUTS</th>\n",
       "      <th>CHICKENPOX</th>\n",
       "      <th>SEASONAL_ALLERGIES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 110 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [CARBOHYDRATE_PER, NONFOODALLERGIES_DRUG, NONFOODALLERGIES_BEESTINGS, AGE, TOT_MASS, QUINOLINE, BODY_SITE, MIGRAINE_FACTOR_2, MIGRAINE_FACTOR_3, TONSILS_REMOVED, MIGRAINE_FACTOR_1, ELEVATION, FLOSSING_FREQUENCY, PROTEIN_PER, CSECTION, COUNTRY_OF_BIRTH, MACRONUTRIENT_PCT_TOTAL, COSMETICS_FREQUENCY, GLUTEN, LONGITUDE, DOG, MIGRAINE_AGGRAVATION, DIABETES_MEDICATION, PENICILLIN, MIGRAINE_FREQUENCY, MIGRAINE_RELATIVES, PLANT_PER, MIGRAINE_PHOTOPHOBIA, SITE_SAMPLED, DRINKING_WATER_SOURCE, COUNTRY, ANTIBIOTIC_SELECT, MIGRAINE_NAUSEA, FOODALLERGIES_TREENUTS, MIGRAINE_AURA, FLU_VACCINE_DATE, PREGNANT, MIGRAINE, BMI, TEETHBRUSHING_FREQUENCY, MIGRAINE_MEDICATIONS, CONTRACEPTIVE, LIVINGWITH, RELATIONS, DIABETES_MEDICATIONS, SUPPLEMENTS, AGE_UNIT, EXERCISE_FREQUENCY, SULFA_DRUG, NONFOODALLERGIES_NO, ENV_MATTER, SKIN_CONDITION, CONDITIONS_MEDICATION, LAST_TRAVEL, HEIGHT_IN, ANTIBIOTIC_CONDITION, PET_CONTACT, NONFOODALLERGIES_SUN, EXERCISE_LOCATION, NITROIMIDAZOLE, DEODORANT_USE, FOODALLERGIES_SHELLFISH, SEX, PERCENTAGE_FROM_CARBS, FAT_PER, SHARED_HOUSING, SOFTENER, MIGRAINE_PHONOPHOBIA, SMOKING_FREQUENCY, WEIGHT_LBS, TANNING_BEDS, ALCOHOL_FREQUENCY, STATE, BODY_HABITAT, DOMINANT_HAND, PRIMARY_VEGETABLE, SPECIAL_RESTRICTIONS, CEPHALOSPORIN, SUPPLEMENTS_FIELDS, APPENDIX_REMOVED, CURRENT_RESIDENCE_DURATION, SLEEP_DURATION, DIET_TYPE, DIET_RESTRICTIONS, LACTOSE, HEIGHT_OR_LENGTH, ANTIBIOTIC_MEDS, IBD, TYPES_OF_PLANTS, MIGRAINEMEDS, FIBER_GRAMS, NAILS, RACE, PETS, MULTIVITAMIN, NONFOODALLERGIES_POISONIVY, ACNE_MEDICATION, ACNE_MEDICATION_OTC, ANIMAL_PER, FOODALLERGIES_OTHER, ...]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 110 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if Switzerland is mentioned in 'STATE' column.\n",
    "\n",
    "df_5[df_5['STATE'] == 'Switzerland']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "absolute number of unique entries in column: CA                    711\n",
      "CO                    324\n",
      "NY                    240\n",
      "MA                    197\n",
      "WA                    175\n",
      "                     ... \n",
      "BUSKERUD                1\n",
      "MA - MASSACHUSETTS      1\n",
      "ALASKA                  1\n",
      "MINNESOTA               1\n",
      "QUEBEC                  1\n",
      "Name: STATE, Length: 118, dtype: int64\n",
      "percentage of unique entries in column: CA                    0.17\n",
      "CO                    0.08\n",
      "NY                    0.06\n",
      "MA                    0.05\n",
      "WA                    0.04\n",
      "                      ... \n",
      "BUSKERUD              0.00\n",
      "MA - MASSACHUSETTS    0.00\n",
      "ALASKA                0.00\n",
      "MINNESOTA             0.00\n",
      "QUEBEC                0.00\n",
      "Name: STATE, Length: 118, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "# Show frequency ranking of unique entries.\n",
    "\n",
    "abs_in_col = df_5.STATE.value_counts()\n",
    "\n",
    "percentage_in_col = round(df_5.STATE.value_counts(normalize=True), 2)\n",
    "\n",
    "print(f'absolute number of unique entries in column: {abs_in_col}')\n",
    "\n",
    "\n",
    "print(f'percentage of unique entries in column: {percentage_in_col}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CO', 'CA', 'NY', 'MA', 'MN', 'VERMONT', 'FL', 'AK', 'PA', 'NC',\n",
       "       'MASSACHUSETTS', 'WI', 'NJ', 'OH', 'KY', 'TX', 'OR', '-', 'AZ',\n",
       "       'MO', 'BANKOK', 'MD', 'GA', 'CT', 'VA', 'IN', 'NH', 'ME',\n",
       "       'unknown', 'SC', 'IL', 'KS', 'NM', 'UT', 'WA', 'ILLINOIS',\n",
       "       'LONDON', 'MB', 'ND', 'HI', 'BRITISH COLUMBIA', 'LA', 'OK',\n",
       "       'TEXAS', 'BC', 'ALASKA', 'CALIFORNIA', 'IA', 'no_data', 'TN',\n",
       "       'WASHINGTON', 'CEREDIGION', 'GREATER LONDON', 'CAMDEN TOWN', 'NV',\n",
       "       'DC', 'MI', 'COLORADO', 'WV', 'AL', 'AR', 'ID', 'VIRGINIA', 'AB',\n",
       "       'TENNESSEE', 'VT', 'MARYLAND', 'SD', 'NE', 'MT', 'MS', 'NEW YORK',\n",
       "       'LONDON, CITY OF', 'WY', 'Bangkok', 'OHIO', 'CA - CALIFORNIA',\n",
       "       'QC', 'RI', 'ONTARIO', 'NSW', 'ROTHERHITHE', 'NONE', 'MAINE',\n",
       "       'OREGON', 'WISCONSIN', 'CONNECTICUT', 'SK', 'BUCKINGHAMSHIRE',\n",
       "       'WESTERN AUSTRALIA', 'FLORIDA', 'HAWAII', 'GA.', 'NEW MEXICO',\n",
       "       'VIC', 'MA - MASSACHUSETTS', 'COLORADO (CO)', 'SOUTH CAROLINA',\n",
       "       'KENTUCKY', 'PR', 'ON', 'MAZOWIECKIE', 'ARKANSAS', 'MISSOURI',\n",
       "       'NC NORTH CAROLINA', 'ZURICH', 'MINNESOTA', 'BEDFORDSHIRE',\n",
       "       'GEORGOA', 'OH- OHIO', 'THURINGIA', 'TX-TEXAS', 'MICHIGAN',\n",
       "       'QUEBEC', 'WELLINGTON', 'MONTANA', 'GVA', 'BUSKERUD'], dtype=object)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show unique entries.\n",
    "\n",
    "df_5.STATE.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>STATE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2603</th>\n",
       "      <td>47.7</td>\n",
       "      <td>ZURICH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     LATITUDE   STATE\n",
       "2603     47.7  ZURICH"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Enter ZURICH instead of no_data\n",
    "\n",
    "df_state_ZURICH = df_5[df_5['STATE'] == 'ZURICH'][['LATITUDE', 'STATE']]\n",
    "df_state_ZURICH.head(8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47.7\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "47.7"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Replace value in a cell.\n",
    "\n",
    "df_5_5 = df_5.loc[863, 'LATITUDE'] = 47.7\n",
    "\n",
    "df_5_6 = df_5.loc[3972, 'LATITUDE'] = 47.7\n",
    "\n",
    "print(df_5_5)\n",
    "\n",
    "df_5_6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>COUNTRY_OF_BIRTH</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>STATE</th>\n",
       "      <th>CURRENT_RESIDENCE_DURATION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [LATITUDE, COUNTRY_OF_BIRTH, COUNTRY, STATE, CURRENT_RESIDENCE_DURATION]\n",
       "Index: []"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_5[df_5['LATITUDE'] == 'no_data'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Replacement worked for those 2 cells."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('O')"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_5['LATITUDE'].dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_6 = df_5.astype({'LATITUDE': 'float64'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_6['LATITUDE'].dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**df_6 will be used for modelling.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Machine Learning Modelling**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clean = df_6 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4232, 110)"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clean.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the value range of a few specially important columns. Here I am choosing columns relevant for diabetes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12.,\n",
       "       13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,\n",
       "       26., 27., 28., 29., 30., 31., 32., 33., 34., 35., 36., 37., 38.,\n",
       "       39., 40., 41., 42., 43., 44., 45., 46., 47., 48., 49., 50., 51.,\n",
       "       52., 53., 54., 55., 56., 57., 58., 59., 60., 61., 62., 63., 64.,\n",
       "       65., 66., 67., 68., 69., 70., 71., 72., 73., 74., 75., 76., 77.,\n",
       "       78., 79., 80., 81., 82., 83., 84., 88., 91., 93.])"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show unique values in a column.\n",
    "# Age range.\n",
    "\n",
    "AGE_unique = np.sort(df_6.AGE.unique())\n",
    "AGE_unique\n",
    "\n",
    "# https://stackoverflow.com/questions/32072076/find-the-unique-values-in-a-column-and-then-sort-them/32072252"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['I do not have diabetes', 'Type I', 'Type II', 'no_data',\n",
       "       'unknown'], dtype=object)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DIABETES_unique = np.sort(df.DIABETES.unique())\n",
    "DIABETES_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['0', '1', '10', '11', '11.4', '12', '13', '13.5', '14', '15', '16',\n",
       "       '17', '17.3', '17.6', '18', '19', '19.2', '19.9', '2', '2.9', '20',\n",
       "       '20.1', '21', '21.3', '22', '22.2', '23', '24', '25', '25.4', '26',\n",
       "       '26.8', '27', '27.8', '28', '28.2', '28.3', '29', '29.2', '3',\n",
       "       '30', '31', '31.4', '31.5', '31.7', '32', '33', '34', '34.2', '35',\n",
       "       '35.01', '35.6', '36', '36.2', '36.3', '36.6', '36.61', '37', '38',\n",
       "       '38.1', '38.7', '38.81', '39', '39.1', '39.4', '39.6', '39.8', '4',\n",
       "       '4.2', '40', '40.2', '41', '41.6', '41.8', '42', '42.1', '42.2',\n",
       "       '42.3', '42.6', '43', '43.3', '43.4', '43.7', '44', '44.3', '44.5',\n",
       "       '44.7', '45', '45.4', '45.5', '46', '46.4', '46.5', '46.7', '47',\n",
       "       '47.1', '47.2', '47.5', '47.6', '47.7', '48', '48.3', '48.5',\n",
       "       '48.7', '49', '49.6', '49.8', '5', '50', '50.2', '50.8', '51',\n",
       "       '51.7', '52', '53', '53.5', '54', '55', '55.5', '56', '56.1', '57',\n",
       "       '57.7', '58', '59', '59.7', '6', '60', '61', '61.76', '62', '63',\n",
       "       '63.2', '64', '65', '66', '67', '68', '69', '7', '70', '71', '72',\n",
       "       '73', '74', '75', '76', '77', '79', '8', '80', '83', '85', '88',\n",
       "       '9', '90', '95', 'no_data', 'unknown'], dtype=object)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CARBOHYDRATE_PER_unique = np.sort(df.CARBOHYDRATE_PER.unique())\n",
    "CARBOHYDRATE_PER_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['0', '10', '11', '11.7', '12', '13', '13.9', '14', '15', '16',\n",
       "       '17', '17.7', '18', '19', '19.01', '2', '20', '21', '22', '22.1',\n",
       "       '23', '24', '25', '25.3', '25.5', '26', '26.8', '26.9', '27', '28',\n",
       "       '28.3', '28.6', '29', '29.1', '29.4', '30', '30.5', '30.7', '30.9',\n",
       "       '31', '31.2', '31.4', '31.6', '31.9', '32', '32.4', '32.7', '32.8',\n",
       "       '33', '33.7', '33.8', '33.9', '34', '34.2', '34.5', '34.6', '35',\n",
       "       '35.1', '35.4', '35.6', '36', '36.1', '36.2', '36.3', '36.5',\n",
       "       '36.6', '37', '37.4', '37.5', '37.9', '38', '38.1', '38.3', '38.5',\n",
       "       '38.8', '39', '39.1', '39.3', '39.5', '39.9', '4', '40', '40.3',\n",
       "       '40.5', '40.94', '41', '41.2', '41.3', '41.4', '41.7', '42',\n",
       "       '42.4', '42.6', '43', '43.2', '43.3', '44', '44.1', '44.8', '44.9',\n",
       "       '45', '45.12', '45.3', '46', '46.58', '46.7', '47', '47.1', '47.2',\n",
       "       '47.9', '48', '48.7', '49', '5', '50', '51', '51.3', '52', '53',\n",
       "       '53.5', '54', '54.7', '55', '55.3', '55.8', '56', '56.1', '57',\n",
       "       '58', '58.9', '59', '6', '60', '61', '62', '62.2', '63', '63.3',\n",
       "       '64', '65', '65.8', '66', '67', '68', '69', '7', '70', '70.6',\n",
       "       '71', '72', '73', '74', '75', '75.3', '76', '78', '79', '8', '80',\n",
       "       '81', '85', '9', 'no_data', 'unknown'], dtype=object)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FAT_PER_unique = np.sort(df.FAT_PER.unique())\n",
    "FAT_PER_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-41.3, -37.9, -37.8, -37.7, -36.8, -33.9, -32. , -31.9, -31.8,\n",
       "       -30.6, -30.5,  13.7,  18.4,  19.6,  20.8,  20.9,  21.3,  21.4,\n",
       "        21.9,  22.1,  22.4,  24.7,  25.5,  25.7,  25.8,  25.9,  26.1,\n",
       "        26.2,  26.3,  26.4,  26.7,  27. ,  27.1,  27.2,  27.5,  27.6,\n",
       "        27.7,  27.9,  28. ,  28.1,  28.2,  28.3,  28.5,  28.6,  28.7,\n",
       "        28.8,  29.5,  29.6,  29.7,  29.8,  29.9,  30. ,  30.1,  30.2,\n",
       "        30.3,  30.4,  30.5,  30.6,  30.7,  31.1,  31.2,  31.3,  31.5,\n",
       "        31.9,  32. ,  32.1,  32.2,  32.3,  32.4,  32.5,  32.6,  32.7,\n",
       "        32.8,  32.9,  33. ,  33.1,  33.2,  33.3,  33.4,  33.5,  33.6,\n",
       "        33.7,  33.8,  33.9,  34. ,  34.1,  34.2,  34.3,  34.4,  34.5,\n",
       "        34.6,  34.7,  34.8,  34.9,  35. ,  35.1,  35.2,  35.3,  35.4,\n",
       "        35.5,  35.6,  35.7,  35.8,  35.9,  36. ,  36.1,  36.2,  36.3,\n",
       "        36.4,  36.5,  36.6,  36.7,  36.8,  36.9,  37. ,  37.1,  37.2,\n",
       "        37.3,  37.4,  37.5,  37.6,  37.7,  37.8,  37.9,  38. ,  38.1,\n",
       "        38.2,  38.3,  38.4,  38.5,  38.6,  38.7,  38.8,  38.9,  39. ,\n",
       "        39.1,  39.2,  39.3,  39.4,  39.5,  39.6,  39.7,  39.8,  39.9,\n",
       "        40. ,  40.1,  40.2,  40.3,  40.4,  40.5,  40.6,  40.7,  40.8,\n",
       "        40.9,  41. ,  41.1,  41.2,  41.3,  41.4,  41.5,  41.6,  41.7,\n",
       "        41.8,  41.9,  42. ,  42.1,  42.2,  42.3,  42.4,  42.5,  42.6,\n",
       "        42.7,  42.8,  42.9,  43. ,  43.1,  43.2,  43.3,  43.4,  43.5,\n",
       "        43.6,  43.7,  43.8,  43.9,  44. ,  44.1,  44.2,  44.3,  44.4,\n",
       "        44.5,  44.6,  44.7,  44.8,  44.9,  45. ,  45.1,  45.2,  45.3,\n",
       "        45.4,  45.5,  45.6,  45.7,  45.8,  46. ,  46.2,  46.5,  46.6,\n",
       "        46.7,  46.8,  46.9,  47. ,  47.1,  47.2,  47.3,  47.4,  47.5,\n",
       "        47.6,  47.7,  47.8,  47.9,  48. ,  48.1,  48.2,  48.3,  48.4,\n",
       "        48.5,  48.8,  49. ,  49.2,  49.3,  49.4,  49.5,  49.9,  50. ,\n",
       "        50.5,  50.6,  50.7,  50.8,  50.9,  51. ,  51.1,  51.3,  51.4,\n",
       "        51.5,  51.6,  51.8,  52. ,  52.1,  52.2,  52.4,  52.5,  53.2,\n",
       "        53.5,  53.6,  54.2,  58.4,  58.5,  59.8,  59.9,  60.2,  61.2,\n",
       "        61.8,  64.8])"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LATITUDE_unique = np.sort(df_6.LATITUDE.unique())\n",
    "LATITUDE_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_6.loc[(df_6.FAT_PER == 'unknown'),'FAT_PER']='33.3'\n",
    "\n",
    "# https://kanoki.org/2019/07/17/pandas-how-to-replace-values-based-on-conditions/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_6.loc[(df_6.CARBOHYDRATE_PER == 'no_data'),'CARBOHYDRATE_PER']='33.3'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['0', '10', '11', '11.4', '12', '13', '13.5', '14', '15', '16',\n",
       "       '17', '17.3', '17.6', '18', '19', '19.2', '19.9', '2', '2.9', '20',\n",
       "       '20.1', '21', '21.3', '22', '22.2', '23', '24', '25', '25.4', '26',\n",
       "       '26.8', '27', '27.8', '28', '28.2', '28.3', '29', '29.2', '3',\n",
       "       '30', '31', '31.4', '31.5', '32', '33', '33.3', '34', '34.2', '35',\n",
       "       '35.01', '35.6', '36', '36.2', '36.3', '36.6', '36.61', '37', '38',\n",
       "       '38.1', '38.7', '38.81', '39', '39.1', '39.4', '39.6', '39.8', '4',\n",
       "       '4.2', '40', '40.2', '41', '41.6', '41.8', '42', '42.1', '42.2',\n",
       "       '42.3', '42.6', '43', '43.3', '43.4', '43.7', '44', '44.3', '44.5',\n",
       "       '44.7', '45', '45.4', '45.5', '46', '46.4', '46.5', '46.7', '47',\n",
       "       '47.1', '47.2', '47.5', '47.6', '47.7', '48', '48.3', '48.5',\n",
       "       '48.7', '49', '49.6', '49.8', '5', '50', '50.2', '50.8', '51',\n",
       "       '51.7', '52', '53', '53.5', '54', '55', '55.5', '56', '56.1', '57',\n",
       "       '57.7', '58', '59', '59.7', '6', '60', '61', '61.76', '62', '63',\n",
       "       '63.2', '64', '65', '66', '67', '68', '69', '7', '70', '71', '72',\n",
       "       '73', '74', '75', '76', '77', '79', '8', '80', '83', '85', '88',\n",
       "       '9', '90', '95'], dtype=object)"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CARBOHYDRATE_PER_unique = np.sort(df_6.CARBOHYDRATE_PER.unique())\n",
    "CARBOHYDRATE_PER_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_6.loc[(df_6.FAT_PER == 'unknown'),'FAT_PER']='33.3'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_6.loc[(df_6.FAT_PER == 'no_data'),'FAT_PER']='33.3'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['0', '10', '11', '11.7', '12', '13', '13.9', '14', '15', '16',\n",
       "       '17', '17.7', '18', '19', '19.01', '20', '21', '22', '22.1', '23',\n",
       "       '24', '25', '25.3', '25.5', '26', '26.8', '26.9', '27', '28',\n",
       "       '28.3', '28.6', '29', '29.1', '29.4', '30', '30.5', '30.7', '30.9',\n",
       "       '31', '31.2', '31.4', '31.6', '31.9', '32', '32.4', '32.7', '32.8',\n",
       "       '33', '33.3', '33.7', '33.8', '33.9', '34', '34.2', '34.5', '34.6',\n",
       "       '35', '35.1', '35.4', '35.6', '36', '36.1', '36.2', '36.3', '36.5',\n",
       "       '36.6', '37', '37.4', '37.5', '37.9', '38', '38.1', '38.3', '38.5',\n",
       "       '38.8', '39', '39.1', '39.3', '39.5', '39.9', '4', '40', '40.3',\n",
       "       '40.5', '40.94', '41', '41.2', '41.3', '41.4', '41.7', '42',\n",
       "       '42.4', '42.6', '43', '43.2', '43.3', '44', '44.1', '44.8', '44.9',\n",
       "       '45', '45.12', '45.3', '46', '46.58', '46.7', '47', '47.1', '47.2',\n",
       "       '47.9', '48', '48.7', '49', '5', '50', '51', '51.3', '52', '53',\n",
       "       '53.5', '54', '54.7', '55', '55.3', '55.8', '56', '56.1', '57',\n",
       "       '58', '58.9', '59', '6', '60', '61', '62', '62.2', '63', '63.3',\n",
       "       '64', '65', '65.8', '66', '67', '68', '69', '7', '70', '70.6',\n",
       "       '71', '72', '73', '74', '75', '75.3', '76', '78', '8', '80', '81',\n",
       "       '85', '9'], dtype=object)"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FAT_PER_unique = np.sort(df_6.FAT_PER.unique())\n",
    "FAT_PER_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('O')"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_6['CARBOHYDRATE_PER'].dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_7 = df_6.astype({'CARBOHYDRATE_PER': 'float64'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_7['LATITUDE'].dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('O')"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_6['FAT_PER'].dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_7 = df_6.astype({'FAT_PER': 'float64'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_7['FAT_PER'].dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**df_7 will be used for modelling.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clean = df_7"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Target: DIABETES**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "# These are all features that can potentially be used.\n",
    "\n",
    "#X = df_clean[['CARBOHYDRATE_PER',\n",
    " 'NONFOODALLERGIES_DRUG',\n",
    " 'NONFOODALLERGIES_BEESTINGS',\n",
    " 'AGE',\n",
    " 'TOT_MASS',\n",
    " 'QUINOLINE',\n",
    " 'BODY_SITE',\n",
    " 'MIGRAINE_FACTOR_2',\n",
    " 'MIGRAINE_FACTOR_3',\n",
    " 'TONSILS_REMOVED',\n",
    " 'MIGRAINE_FACTOR_1',\n",
    " 'ELEVATION',\n",
    " 'FLOSSING_FREQUENCY',\n",
    " 'PROTEIN_PER',\n",
    " 'CSECTION',\n",
    " 'COUNTRY_OF_BIRTH',\n",
    " 'MACRONUTRIENT_PCT_TOTAL',\n",
    " 'COSMETICS_FREQUENCY',\n",
    " 'GLUTEN',\n",
    " 'LONGITUDE',\n",
    " 'DOG',\n",
    " 'MIGRAINE_AGGRAVATION',\n",
    " 'DIABETES_MEDICATION',\n",
    " 'PENICILLIN',\n",
    " 'MIGRAINE_FREQUENCY',\n",
    " 'MIGRAINE_RELATIVES',\n",
    " 'PLANT_PER',\n",
    " 'MIGRAINE_PHOTOPHOBIA',\n",
    " 'SITE_SAMPLED',\n",
    " 'DRINKING_WATER_SOURCE',\n",
    " 'COUNTRY',\n",
    " 'ANTIBIOTIC_SELECT',\n",
    " 'MIGRAINE_NAUSEA',\n",
    " 'FOODALLERGIES_TREENUTS',\n",
    " 'MIGRAINE_AURA',\n",
    " 'FLU_VACCINE_DATE',\n",
    " 'PREGNANT',\n",
    " 'MIGRAINE',\n",
    " 'BMI',\n",
    " 'TEETHBRUSHING_FREQUENCY',\n",
    " 'MIGRAINE_MEDICATIONS',\n",
    " 'CONTRACEPTIVE',\n",
    " 'LIVINGWITH',\n",
    " 'RELATIONS',\n",
    " 'DIABETES_MEDICATIONS',\n",
    " 'SUPPLEMENTS',\n",
    " 'AGE_UNIT',\n",
    " 'EXERCISE_FREQUENCY',\n",
    " 'SULFA_DRUG',\n",
    " 'NONFOODALLERGIES_NO',\n",
    " 'ENV_MATTER',\n",
    " 'SKIN_CONDITION',\n",
    " 'CONDITIONS_MEDICATION',\n",
    " 'LAST_TRAVEL',\n",
    " 'HEIGHT_IN',\n",
    " 'ANTIBIOTIC_CONDITION',\n",
    " 'PET_CONTACT',\n",
    " 'NONFOODALLERGIES_SUN',\n",
    " 'EXERCISE_LOCATION',\n",
    " 'NITROIMIDAZOLE',\n",
    " 'DEODORANT_USE',\n",
    " 'FOODALLERGIES_SHELLFISH',\n",
    " 'SEX',\n",
    " 'PERCENTAGE_FROM_CARBS',\n",
    " 'FAT_PER',\n",
    " 'SHARED_HOUSING',\n",
    " 'SOFTENER',\n",
    " 'MIGRAINE_PHONOPHOBIA',\n",
    " 'SMOKING_FREQUENCY',\n",
    " 'WEIGHT_LBS',\n",
    " 'TANNING_BEDS',\n",
    " 'ALCOHOL_FREQUENCY',\n",
    " 'STATE',\n",
    " 'BODY_HABITAT',\n",
    " 'DOMINANT_HAND',\n",
    " 'PRIMARY_VEGETABLE',\n",
    " 'SPECIAL_RESTRICTIONS',\n",
    " 'CEPHALOSPORIN',\n",
    " 'SUPPLEMENTS_FIELDS',\n",
    " 'APPENDIX_REMOVED',\n",
    " 'CURRENT_RESIDENCE_DURATION',\n",
    " 'SLEEP_DURATION',\n",
    " 'DIET_TYPE',\n",
    " 'DIET_RESTRICTIONS',\n",
    " 'LACTOSE',\n",
    " 'HEIGHT_OR_LENGTH',\n",
    " 'ANTIBIOTIC_MEDS',\n",
    " 'IBD',\n",
    " 'TYPES_OF_PLANTS',\n",
    " 'MIGRAINEMEDS',\n",
    " 'FIBER_GRAMS',\n",
    " 'NAILS',\n",
    " 'RACE',\n",
    " 'PETS',\n",
    " 'MULTIVITAMIN',\n",
    " 'NONFOODALLERGIES_POISONIVY',\n",
    " 'ACNE_MEDICATION',\n",
    " 'ACNE_MEDICATION_OTC',\n",
    " 'ANIMAL_PER',\n",
    " 'FOODALLERGIES_OTHER',\n",
    " 'WEIGHT_CHANGE',\n",
    " 'PRIMARY_CARB',\n",
    " 'ASTHMA',\n",
    " 'POOL_FREQUENCY',\n",
    " 'CAT',\n",
    " 'LATITUDE',\n",
    " 'FOODALLERGIES_PEANUTS',\n",
    " 'CHICKENPOX',\n",
    " 'SEASONAL_ALLERGIES']] # independent variables or features\n",
    "\n",
    "#y = df_clean['DIABETES'] # dependent variable or target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {},
   "outputs": [],
   "source": [
    "# store clean data in X and y.\n",
    "\n",
    "X = df_clean[['AGE',\n",
    " 'LATITUDE',            \n",
    " 'CARBOHYDRATE_PER',\n",
    " 'FAT_PER']] # independent variables or features\n",
    "\n",
    "y = df_clean['DIABETES'] # dependent variable or target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGE</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>FAT_PER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>51.0</td>\n",
       "      <td>39.7</td>\n",
       "      <td>55</td>\n",
       "      <td>30.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>35.0</td>\n",
       "      <td>32.7</td>\n",
       "      <td>56</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>56.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>16</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>37.0</td>\n",
       "      <td>40.7</td>\n",
       "      <td>33.3</td>\n",
       "      <td>33.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>72.0</td>\n",
       "      <td>43.1</td>\n",
       "      <td>61</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    AGE  LATITUDE CARBOHYDRATE_PER  FAT_PER\n",
       "0  51.0      39.7               55     30.0\n",
       "1  35.0      32.7               56     21.0\n",
       "2  56.0      40.0               16     67.0\n",
       "3  37.0      40.7             33.3     33.3\n",
       "4  72.0      43.1               61     23.0"
      ]
     },
     "execution_count": 277,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4232, 4)"
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    I do not have diabetes\n",
       "1    I do not have diabetes\n",
       "2    I do not have diabetes\n",
       "3    I do not have diabetes\n",
       "4    I do not have diabetes\n",
       "Name: DIABETES, dtype: object"
      ]
     },
     "execution_count": 279,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4232,)"
      ]
     },
     "execution_count": 280,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [],
   "source": [
    "# splitting our data into train and test sets\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=19, stratify = y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2835, 4)"
      ]
     },
     "execution_count": 282,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# size of training data\n",
    "\n",
    "X_train.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1397, 4)"
      ]
     },
     "execution_count": 283,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# size of test data\n",
    "\n",
    "X_test.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGE</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>FAT_PER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1958</th>\n",
       "      <td>43.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>40</td>\n",
       "      <td>30.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906</th>\n",
       "      <td>32.0</td>\n",
       "      <td>29.6</td>\n",
       "      <td>60</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>986</th>\n",
       "      <td>56.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>25</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>46.0</td>\n",
       "      <td>47.5</td>\n",
       "      <td>30</td>\n",
       "      <td>30.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3327</th>\n",
       "      <td>32.0</td>\n",
       "      <td>39.1</td>\n",
       "      <td>59</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       AGE  LATITUDE CARBOHYDRATE_PER  FAT_PER\n",
       "1958  43.0      43.0               40     30.0\n",
       "906   32.0      29.6               60     20.0\n",
       "986   56.0      38.0               25     35.0\n",
       "1202  46.0      47.5               30     30.0\n",
       "3327  32.0      39.1               59     21.0"
      ]
     },
     "execution_count": 284,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1958    I do not have diabetes\n",
       "906     I do not have diabetes\n",
       "986     I do not have diabetes\n",
       "1202    I do not have diabetes\n",
       "3327                    Type I\n",
       "Name: DIABETES, dtype: object"
      ]
     },
     "execution_count": 285,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_train.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGE</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>FAT_PER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1649</th>\n",
       "      <td>43.0</td>\n",
       "      <td>33.9</td>\n",
       "      <td>50.2</td>\n",
       "      <td>36.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2042</th>\n",
       "      <td>35.0</td>\n",
       "      <td>41.7</td>\n",
       "      <td>34</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3534</th>\n",
       "      <td>21.0</td>\n",
       "      <td>40.7</td>\n",
       "      <td>56</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1902</th>\n",
       "      <td>46.0</td>\n",
       "      <td>47.5</td>\n",
       "      <td>30</td>\n",
       "      <td>30.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3531</th>\n",
       "      <td>45.0</td>\n",
       "      <td>30.2</td>\n",
       "      <td>75</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       AGE  LATITUDE CARBOHYDRATE_PER  FAT_PER\n",
       "1649  43.0      33.9             50.2     36.1\n",
       "2042  35.0      41.7               34     18.0\n",
       "3534  21.0      40.7               56     27.0\n",
       "1902  46.0      47.5               30     30.0\n",
       "3531  45.0      30.2               75     15.0"
      ]
     },
     "execution_count": 286,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_test.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1649    I do not have diabetes\n",
       "2042    I do not have diabetes\n",
       "3534    I do not have diabetes\n",
       "1902    I do not have diabetes\n",
       "3531    I do not have diabetes\n",
       "Name: DIABETES, dtype: object"
      ]
     },
     "execution_count": 287,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_test.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Logistic Regression model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.linear_model import LogisticRegression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train accuracy: 0.9488536155202821\n",
      "test accuracy: 0.9491768074445239\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):\n",
      "STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.\n",
      "\n",
      "Increase the number of iterations (max_iter) or scale the data as shown in:\n",
      "    https://scikit-learn.org/stable/modules/preprocessing.html\n",
      "Please also refer to the documentation for alternative solver options:\n",
      "    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression\n",
      "  n_iter_i = _check_optimize_result(\n"
     ]
    }
   ],
   "source": [
    "# instantiate a logistic regression\n",
    "logreg_DIABETES = LogisticRegression()\n",
    "\n",
    "# fit the model to train set only\n",
    "logreg_DIABETES.fit(X_train, y_train)\n",
    "\n",
    "# score the model\n",
    "print(f\"train accuracy: {logreg_DIABETES.score(X_train, y_train)}\")\n",
    "print(f\"test accuracy: {logreg_DIABETES.score(X_test, y_test)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "I do not have diabetes    4016\n",
       "no_data                     89\n",
       "Type II                     78\n",
       "Type I                      34\n",
       "unknown                     15\n",
       "Name: DIABETES, dtype: int64"
      ]
     },
     "execution_count": 289,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show frequency of distinct values in a column.\n",
    "\n",
    "df_7['DIABETES'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4232,)"
      ]
     },
     "execution_count": 290,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_7['DIABETES'].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Add Scaling**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import scalar.\n",
    "from sklearn.preprocessing import StandardScaler \n",
    "\n",
    "# The scaler is FIT ON THE TRAINING DATA ONLY \n",
    "\n",
    "# 2. make a scaler\n",
    "ss_scaler = StandardScaler()\n",
    "\n",
    "# 3. fit the scaler\n",
    "ss_scaler.fit(X_train)\n",
    "\n",
    "# 4. transform the data\n",
    "X_train_scaled = ss_scaler.transform(X_train)\n",
    "X_test_scaled = ss_scaler.transform(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train accuracy: 0.9488536155202821\n",
      "test accuracy: 0.9491768074445239\n"
     ]
    }
   ],
   "source": [
    "# instantiate a logistic regression\n",
    "logreg_scaled_DIABETES = LogisticRegression()\n",
    "\n",
    "# fit the model to train set only\n",
    "logreg_scaled_DIABETES.fit(X_train_scaled, y_train)\n",
    "\n",
    "# score the model\n",
    "print(f\"train accuracy: {logreg_scaled_DIABETES.score(X_train_scaled, y_train)}\")\n",
    "print(f\"test accuracy: {logreg_scaled_DIABETES.score(X_test_scaled, y_test)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No improvement of accuracy scores."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Adding (C=0.01) as a definition of a hyperparameter**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train accuracy: 0.9488536155202821\n",
      "test accuracy: 0.9491768074445239\n"
     ]
    }
   ],
   "source": [
    "# instantiate a logistic regression\n",
    "logreg_scaled_DIABETES = LogisticRegression(C=1)\n",
    "# C=1 is default value.\n",
    "\n",
    "# fit the model to train set only\n",
    "logreg_scaled_DIABETES.fit(X_train_scaled, y_train)\n",
    "\n",
    "# score the model\n",
    "print(f\"train accuracy: {logreg_scaled_DIABETES.score(X_train_scaled, y_train)}\")\n",
    "print(f\"test accuracy: {logreg_scaled_DIABETES.score(X_test_scaled, y_test)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No improvement of accuracy scores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train accuracy: 0.9488536155202821\n",
      "test accuracy: 0.9491768074445239\n"
     ]
    }
   ],
   "source": [
    "# instantiate a logistic regression\n",
    "logreg_scaled_DIABETES = LogisticRegression(C=0.001)\n",
    "\n",
    "# fit the model to train set only\n",
    "logreg_scaled_DIABETES.fit(X_train_scaled, y_train)\n",
    "\n",
    "# score the model\n",
    "print(f\"train accuracy: {logreg_scaled_DIABETES.score(X_train_scaled, y_train)}\")\n",
    "print(f\"test accuracy: {logreg_scaled_DIABETES.score(X_test_scaled, y_test)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No improvement of accuracy scores."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The accuracy score on the test set is 0.949."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coefficients: [-0.03618958  0.00222874  0.01805508  0.0161957 ]\n"
     ]
    }
   ],
   "source": [
    "# Coefficient: it returns an array with one spot for each feature.\n",
    "\n",
    "coefficient = logreg_scaled_DIABETES.coef_[0]\n",
    "\n",
    "print(f'Coefficients: {coefficient}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4,)"
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coefficient.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AGE', 'LATITUDE', 'CARBOHYDRATE_PER', 'FAT_PER']"
      ]
     },
     "execution_count": 297,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(X_train.columns.values) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        precision    recall  f1-score   support\n",
      "\n",
      "I do not have diabetes       0.95      1.00      0.97      1326\n",
      "                Type I       0.00      0.00      0.00        11\n",
      "               Type II       0.00      0.00      0.00        26\n",
      "               no_data       0.00      0.00      0.00        29\n",
      "               unknown       0.00      0.00      0.00         5\n",
      "\n",
      "              accuracy                           0.95      1397\n",
      "             macro avg       0.19      0.20      0.19      1397\n",
      "          weighted avg       0.90      0.95      0.92      1397\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "\n",
    "predicted = logreg_scaled_DIABETES.predict(X_test_scaled)\n",
    "report = classification_report(y_test, predicted)\n",
    "print(report)\n",
    "\n",
    "# https://machinelearningmastery.com/metrics-evaluate-machine-learning-algorithms-python/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **DecisionTreeClassifier**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.tree import DecisionTreeClassifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The TRAIN classification accuracy is:  0.9982363315696648\n",
      "The TEST classification accuracy is:  0.9377236936292055\n"
     ]
    }
   ],
   "source": [
    "#Fit to the training data\n",
    "DT_model_scaled_DIABETES = DecisionTreeClassifier()\n",
    "DT_model_scaled_DIABETES.fit(X_train_scaled, y_train)\n",
    "\n",
    "print(f\"The TRAIN classification accuracy is:  {DT_model_scaled_DIABETES.score(X_train_scaled,y_train)}\")\n",
    "print(f\"The TEST classification accuracy is:  {DT_model_scaled_DIABETES.score(X_test_scaled,y_test)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The accuracy score on the test set is 0.938."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Confusion matrix**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Found input variables with inconsistent numbers of samples: [1397, 2835]",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-237-d845150dfa68>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# Generate confusion matrix\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mcf_matrix\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconfusion_matrix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_test\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0mcf_matrix\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36minner_f\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     71\u001b[0m                           FutureWarning)\n\u001b[1;32m     72\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0marg\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marg\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparameters\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 73\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     74\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0minner_f\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     75\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py\u001b[0m in \u001b[0;36mconfusion_matrix\u001b[0;34m(y_true, y_pred, labels, sample_weight, normalize)\u001b[0m\n\u001b[1;32m    274\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    275\u001b[0m     \"\"\"\n\u001b[0;32m--> 276\u001b[0;31m     \u001b[0my_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_check_targets\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    277\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0my_type\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m\"binary\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"multiclass\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    278\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"%s is not supported\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0my_type\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py\u001b[0m in \u001b[0;36m_check_targets\u001b[0;34m(y_true, y_pred)\u001b[0m\n\u001b[1;32m     79\u001b[0m     \u001b[0my_pred\u001b[0m \u001b[0;34m:\u001b[0m \u001b[0marray\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mindicator\u001b[0m \u001b[0mmatrix\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     80\u001b[0m     \"\"\"\n\u001b[0;32m---> 81\u001b[0;31m     \u001b[0mcheck_consistent_length\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     82\u001b[0m     \u001b[0mtype_true\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtype_of_target\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     83\u001b[0m     \u001b[0mtype_pred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtype_of_target\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36mcheck_consistent_length\u001b[0;34m(*arrays)\u001b[0m\n\u001b[1;32m    254\u001b[0m     \u001b[0muniques\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlengths\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    255\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muniques\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 256\u001b[0;31m         raise ValueError(\"Found input variables with inconsistent numbers of\"\n\u001b[0m\u001b[1;32m    257\u001b[0m                          \" samples: %r\" % [int(l) for l in lengths])\n\u001b[1;32m    258\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Found input variables with inconsistent numbers of samples: [1397, 2835]"
     ]
    }
   ],
   "source": [
    "# Get class predictions\n",
    "y_pred = logreg_scaled_DIABETES.predict(X_train_scaled)\n",
    "\n",
    "# Generate confusion matrix\n",
    "cf_matrix = confusion_matrix(y_test, y_pred)\n",
    "cf_matrix\n",
    "\n",
    "# Plot\n",
    "ConfusionMatrixDisplay(cf_matrix, display_labels=['0','1']).plot()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Found input variables with inconsistent numbers of samples: [1397, 2835]",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-238-b01c2225a1bd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Importing confusion matrix.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mconfusion\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconfusion_matrix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_test\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Confusion Matrix\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconfusion\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36minner_f\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     71\u001b[0m                           FutureWarning)\n\u001b[1;32m     72\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0marg\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marg\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparameters\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 73\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     74\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0minner_f\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     75\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py\u001b[0m in \u001b[0;36mconfusion_matrix\u001b[0;34m(y_true, y_pred, labels, sample_weight, normalize)\u001b[0m\n\u001b[1;32m    274\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    275\u001b[0m     \"\"\"\n\u001b[0;32m--> 276\u001b[0;31m     \u001b[0my_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_check_targets\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    277\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0my_type\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m\"binary\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"multiclass\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    278\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"%s is not supported\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0my_type\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py\u001b[0m in \u001b[0;36m_check_targets\u001b[0;34m(y_true, y_pred)\u001b[0m\n\u001b[1;32m     79\u001b[0m     \u001b[0my_pred\u001b[0m \u001b[0;34m:\u001b[0m \u001b[0marray\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mindicator\u001b[0m \u001b[0mmatrix\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     80\u001b[0m     \"\"\"\n\u001b[0;32m---> 81\u001b[0;31m     \u001b[0mcheck_consistent_length\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     82\u001b[0m     \u001b[0mtype_true\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtype_of_target\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     83\u001b[0m     \u001b[0mtype_pred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtype_of_target\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36mcheck_consistent_length\u001b[0;34m(*arrays)\u001b[0m\n\u001b[1;32m    254\u001b[0m     \u001b[0muniques\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlengths\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    255\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muniques\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 256\u001b[0;31m         raise ValueError(\"Found input variables with inconsistent numbers of\"\n\u001b[0m\u001b[1;32m    257\u001b[0m                          \" samples: %r\" % [int(l) for l in lengths])\n\u001b[1;32m    258\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Found input variables with inconsistent numbers of samples: [1397, 2835]"
     ]
    }
   ],
   "source": [
    "# Importing confusion matrix.\n",
    "\n",
    "confusion = confusion_matrix(y_test, y_pred)\n",
    "print('Confusion Matrix\\n')\n",
    "print(confusion)\n",
    "\n",
    "# https://towardsdatascience.com/confusion-matrix-for-your-multi-class-machine-learning-model-ff9aa3bf7826"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## IBD - Crohn's Disease and Ulcerative Colitis are the two main forms of Inflammatory Bowel Disease (IBD).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [],
   "source": [
    "# store clean data in X and y.\n",
    "\n",
    "X = df_clean[['AGE',\n",
    " 'LATITUDE',            \n",
    " 'CARBOHYDRATE_PER',\n",
    " 'FAT_PER']] # independent variables or features\n",
    "\n",
    "y = df_clean['IBD'] # dependent variable or target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGE</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>FAT_PER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>51.0</td>\n",
       "      <td>39.7</td>\n",
       "      <td>55</td>\n",
       "      <td>30.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>35.0</td>\n",
       "      <td>32.7</td>\n",
       "      <td>56</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>56.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>16</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>37.0</td>\n",
       "      <td>40.7</td>\n",
       "      <td>33.3</td>\n",
       "      <td>33.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>72.0</td>\n",
       "      <td>43.1</td>\n",
       "      <td>61</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    AGE  LATITUDE CARBOHYDRATE_PER  FAT_PER\n",
       "0  51.0      39.7               55     30.0\n",
       "1  35.0      32.7               56     21.0\n",
       "2  56.0      40.0               16     67.0\n",
       "3  37.0      40.7             33.3     33.3\n",
       "4  72.0      43.1               61     23.0"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4232, 4)"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    I do not have IBD\n",
       "1    I do not have IBD\n",
       "2    I do not have IBD\n",
       "3    I do not have IBD\n",
       "4    I do not have IBD\n",
       "Name: IBD, dtype: object"
      ]
     },
     "execution_count": 243,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4232,)"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [],
   "source": [
    "# splitting our data into train and test sets\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=19, stratify = y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2835, 4)"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# size of training data\n",
    "\n",
    "X_train.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1397, 4)"
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# size of test data\n",
    "\n",
    "X_test.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGE</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>FAT_PER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3213</th>\n",
       "      <td>31.0</td>\n",
       "      <td>34.1</td>\n",
       "      <td>20</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4163</th>\n",
       "      <td>37.0</td>\n",
       "      <td>37.4</td>\n",
       "      <td>43</td>\n",
       "      <td>39.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2186</th>\n",
       "      <td>65.0</td>\n",
       "      <td>20.8</td>\n",
       "      <td>64</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3879</th>\n",
       "      <td>60.0</td>\n",
       "      <td>32.8</td>\n",
       "      <td>33.3</td>\n",
       "      <td>33.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2778</th>\n",
       "      <td>54.0</td>\n",
       "      <td>40.7</td>\n",
       "      <td>49</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       AGE  LATITUDE CARBOHYDRATE_PER  FAT_PER\n",
       "3213  31.0      34.1               20     50.0\n",
       "4163  37.0      37.4               43     39.0\n",
       "2186  65.0      20.8               64     21.0\n",
       "3879  60.0      32.8             33.3     33.3\n",
       "2778  54.0      40.7               49     36.0"
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3213    I do not have IBD\n",
       "4163    I do not have IBD\n",
       "2186    I do not have IBD\n",
       "3879    I do not have IBD\n",
       "2778    I do not have IBD\n",
       "Name: IBD, dtype: object"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_train.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGE</th>\n",
       "      <th>LATITUDE</th>\n",
       "      <th>CARBOHYDRATE_PER</th>\n",
       "      <th>FAT_PER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4724</th>\n",
       "      <td>38.0</td>\n",
       "      <td>42.4</td>\n",
       "      <td>56</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2549</th>\n",
       "      <td>20.0</td>\n",
       "      <td>41.8</td>\n",
       "      <td>52</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2040</th>\n",
       "      <td>70.0</td>\n",
       "      <td>43.1</td>\n",
       "      <td>53</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1577</th>\n",
       "      <td>35.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>19.9</td>\n",
       "      <td>62.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>41.0</td>\n",
       "      <td>37.6</td>\n",
       "      <td>17</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       AGE  LATITUDE CARBOHYDRATE_PER  FAT_PER\n",
       "4724  38.0      42.4               56     25.0\n",
       "2549  20.0      41.8               52     34.0\n",
       "2040  70.0      43.1               53     26.0\n",
       "1577  35.0      40.0             19.9     62.2\n",
       "432   41.0      37.6               17     53.0"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_test.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4724    I do not have IBD\n",
       "2549    I do not have IBD\n",
       "2040    I do not have IBD\n",
       "1577    I do not have IBD\n",
       "432     I do not have IBD\n",
       "Name: IBD, dtype: object"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_test.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Logistic Regression model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.linear_model import LogisticRegression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train accuracy: 0.9199294532627866\n",
      "test accuracy: 0.9198282032927703\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):\n",
      "STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.\n",
      "\n",
      "Increase the number of iterations (max_iter) or scale the data as shown in:\n",
      "    https://scikit-learn.org/stable/modules/preprocessing.html\n",
      "Please also refer to the documentation for alternative solver options:\n",
      "    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression\n",
      "  n_iter_i = _check_optimize_result(\n"
     ]
    }
   ],
   "source": [
    "# instantiate a logistic regression\n",
    "logreg_IBD = LogisticRegression()\n",
    "\n",
    "# fit the model to train set only\n",
    "logreg_IBD.fit(X_train, y_train)\n",
    "\n",
    "# score the model\n",
    "print(f\"train accuracy: {logreg_IBD.score(X_train, y_train)}\")\n",
    "print(f\"test accuracy: {logreg_IBD.score(X_test, y_test)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "I do not have IBD     3893\n",
       "no_data                127\n",
       "Ulcerative colitis     116\n",
       "Crohn's disease         69\n",
       "unknown                 27\n",
       "Name: IBD, dtype: int64"
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show frequency of distinct values in a column.\n",
    "\n",
    "df_7['IBD'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4232,)"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_7['IBD'].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Add Scaling**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import scalar.\n",
    "from sklearn.preprocessing import StandardScaler \n",
    "\n",
    "# The scaler is FIT ON THE TRAINING DATA ONLY \n",
    "\n",
    "# 2. make a scaler\n",
    "ss_scaler = StandardScaler()\n",
    "\n",
    "# 3. fit the scaler\n",
    "ss_scaler.fit(X_train)\n",
    "\n",
    "# 4. transform the data\n",
    "X_train_scaled = ss_scaler.transform(X_train)\n",
    "X_test_scaled = ss_scaler.transform(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train accuracy: 0.9199294532627866\n",
      "test accuracy: 0.9198282032927703\n"
     ]
    }
   ],
   "source": [
    "# instantiate a logistic regression\n",
    "logreg_scaled_IBD = LogisticRegression()\n",
    "\n",
    "# fit the model to train set only\n",
    "logreg_scaled_IBD.fit(X_train_scaled, y_train)\n",
    "\n",
    "# score the model\n",
    "print(f\"train accuracy: {logreg_scaled_IBD.score(X_train_scaled, y_train)}\")\n",
    "print(f\"test accuracy: {logreg_scaled_IBD.score(X_test_scaled, y_test)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No improvement of accuracy scores."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Adding (C=0.01) as a definition of a hyperparameter**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train accuracy: 0.9199294532627866\n",
      "test accuracy: 0.9198282032927703\n"
     ]
    }
   ],
   "source": [
    "# instantiate a logistic regression\n",
    "logreg_scaled_IBD = LogisticRegression(C=1)\n",
    "# C=1 is default value.\n",
    "\n",
    "# fit the model to train set only\n",
    "logreg_scaled_IBD.fit(X_train_scaled, y_train)\n",
    "\n",
    "# score the model\n",
    "print(f\"train accuracy: {logreg_scaled_IBD.score(X_train_scaled, y_train)}\")\n",
    "print(f\"test accuracy: {logreg_scaled_IBD.score(X_test_scaled, y_test)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No improvement of accuracy scores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train accuracy: 0.9199294532627866\n",
      "test accuracy: 0.9198282032927703\n"
     ]
    }
   ],
   "source": [
    "# instantiate a logistic regression\n",
    "logreg_scaled_IBD = LogisticRegression(C=0.001)\n",
    "\n",
    "# fit the model to train set only\n",
    "logreg_scaled_IBD.fit(X_train_scaled, y_train)\n",
    "\n",
    "# score the model\n",
    "print(f\"train accuracy: {logreg_scaled_IBD.score(X_train_scaled, y_train)}\")\n",
    "print(f\"test accuracy: {logreg_scaled_IBD.score(X_test_scaled, y_test)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No improvement of accuracy scores."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The accuracy score on the test set is 0.920."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coefficients: [-0.00155367  0.00063138 -0.01024778  0.00301673]\n"
     ]
    }
   ],
   "source": [
    "# Coefficient: it returns an array with one spot for each feature.\n",
    "\n",
    "coefficient = logreg_scaled_IBD.coef_[0]\n",
    "\n",
    "print(f'Coefficients: {coefficient}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4,)"
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coefficient.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AGE', 'LATITUDE', 'CARBOHYDRATE_PER', 'FAT_PER']"
      ]
     },
     "execution_count": 263,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(X_train.columns.values) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "I do not have IBD     3893\n",
       "no_data                127\n",
       "Ulcerative colitis     116\n",
       "Crohn's disease         69\n",
       "unknown                 27\n",
       "Name: IBD, dtype: int64"
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show frequency of distinct values in a column.\n",
    "\n",
    "df_7['IBD'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    precision    recall  f1-score   support\n",
      "\n",
      "   Crohn's disease       0.00      0.00      0.00        23\n",
      " I do not have IBD       0.92      1.00      0.96      1285\n",
      "Ulcerative colitis       0.00      0.00      0.00        38\n",
      "           no_data       0.00      0.00      0.00        42\n",
      "           unknown       0.00      0.00      0.00         9\n",
      "\n",
      "          accuracy                           0.92      1397\n",
      "         macro avg       0.18      0.20      0.19      1397\n",
      "      weighted avg       0.85      0.92      0.88      1397\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "\n",
    "predicted = logreg_scaled_IBD.predict(X_test_scaled)\n",
    "report = classification_report(y_test, predicted)\n",
    "print(report)\n",
    "\n",
    "# https://machinelearningmastery.com/metrics-evaluate-machine-learning-algorithms-python/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **DecisionTreeClassifier**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.tree import DecisionTreeClassifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The TRAIN classification accuracy is:  0.9982363315696648\n",
      "The TEST classification accuracy is:  0.9377236936292055\n"
     ]
    }
   ],
   "source": [
    "#Fit to the training data\n",
    "DT_model_scaled_DIABETES = DecisionTreeClassifier()\n",
    "DT_model_scaled_DIABETES.fit(X_train_scaled, y_train)\n",
    "\n",
    "print(f\"The TRAIN classification accuracy is:  {DT_model_scaled_DIABETES.score(X_train_scaled,y_train)}\")\n",
    "print(f\"The TEST classification accuracy is:  {DT_model_scaled_DIABETES.score(X_test_scaled,y_test)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The accuracy score on the test set is 0.938."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4232, 110)"
      ]
     },
     "execution_count": 299,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clean.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
